{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary Statistics \n",
    "2023-08-08 ZD  \n",
    "\n",
    "This notebook will explore options to gather summary statistics and other reporting data calculated from data within processed grants data.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "import numpy as np\n",
    "import re\n",
    "from itertools import combinations\n",
    "\n",
    "\n",
    "# Method to import from parent directory\n",
    "import os\n",
    "import sys\n",
    "root_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "sys.path.append(root_dir)\n",
    "\n",
    "import config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load data from all Key Program grants output CSVs into a single dataframe for stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Project data pulled from: ../data/processed/2023-08-30/api-gathered-2023-08-30\n",
      "Reports will be output to: ../reports/2023-08-30/api-gathered-2023-08-30\n"
     ]
    }
   ],
   "source": [
    "# Define directory containing processed grant data\n",
    "processed_dir = '../'+config.PROCESSED_DIR\n",
    "\n",
    "# Define directory to store reports. Create if doesn't already exist\n",
    "reports_dir = '../'+config.REPORTS_DIR\n",
    "if not os.path.exists(reports_dir):\n",
    "    os.makedirs(reports_dir)\n",
    "\n",
    "# # Backup paths to allow for using same set of gathered data during notebook development\n",
    "# processed_dir = '../' + 'data/processed/2023-07-19/api-gathered-2023-08-25'\n",
    "# reports_dir = '../' + 'reports/2023-07-19/api-gathered-2023-08-25'\n",
    "if not os.path.exists(reports_dir):\n",
    "    os.makedirs(reports_dir)\n",
    "\n",
    "print(f\"Project data pulled from: {processed_dir}\")\n",
    "print(f\"Reports will be output to: {reports_dir}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load grants data\n",
    "grants_filename = os.path.join(processed_dir, 'project.tsv')\n",
    "df = pd.read_csv(grants_filename, sep='\\t')\n",
    "\n",
    "# Rename program.program_id column\n",
    "df.rename(columns={'program.program_id':'program'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quick detour to handle abstract text cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Project Summary - Overall The mortality rates of breast and pancreatic cancers are intrinsically tied to metastasis. In pancreatic cancer, the 5-year survival rate is only 9% and in ~70% pathological evaluations of the resected tumor, instances of venous invasion are found. Metastasis is a complex multi-step process involving cancer cells, local vasculature, and the surrounding microenvironment at multiple sites. Venous invasion in pancreatic cancer, in which cancer cells gain often invade the portal vein, is an early step in this process and provides the cells a direct path to the liver, the most common site of pancreatic cancer metastasis. As in pancreatic cancer, invasion past normal breast tissue barriers is critically tied to breast cancer outcomes. Most breast tumors can be surgically removed, and so mortality is closely tied to the extent of distant metastasis through lymphovascular invasion. The detection of lymphovascular invasion in a breast tumor correlates with poor prognosis and is not captured in current molecular analyses. The spatial organization of cancer cells, and cellular and stromal components of the tumor microenvironment near and far away from blood vessels is intrinsically three-dimensional, non-symmetric, and highly heterogeneous. In the TECH units of the Johns Hopkins Center for 3D Multiscale Cancer Imaging, we will develop a versatile 3D multiscale imaging method, CODA, which will allow us to probe the phenotypic heterogeneity of tumors from the multi-cm to the micron scale via multiplexing serial imaging. CODA can readily incorporate other imaging modalities to extract high cellular/molecular content from 3D samples. These include immunocytochemistry (CODA+IHC), immunofluorescence (CODA+IF), imaging mass cytometry (CODA+IMC) and spatial transcriptomics/proteomics (CODA+DBiT-seq). These proposed expanded versions of CODA offer a unique opportunity to produce new 3D multi-omic maps of human PDAC and breast tumors near and far from blood vessels. CODA and its integrated versions CODA+X will be tested in the RTB units of the Center in both human/mouse tissue samples and organoids, in breast and pancreatic cancer. Results from these test beds will provide novel mechanistic insights into venous invasion in breast and pancreatic cancer. Exploiting our extensive experience in data dissemination, we will make the large datasets and software produced by our CODA+X platforms and software widely available to the larger cancer research community. The units of the Center will be co-led by Johns Hopkins/Yale engineers, scientists and physicians. Substantial additional support will be provided by the Johns Hopkins U., the JH School of Medicine, the Institute for Convergence, the Department of Pathology, and the Whiting School of Engineering.'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pull one abstract to check for odd character encoding\n",
    "text = df.abstract_text[1]\n",
    "text"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lots of odd characters in here like \"\\xad\" and \"\\xa0\". Double-spaces were an issue before as well.  \n",
    "Make a function that can clean the column with regex and space replacement.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_abstract(text):\n",
    "    # Define a list of characters to be removed or replaced\n",
    "    chars_to_remove = ['\\xad']\n",
    "    chars_to_space = ['  ']\n",
    "\n",
    "    if isinstance(text, str):\n",
    "        # Remove unwanted characters (no space added)\n",
    "        for char in chars_to_remove:\n",
    "            text = text.replace(char, '')\n",
    "        # Replace unwanted characters with a space\n",
    "        for char in chars_to_space:\n",
    "            text = text.replace(char, ' ')\n",
    "\n",
    "        # Remove non-breaking spaces, newlines, and other unwanted characters\n",
    "        cleaned_text = re.sub(r'[\\s\\xa0]+', ' ', text).strip()\n",
    "        return cleaned_text\n",
    "        \n",
    "    # Return original value if it is NaN or float\n",
    "    else: \n",
    "        return text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Project Summary - Overall The mortality rates of breast and pancreatic cancers are intrinsically tied to metastasis. In pancreatic cancer, the 5-year survival rate is only 9% and in ~70% pathological evaluations of the resected tumor, instances of venous invasion are found. Metastasis is a complex multi-step process involving cancer cells, local vasculature, and the surrounding microenvironment at multiple sites. Venous invasion in pancreatic cancer, in which cancer cells gain often invade the portal vein, is an early step in this process and provides the cells a direct path to the liver, the most common site of pancreatic cancer metastasis. As in pancreatic cancer, invasion past normal breast tissue barriers is critically tied to breast cancer outcomes. Most breast tumors can be surgically removed, and so mortality is closely tied to the extent of distant metastasis through lymphovascular invasion. The detection of lymphovascular invasion in a breast tumor correlates with poor prognosis and is not captured in current molecular analyses. The spatial organization of cancer cells, and cellular and stromal components of the tumor microenvironment near and far away from blood vessels is intrinsically three-dimensional, non-symmetric, and highly heterogeneous. In the TECH units of the Johns Hopkins Center for 3D Multiscale Cancer Imaging, we will develop a versatile 3D multiscale imaging method, CODA, which will allow us to probe the phenotypic heterogeneity of tumors from the multi-cm to the micron scale via multiplexing serial imaging. CODA can readily incorporate other imaging modalities to extract high cellular/molecular content from 3D samples. These include immunocytochemistry (CODA+IHC), immunofluorescence (CODA+IF), imaging mass cytometry (CODA+IMC) and spatial transcriptomics/proteomics (CODA+DBiT-seq). These proposed expanded versions of CODA offer a unique opportunity to produce new 3D multi-omic maps of human PDAC and breast tumors near and far from blood vessels. CODA and its integrated versions CODA+X will be tested in the RTB units of the Center in both human/mouse tissue samples and organoids, in breast and pancreatic cancer. Results from these test beds will provide novel mechanistic insights into venous invasion in breast and pancreatic cancer. Exploiting our extensive experience in data dissemination, we will make the large datasets and software produced by our CODA+X platforms and software widely available to the larger cancer research community. The units of the Center will be co-led by Johns Hopkins/Yale engineers, scientists and physicians. Substantial additional support will be provided by the Johns Hopkins U., the JH School of Medicine, the Institute for Convergence, the Department of Pathology, and the Whiting School of Engineering.'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Try it on the example abstract from before\n",
    "clean = clean_abstract(text)\n",
    "clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean abstract column\n",
    "df['abstract_text'] = df['abstract_text'].apply(clean_abstract)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This project proposes an integrated suite of microscopy and data analysis advances that would enable quantitative, mechanistic analysis of immune-microenvironment dynamics in poor prognosis solid tumors. While immunotherapies are showing remarkable clinical responses in some advanced cancers, to date, their impact on many solid tumors has been modest. This is due, in part, to solid tumor microenvironments limiting the effectiveness of natural immune responses and immunotherapies. Yet, our understanding of the physical and molecular mechanisms governing T cell infiltration, distribution, and function in native tumor microenvironments remains extremely limited. As such, defining key T cell behaviors as a function of complex tumor microenvironments will identify design criteria that can be used to develop novel cell engineering strategies that optimize T cell-centric therapies for solid tumors. To this end the research theme of the U54 Center for Multiparametric Imaging of Tumor Immune Microenvironments (C-MITIE) is to define physical and molecular barriers to effective anti-tumor immunity and immunotherapies through advancement and development of state-of-the-art live cell and tissue optical imaging platforms and quantitative analyses. To achieve our goals, our framework brings together advanced optical imaging platforms, nano- and micro- fabrication, genome engineering, cancer immunology, and biophysical modeling. Thus, from this integrated effort we seek to define mechanisms of immune suppression that guide the development of next-generation cell-based immunotherapies.\n",
      "Project Summary - Overall The mortality rates of breast and pancreatic cancers are intrinsically tied to metastasis. In pancreatic cancer, the 5-year survival rate is only 9% and in ~70% pathological evaluations of the resected tumor, instances of venous invasion are found. Metastasis is a complex multi-step process involving cancer cells, local vasculature, and the surrounding microenvironment at multiple sites. Venous invasion in pancreatic cancer, in which cancer cells gain often invade the portal vein, is an early step in this process and provides the cells a direct path to the liver, the most common site of pancreatic cancer metastasis. As in pancreatic cancer, invasion past normal breast tissue barriers is critically tied to breast cancer outcomes. Most breast tumors can be surgically removed, and so mortality is closely tied to the extent of distant metastasis through lymphovascular invasion. The detection of lymphovascular invasion in a breast tumor correlates with poor prognosis and is not captured in current molecular analyses. The spatial organization of cancer cells, and cellular and stromal components of the tumor microenvironment near and far away from blood vessels is intrinsically three-dimensional, non-symmetric, and highly heterogeneous. In the TECH units of the Johns Hopkins Center for 3D Multiscale Cancer Imaging, we will develop a versatile 3D multiscale imaging method, CODA, which will allow us to probe the phenotypic heterogeneity of tumors from the multi-cm to the micron scale via multiplexing serial imaging. CODA can readily incorporate other imaging modalities to extract high cellular/molecular content from 3D samples. These include immunocytochemistry (CODA+IHC), immunofluorescence (CODA+IF), imaging mass cytometry (CODA+IMC) and spatial transcriptomics/proteomics (CODA+DBiT-seq). These proposed expanded versions of CODA offer a unique opportunity to produce new 3D multi-omic maps of human PDAC and breast tumors near and far from blood vessels. CODA and its integrated versions CODA+X will be tested in the RTB units of the Center in both human/mouse tissue samples and organoids, in breast and pancreatic cancer. Results from these test beds will provide novel mechanistic insights into venous invasion in breast and pancreatic cancer. Exploiting our extensive experience in data dissemination, we will make the large datasets and software produced by our CODA+X platforms and software widely available to the larger cancer research community. The units of the Center will be co-led by Johns Hopkins/Yale engineers, scientists and physicians. Substantial additional support will be provided by the Johns Hopkins U., the JH School of Medicine, the Institute for Convergence, the Department of Pathology, and the Whiting School of Engineering.\n",
      "Overall: PROJECT SUMMARY Cancer stem cells (CSCs) play a critical role in fostering tumor resistance to therapies and relapse after treatment. This presents a crucial barrier to the development of successful anti-cancer therapeutics. Transcriptional reprogramming and plasticity play a critical role in and out of the CSC state, which in turn are interdependent on the regulatory function of the three-dimensional (3D) structure of chromatin, epigenetic states, and other molecular events. Our understanding of fundamental CSC biology has been hampered by the need for cellular nanoscale imaging technologies that provide both highly detailed structural information regarding 3D chromatin organization and highly multiplexed molecular imaging of the many molecular regulators and events involved in CSC processes. We propose to establish the Northwestern University Center for Chromatin Nanoimaging in Cancer (NU-CCNIC) to address this fundamental technology gap in cellular nanoscale imaging and deploy the new technologies to address the fundamental knowledge gap in CSC biology. The Center converges experts in cellular nanoscale imaging, computational imaging, molecular modeling, computational genomics, CSC biology, and oncology. The Center will develop, test, validate, iterate, and deploy an integrated and co-registered Multi-scale Chromatin Nanoimaging Platform that will comprise three “nested-doll” imaging techniques: chromatin scanning transmission electron microscopy, optical spectroscopic super-resolution nanoscopy, and optical spectroscopic nanosensing. The Nanoimaging Platform will enable quantitative imaging of chromatin structure and highly multiplexed molecular and gene-specific localization, at the most fundamental length-scale approaching 1 nm resolution, including the imaging of statistically significant cell populations and live cells with high temporal resolution over prolonged temporal follow-up times. The Nanoimaging Platform will be bridged to computational genomics, epigenomics, genome mapping, and predictive transcriptional modeling datasets. These technologies will be deployed to answer several long-standing open questions in CSC biology. We will elucidate whether CSCs can originate from non-CSCs via transcriptional reprogramming, test the role of chromatin structure in fostering transcriptional plasticity in CSC processes, and explore the possibility of transcriptionally reprogramming CSCs to exit the stem-state as a new therapeutic strategy. All aspects of the technology development will be guided by the needs of the CSC biology testbed through a series of research feedback loops. In the long term, such single-cell nanoimaging technologies will help comprehensive understanding of the complex interplay between structural, physico-chemical, and molecular genomic events. We anticipate that these convergence studies will provide new insights into CSC biology, which are impossible to reveal with the use of any single method, and open new opportunities for identifying therapeutic strategies.\n",
      "Overall: PROJECT SUMMARY Cancer stem cells (CSCs) play a critical role in fostering tumor resistance to therapies and relapse after treatment. This presents a crucial barrier to the development of successful anti-cancer therapeutics. Transcriptional reprogramming and plasticity play a critical role in and out of the CSC state, which in turn are interdependent on the regulatory function of the three-dimensional (3D) structure of chromatin, epigenetic states, and other molecular events. Our understanding of fundamental CSC biology has been hampered by the need for cellular nanoscale imaging technologies that provide both highly detailed structural information regarding 3D chromatin organization and highly multiplexed molecular imaging of the many molecular regulators and events involved in CSC processes. We propose to establish the Northwestern University Center for Chromatin Nanoimaging in Cancer (NU-CCNIC) to address this fundamental technology gap in cellular nanoscale imaging and deploy the new technologies to address the fundamental knowledge gap in CSC biology. The Center converges experts in cellular nanoscale imaging, computational imaging, molecular modeling, computational genomics, CSC biology, and oncology. The Center will develop, test, validate, iterate, and deploy an integrated and co-registered Multi-scale Chromatin Nanoimaging Platform that will comprise three “nested-doll” imaging techniques: chromatin scanning transmission electron microscopy, optical spectroscopic super-resolution nanoscopy, and optical spectroscopic nanosensing. The Nanoimaging Platform will enable quantitative imaging of chromatin structure and highly multiplexed molecular and gene-specific localization, at the most fundamental length-scale approaching 1 nm resolution, including the imaging of statistically significant cell populations and live cells with high temporal resolution over prolonged temporal follow-up times. The Nanoimaging Platform will be bridged to computational genomics, epigenomics, genome mapping, and predictive transcriptional modeling datasets. These technologies will be deployed to answer several long-standing open questions in CSC biology. We will elucidate whether CSCs can originate from non-CSCs via transcriptional reprogramming, test the role of chromatin structure in fostering transcriptional plasticity in CSC processes, and explore the possibility of transcriptionally reprogramming CSCs to exit the stem-state as a new therapeutic strategy. All aspects of the technology development will be guided by the needs of the CSC biology testbed through a series of research feedback loops. In the long term, such single-cell nanoimaging technologies will help comprehensive understanding of the complex interplay between structural, physico-chemical, and molecular genomic events. We anticipate that these convergence studies will provide new insights into CSC biology, which are impossible to reveal with the use of any single method, and open new opportunities for identifying therapeutic strategies.\n",
      "Project Summary In response to the RFA for a Cellular Cancer Biology Imaging Program we propose a program focused on imaging and molecularly probing the cell biological events that drive the formation of new metastatic tumors. Specifically, we will address two questions: 1) How does the intersection of shifts in cell-intrinsic and cell- extrinsic signals associated with shifts in expression of the membrane adaptor protein Caveolin-1 affect the metastatic propensity of pediatric sarcoma (Research Testbed Unit 1)? 2) What are the effects of cell-intrinsic and cell-extrinsic variation in lipid metabolism on melanoma metastasis patterns (Research Testbed Unit 2)? Answers to both questions depend on technology to capture the molecular, metabolic, and morphological states of individual metastatic cells as they colonize the distant site: In the Technology Development Unit-1 we will develop a multi-modal, multi-scale live imaging platform to investigate the effects of intersecting microenvironmental variation across an organism and cell intrinsic heterogeneity on metastatic spreading. The platform will leverage the exquisite optical and physiological properties of the zebrafish embryos to ‘watch’ at once how cells form human tumor xenografts spread to multiple distant sites where they form metastatic tumors. The microscope will allow seamless switching between a high-throughput screening mode observing the metastatic patterns in tens to hundreds of embryos in one experiment and a high-resolution imaging mode with fully isotropic resolution of 300 nm in XYZ that allows detailed analysis of the molecular, metabolic, morphologic, and proliferation/survival states of individual cells within an emerging metastatic niche. In the Technology Development Unit-2 we will develop a multi-scale imaging platform to investigate by hyper-spectral analysis the molecular, metabolic, morphological, and functional states of metastatic cells across entire mouse organs. The platform will leverage advances in tissue clearing, fully automated high-speed and high-resolution light-sheet fluorescence imaging, and computer vision, to integrate a mesoscopic imaging mode for fast acquisition of volumes of up to 20 x 20 x 20 mm at a ~5-10 micron isotropic resolution with a nanoscopic imaging mode providing 300 nm XYZ-resolution throughout a 300 micron field of view anywhere in the organ. Biological features can thus be rapidly identified and immediately interrogated with high subcellular resolution. We will then develop physically and chemically accelerated 60-plex cyclic immunofluorescence assays to comprehensively characterize the molecular, metabolic and architectural states of colonizing cells and their surroundings in the metastatic niche in thick (~200 microns) tissue sections. To accurately describe metastatic heterogeneity, the entire system, including sample handling, labeling, and imaging, will be fully automated and operated in a high-throughput fashion. Our goal with this system is to enable comprehensive profiling of heterogeneous cell metastatic cell behavior in 100’s of intact tissue specimens. Together, these platforms will generate versatile imaging tools for a new era of in situ cancer cell biology.\n",
      "This project proposes an integrated suite of microscopy and data analysis advances that would enable quantitative, mechanistic analysis of immune-microenvironment dynamics in poor prognosis solid tumors. While immunotherapies are showing remarkable clinical responses in some advanced cancers, to date, their impact on many solid tumors has been modest. This is due, in part, to solid tumor microenvironments limiting the effectiveness of natural immune responses and immunotherapies. Yet, our understanding of the physical and molecular mechanisms governing T cell infiltration, distribution, and function in native tumor microenvironments remains extremely limited. As such, defining key T cell behaviors as a function of complex tumor microenvironments will identify design criteria that can be used to develop novel cell engineering strategies that optimize T cell-centric therapies for solid tumors. To this end the research theme of the U54 Center for Multiparametric Imaging of Tumor Immune Microenvironments (C-MITIE) is to define physical and molecular barriers to effective anti-tumor immunity and immunotherapies through advancement and development of state-of-the-art live cell and tissue optical imaging platforms and quantitative analyses. To achieve our goals, our framework brings together advanced optical imaging platforms, nano- and micro- fabrication, genome engineering, cancer immunology, and biophysical modeling. Thus, from this integrated effort we seek to define mechanisms of immune suppression that guide the development of next-generation cell-based immunotherapies.\n",
      "Project Summary - Overall The mortality rates of breast and pancreatic cancers are intrinsically tied to metastasis. In pancreatic cancer, the 5-year survival rate is only 9% and in ~70% pathological evaluations of the resected tumor, instances of venous invasion are found. Metastasis is a complex multi-step process involving cancer cells, local vasculature, and the surrounding microenvironment at multiple sites. Venous invasion in pancreatic cancer, in which cancer cells gain often invade the portal vein, is an early step in this process and provides the cells a direct path to the liver, the most common site of pancreatic cancer metastasis. As in pancreatic cancer, invasion past normal breast tissue barriers is critically tied to breast cancer outcomes. Most breast tumors can be surgically removed, and so mortality is closely tied to the extent of distant metastasis through lymphovascular invasion. The detection of lymphovascular invasion in a breast tumor correlates with poor prognosis and is not captured in current molecular analyses. The spatial organization of cancer cells, and cellular and stromal components of the tumor microenvironment near and far away from blood vessels is intrinsically three-dimensional, non-symmetric, and highly heterogeneous. In the TECH units of the Johns Hopkins Center for 3D Multiscale Cancer Imaging, we will develop a versatile 3D multiscale imaging method, CODA, which will allow us to probe the phenotypic heterogeneity of tumors from the multi-cm to the micron scale via multiplexing serial imaging. CODA can readily incorporate other imaging modalities to extract high cellular/molecular content from 3D samples. These include immunocytochemistry (CODA+IHC), immunofluorescence (CODA+IF), imaging mass cytometry (CODA+IMC) and spatial transcriptomics/proteomics (CODA+DBiT-seq). These proposed expanded versions of CODA offer a unique opportunity to produce new 3D multi-omic maps of human PDAC and breast tumors near and far from blood vessels. CODA and its integrated versions CODA+X will be tested in the RTB units of the Center in both human/mouse tissue samples and organoids, in breast and pancreatic cancer. Results from these test beds will provide novel mechanistic insights into venous invasion in breast and pancreatic cancer. Exploiting our extensive experience in data dissemination, we will make the large datasets and software produced by our CODA+X platforms and software widely available to the larger cancer research community. The units of the Center will be co-led by Johns Hopkins/Yale engineers, scientists and physicians. Substantial additional support will be provided by the Johns Hopkins U., the JH School of Medicine, the Institute for Convergence, the Department of Pathology, and the Whiting School of Engineering.\n",
      "Project Summary In response to the RFA for a Cellular Cancer Biology Imaging Program we propose a program focused on imaging and molecularly probing the cell biological events that drive the formation of new metastatic tumors. Specifically, we will address two questions: 1) How does the intersection of shifts in cell-intrinsic and cell- extrinsic signals associated with shifts in expression of the membrane adaptor protein Caveolin-1 affect the metastatic propensity of pediatric sarcoma (Research Testbed Unit 1)? 2) What are the effects of cell-intrinsic and cell-extrinsic variation in lipid metabolism on melanoma metastasis patterns (Research Testbed Unit 2)? Answers to both questions depend on technology to capture the molecular, metabolic, and morphological states of individual metastatic cells as they colonize the distant site: In the Technology Development Unit-1 we will develop a multi-modal, multi-scale live imaging platform to investigate the effects of intersecting microenvironmental variation across an organism and cell intrinsic heterogeneity on metastatic spreading. The platform will leverage the exquisite optical and physiological properties of the zebrafish embryos to ‘watch’ at once how cells form human tumor xenografts spread to multiple distant sites where they form metastatic tumors. The microscope will allow seamless switching between a high-throughput screening mode observing the metastatic patterns in tens to hundreds of embryos in one experiment and a high-resolution imaging mode with fully isotropic resolution of 300 nm in XYZ that allows detailed analysis of the molecular, metabolic, morphologic, and proliferation/survival states of individual cells within an emerging metastatic niche. In the Technology Development Unit-2 we will develop a multi-scale imaging platform to investigate by hyper-spectral analysis the molecular, metabolic, morphological, and functional states of metastatic cells across entire mouse organs. The platform will leverage advances in tissue clearing, fully automated high-speed and high-resolution light-sheet fluorescence imaging, and computer vision, to integrate a mesoscopic imaging mode for fast acquisition of volumes of up to 20 x 20 x 20 mm at a ~5-10 micron isotropic resolution with a nanoscopic imaging mode providing 300 nm XYZ-resolution throughout a 300 micron field of view anywhere in the organ. Biological features can thus be rapidly identified and immediately interrogated with high subcellular resolution. We will then develop physically and chemically accelerated 60-plex cyclic immunofluorescence assays to comprehensively characterize the molecular, metabolic and architectural states of colonizing cells and their surroundings in the metastatic niche in thick (~200 microns) tissue sections. To accurately describe metastatic heterogeneity, the entire system, including sample handling, labeling, and imaging, will be fully automated and operated in a high-throughput fashion. Our goal with this system is to enable comprehensive profiling of heterogeneous cell metastatic cell behavior in 100’s of intact tissue specimens. Together, these platforms will generate versatile imaging tools for a new era of in situ cancer cell biology.\n",
      "PROJECT SUMMARY Although esophageal adenocarcinoma (EAC) is a relatively uncommon cancer, the incidence of this disease has increased rapidly in recent decades; moreover, at the time of detection of EA, most patients have very poor prognosis. The alarming increase in EA over the past several decades, along with the poor prognosis for those affected, demand investigation into the risk factors for this disease, improved management of patients at risk, and development of preventive measures. Advances in these areas will have significant public health benefit in terms of reducing morbidity and mortality for those affected, as well as reducing the public health costs associated with managing EA at advanced stages. To better manage patients at risk for EA will require an improved understanding of the mechanisms of neoplastic progression in the esophagus. Such progression involves the development of Barrett's esophagus, a premalignant lesion in which the normal stratified squamous epithelium of the distal esophagus is replaced by columnar epithelium, visible as an abnormality on endoscopy, and confirmed on biopsy as specialized columnar epithelium with intestinal metaplasia. For these reasons, the NCI has developed the Barrett's Esophagus Translational Research Network (BETRNet), a consortium research initiative with the goal of reducing the incidence, morbidity, and mortality of EAC, through answering key questions related to the progression of this disease, especially in the premalignant stage of BE, the only precursor lesion definitively associated with EA. In this application we describe the accomplishments to date, as well as the vision forward, for the BETRNet Coordinating Center (BETRNet-CC), with the following critical functions: • Administrative and logistical support for the BETRNet program, including efficient communications and coordination of efforts to connect available resources, serve as a data management center, and integrate activities of the research centers; • Coordination of cross-network research through leadership and management of diverse teams, as well as providing logistical and administrative support to the governing body of the BETRNet (the Steering Committee); and • Ongoing development, maintenance, and administration of the BETRNet patient registry with virtual biorepository (BETRNet PR-VB).\n",
      "PROJECT SUMMARY Although esophageal adenocarcinoma (EAC) is a relatively uncommon cancer, the incidence of this disease has increased rapidly in recent decades; moreover, at the time of detection of EA, most patients have very poor prognosis. The alarming increase in EA over the past several decades, along with the poor prognosis for those affected, demand investigation into the risk factors for this disease, improved management of patients at risk, and development of preventive measures. Advances in these areas will have significant public health benefit in terms of reducing morbidity and mortality for those affected, as well as reducing the public health costs associated with managing EA at advanced stages. To better manage patients at risk for EA will require an improved understanding of the mechanisms of neoplastic progression in the esophagus. Such progression involves the development of Barrett's esophagus, a premalignant lesion in which the normal stratified squamous epithelium of the distal esophagus is replaced by columnar epithelium, visible as an abnormality on endoscopy, and confirmed on biopsy as specialized columnar epithelium with intestinal metaplasia. For these reasons, the NCI has developed the Barrett's Esophagus Translational Research Network (BETRNet), a consortium research initiative with the goal of reducing the incidence, morbidity, and mortality of EAC, through answering key questions related to the progression of this disease, especially in the premalignant stage of BE, the only precursor lesion definitively associated with EA. In this application we describe the accomplishments to date, as well as the vision forward, for the BETRNet Coordinating Center (BETRNet-CC), with the following critical functions: • Administrative and logistical support for the BETRNet program, including efficient communications and coordination of efforts to connect available resources, serve as a data management center, and integrate activities of the research centers; • Coordination of cross-network research through leadership and management of diverse teams, as well as providing logistical and administrative support to the governing body of the BETRNet (the Steering Committee); and • Ongoing development, maintenance, and administration of the BETRNet patient registry with virtual biorepository (BETRNet PR-VB).\n"
     ]
    }
   ],
   "source": [
    "# Check top 10 abstracts \n",
    "for abstract in df['abstract_text'][0:10]: print(abstract)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No sign of the problem characters. There is likely a better way to clean this in a less explicitly defined way, but this works well enough to get rid of the egregious encoding oddities and can be expanded as others are identified."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Start exploring for patterns and stats to report out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "project_num                                                  5U54CA268069-02\n",
       "core_project_num                                                 U54CA268069\n",
       "appl_id                                                             10538588\n",
       "fiscal_year                                                             2023\n",
       "project_title              Center for Multiparametric Imaging of Tumor Im...\n",
       "abstract_text              This project proposes an integrated suite of m...\n",
       "pref_terms                 Address;Advanced Malignant Neoplasm;Behavior;B...\n",
       "org_name                                             UNIVERSITY OF MINNESOTA\n",
       "org_city                                                         MINNEAPOLIS\n",
       "org_state                                                                 MN\n",
       "org_country                                                    UNITED STATES\n",
       "principal_investigators             Kevin William Eliceiri, Paolo Provenzano\n",
       "program_officers                                               Steven Becker\n",
       "award_amount                                                         1471693\n",
       "agency_ic_fundings                                                   1471693\n",
       "award_notice_date                                       2023-02-08T12:02:00Z\n",
       "project_start_date                                      2021-12-09T12:12:00Z\n",
       "project_end_date                                        2026-11-30T12:11:00Z\n",
       "full_foa                                                       RFA-CA-21-002\n",
       "api_source_search                                         nofo_RFA-CA-21-002\n",
       "program                                 CellularCancerBiologyImagingResearch\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look at a single row in detail\n",
    "df.loc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "project_num                object\n",
       "core_project_num           object\n",
       "appl_id                     int64\n",
       "fiscal_year                 int64\n",
       "project_title              object\n",
       "abstract_text              object\n",
       "pref_terms                 object\n",
       "org_name                   object\n",
       "org_city                   object\n",
       "org_state                  object\n",
       "org_country                object\n",
       "principal_investigators    object\n",
       "program_officers           object\n",
       "award_amount                int64\n",
       "agency_ic_fundings          int64\n",
       "award_notice_date          object\n",
       "project_start_date         object\n",
       "project_end_date           object\n",
       "full_foa                   object\n",
       "api_source_search          object\n",
       "program                    object\n",
       "dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>fiscal_year</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2000</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>2001</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2002</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>2003</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2004</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2005</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2006</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2007</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2008</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2009</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2010</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2011</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2012</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2013</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2014</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2015</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2016</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2017</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2018</td>\n",
       "      <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2019</td>\n",
       "      <td>241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2020</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2021</td>\n",
       "      <td>906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2022</td>\n",
       "      <td>1477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2023</td>\n",
       "      <td>1578</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index  fiscal_year\n",
       "23   2000            8\n",
       "21   2001           10\n",
       "22   2002            8\n",
       "20   2003           11\n",
       "15   2004           14\n",
       "11   2005           19\n",
       "14   2006           15\n",
       "17   2007           13\n",
       "18   2008           12\n",
       "13   2009           16\n",
       "19   2010           12\n",
       "16   2011           14\n",
       "12   2012           16\n",
       "10   2013           21\n",
       "9    2014           23\n",
       "8    2015           35\n",
       "7    2016           96\n",
       "6    2017          144\n",
       "5    2018          196\n",
       "4    2019          241\n",
       "3    2020          258\n",
       "2    2021          906\n",
       "1    2022         1477\n",
       "0    2023         1578"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check fiscal years of all grants\n",
    "df['fiscal_year'].value_counts().reset_index().sort_values(by='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['project_num',\n",
       " 'core_project_num',\n",
       " 'appl_id',\n",
       " 'fiscal_year',\n",
       " 'project_title',\n",
       " 'abstract_text',\n",
       " 'pref_terms',\n",
       " 'org_name',\n",
       " 'org_city',\n",
       " 'org_state',\n",
       " 'org_country',\n",
       " 'principal_investigators',\n",
       " 'program_officers',\n",
       " 'award_amount',\n",
       " 'agency_ic_fundings',\n",
       " 'award_notice_date',\n",
       " 'project_start_date',\n",
       " 'project_end_date',\n",
       " 'full_foa',\n",
       " 'api_source_search',\n",
       " 'program']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check columns\n",
    "df.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program</th>\n",
       "      <th>core_project_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADataResourceforBloodandMarrowTransplantsandAd...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AIDSandCancerSpecimenResourceACSR</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AcceleratingColorectalCancerScreeningandfollow...</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>TheUniversityofTexasMDAndersonCancerCenterSPOR...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>TranslationalandBasicScienceResearchinEarlyLes...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>UniversityofTexasSPOREinLungCancer</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>VanderbiltIngramCancerCenterSPOREinGastrointes...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              program  core_project_num\n",
       "0   ADMIRALStudyAdmixtureanalysisofacutelymphoblas...                 1\n",
       "1   ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...                56\n",
       "2   ADataResourceforBloodandMarrowTransplantsandAd...                 1\n",
       "3                   AIDSandCancerSpecimenResourceACSR                 1\n",
       "4   AcceleratingColorectalCancerScreeningandfollow...                11\n",
       "..                                                ...               ...\n",
       "63  TheUniversityofTexasMDAndersonCancerCenterSPOR...                 1\n",
       "64  TranslationalandBasicScienceResearchinEarlyLes...                 6\n",
       "65                 UniversityofTexasSPOREinLungCancer                 1\n",
       "66  VanderbiltIngramCancerCenterSPOREinGastrointes...                 1\n",
       "67  YaleSPOREinLungCancerYSILCTheBiologyandPersona...                 1\n",
       "\n",
       "[68 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get number of core projects for each program\n",
    "df.groupby('program')['core_project_num'].nunique().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build a summary of all Programs and the number of projects for each"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the year from the date objects\n",
    "# Lose some resolution but make the data easier to read for stats\n",
    "df['project_start_date_year'] = df['project_start_date'].apply(lambda x: int(x[:4]))\n",
    "df['project_end_date_year'] = df['project_end_date'].apply(lambda x: int(x[:4]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy fiscal year column for later min and max stats\n",
    "df['fiscal_year_copy'] = df['fiscal_year']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define functions to apply to each column\n",
    "agg_funcs = {\n",
    "    'api_source_search': 'nunique',\n",
    "    'core_project_num': 'nunique',\n",
    "    'project_num': 'nunique',\n",
    "    'agency_ic_fundings': 'sum', # This will be a little off due to duplicates!\n",
    "    'project_start_date_year': 'min',\n",
    "    'project_end_date_year': 'max',\n",
    "    'fiscal_year': 'min',\n",
    "    'fiscal_year_copy': 'max'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Define column titles better suited for reporting\n",
    "# rename_dict = {\n",
    "#     \"program\": \"Program\",\n",
    "#     \"core_project_num\": \"Core Project Count\",\n",
    "#     \"project_num\": \"Grant/Award Count\",\n",
    "#     'agency_ic_fundings': \"Total NCI Funding (since 2000)\", # This will be a little off due to duplicates! \n",
    "#     \"api_source_search\": \"Provided NOFOs/Awards with Associated Grants\",\n",
    "#     \"project_start_date_year\": \"Earliest Project Start Date\",\n",
    "#     \"project_end_date_year\": \"Latest Project End Date\",\n",
    "#     \"fiscal_year\": \"Earliest Fiscal Year\",\n",
    "#     \"fiscal_year_copy\": \"Latest Fiscal Year\"\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group by 'program' and apply aggregation functions defined above\n",
    "summary_stat_df = df.groupby('program').agg(agg_funcs).reset_index()\n",
    "\n",
    "# # Rename columns for better presentation as defined above\n",
    "# summary_stat_df.rename(columns=rename_dict, inplace=True)\n",
    "\n",
    "# # Store program summary \n",
    "# program_summary_filename = reports_dir + '/' + 'programSummaryStats.csv'\n",
    "# summary_stat_df.to_csv(program_summary_filename, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program</th>\n",
       "      <th>api_source_search</th>\n",
       "      <th>core_project_num</th>\n",
       "      <th>project_num</th>\n",
       "      <th>agency_ic_fundings</th>\n",
       "      <th>project_start_date_year</th>\n",
       "      <th>project_end_date_year</th>\n",
       "      <th>fiscal_year</th>\n",
       "      <th>fiscal_year_copy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2107456</td>\n",
       "      <td>2020</td>\n",
       "      <td>2023</td>\n",
       "      <td>2020</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...</td>\n",
       "      <td>4</td>\n",
       "      <td>56</td>\n",
       "      <td>187</td>\n",
       "      <td>100883313</td>\n",
       "      <td>2013</td>\n",
       "      <td>2024</td>\n",
       "      <td>2016</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADataResourceforBloodandMarrowTransplantsandAd...</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>28797489</td>\n",
       "      <td>1998</td>\n",
       "      <td>2028</td>\n",
       "      <td>2013</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AIDSandCancerSpecimenResourceACSR</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>18189601</td>\n",
       "      <td>2013</td>\n",
       "      <td>2024</td>\n",
       "      <td>2019</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AcceleratingColorectalCancerScreeningandfollow...</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>26</td>\n",
       "      <td>27008270</td>\n",
       "      <td>2018</td>\n",
       "      <td>2024</td>\n",
       "      <td>2018</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>TheUniversityofTexasMDAndersonCancerCenterSPOR...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>9063325</td>\n",
       "      <td>2019</td>\n",
       "      <td>2024</td>\n",
       "      <td>2019</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>TranslationalandBasicScienceResearchinEarlyLes...</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>9095020</td>\n",
       "      <td>2022</td>\n",
       "      <td>2027</td>\n",
       "      <td>2022</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>UniversityofTexasSPOREinLungCancer</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>35</td>\n",
       "      <td>52951102</td>\n",
       "      <td>1996</td>\n",
       "      <td>2025</td>\n",
       "      <td>2000</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>VanderbiltIngramCancerCenterSPOREinGastrointes...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>11590546</td>\n",
       "      <td>2019</td>\n",
       "      <td>2024</td>\n",
       "      <td>2019</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>20540459</td>\n",
       "      <td>2015</td>\n",
       "      <td>2025</td>\n",
       "      <td>2015</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              program  api_source_search  \\\n",
       "0   ADMIRALStudyAdmixtureanalysisofacutelymphoblas...                  1   \n",
       "1   ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...                  4   \n",
       "2   ADataResourceforBloodandMarrowTransplantsandAd...                  3   \n",
       "3                   AIDSandCancerSpecimenResourceACSR                  1   \n",
       "4   AcceleratingColorectalCancerScreeningandfollow...                  3   \n",
       "..                                                ...                ...   \n",
       "63  TheUniversityofTexasMDAndersonCancerCenterSPOR...                  1   \n",
       "64  TranslationalandBasicScienceResearchinEarlyLes...                  2   \n",
       "65                 UniversityofTexasSPOREinLungCancer                  1   \n",
       "66  VanderbiltIngramCancerCenterSPOREinGastrointes...                  1   \n",
       "67  YaleSPOREinLungCancerYSILCTheBiologyandPersona...                  1   \n",
       "\n",
       "    core_project_num  project_num  agency_ic_fundings  \\\n",
       "0                  1            4             2107456   \n",
       "1                 56          187           100883313   \n",
       "2                  1           12            28797489   \n",
       "3                  1            4            18189601   \n",
       "4                 11           26            27008270   \n",
       "..               ...          ...                 ...   \n",
       "63                 1            4             9063325   \n",
       "64                 6            6             9095020   \n",
       "65                 1           35            52951102   \n",
       "66                 1            5            11590546   \n",
       "67                 1           11            20540459   \n",
       "\n",
       "    project_start_date_year  project_end_date_year  fiscal_year  \\\n",
       "0                      2020                   2023         2020   \n",
       "1                      2013                   2024         2016   \n",
       "2                      1998                   2028         2013   \n",
       "3                      2013                   2024         2019   \n",
       "4                      2018                   2024         2018   \n",
       "..                      ...                    ...          ...   \n",
       "63                     2019                   2024         2019   \n",
       "64                     2022                   2027         2022   \n",
       "65                     1996                   2025         2000   \n",
       "66                     2019                   2024         2019   \n",
       "67                     2015                   2025         2015   \n",
       "\n",
       "    fiscal_year_copy  \n",
       "0               2022  \n",
       "1               2023  \n",
       "2               2023  \n",
       "3               2022  \n",
       "4               2022  \n",
       "..               ...  \n",
       "63              2022  \n",
       "64              2022  \n",
       "65              2023  \n",
       "66              2023  \n",
       "67              2023  \n",
       "\n",
       "[68 rows x 9 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_stat_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the most part, Earliest Grant Date aligns with earliest Fiscal Year. Outliers are the very early start dates for All of Us, ALCHEMIST, and EDRN programs compared to the earlist fiscal year. This might indicate a project that received additional funds or supplements many years after the proejct began. Could this be data deposition or a similar modernization effort?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Enrich Key Program CSV with Grants data\n",
    "This will begin to mock up the INS Key Programs page as requested by ODS. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use config to get path and load clean key programs csv \n",
    "key_programs_filename = '../' + config.CLEANED_KEY_PROGRAMS_CSV\n",
    "key_programs_df = pd.read_csv(key_programs_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a temporary program ID column to use for mapping stats to key programs\n",
    "key_programs_df['program'] = key_programs_df['program_name'].apply(lambda name: ''.join(filter(str.isalnum, name)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['program_name',\n",
       " 'program_acronym',\n",
       " 'focus_area',\n",
       " 'doc',\n",
       " 'contact_pi',\n",
       " 'contact_pi_email',\n",
       " 'contact_nih',\n",
       " 'contact_nih_email',\n",
       " 'nofo',\n",
       " 'award',\n",
       " 'program_link',\n",
       " 'data_link',\n",
       " 'cancer_type',\n",
       " 'program']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check columns in key programs df\n",
    "key_programs_df.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['program',\n",
       " 'api_source_search',\n",
       " 'core_project_num',\n",
       " 'project_num',\n",
       " 'agency_ic_fundings',\n",
       " 'project_start_date_year',\n",
       " 'project_end_date_year',\n",
       " 'fiscal_year',\n",
       " 'fiscal_year_copy']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check columns in program summary stats df\n",
    "summary_stat_df.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add stats to key programs df using temp program ID as connector\n",
    "key_programs_enriched = pd.merge(left=key_programs_df, \n",
    "                                right=summary_stat_df[['program',\n",
    "                                                       'core_project_num', \n",
    "                                                       'project_num',\n",
    "                                                       'agency_ic_fundings']], \n",
    "                                on='program')\n",
    "# Drop temp program ID column\n",
    "key_programs_enriched.drop(columns='program', inplace=True)\n",
    "\n",
    "# Rename stat columns \n",
    "stat_renamer_dict = {\n",
    "    'core_project_num': 'core_project_count',\n",
    "    'project_num': 'grant_count',\n",
    "    'agency_ic_fundings': 'total_nci_funding',\n",
    "}\n",
    "key_programs_enriched.rename(columns=stat_renamer_dict, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['program_name',\n",
       " 'program_acronym',\n",
       " 'focus_area',\n",
       " 'doc',\n",
       " 'contact_pi',\n",
       " 'contact_pi_email',\n",
       " 'contact_nih',\n",
       " 'contact_nih_email',\n",
       " 'nofo',\n",
       " 'award',\n",
       " 'program_link',\n",
       " 'data_link',\n",
       " 'cancer_type',\n",
       " 'core_project_count',\n",
       " 'grant_count',\n",
       " 'total_nci_funding']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "key_programs_enriched.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store enriched Key Programs Stats Table\n",
    "key_program_stats_filename = reports_dir + '/' + 'keyProgramStats.csv'\n",
    "key_programs_enriched.to_csv(key_program_stats_filename, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Continue exploring data report options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a df of all grants with project start dates before year 2000\n",
    "early_project_df = df[df['project_start_date_year'] < 2000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Export to csv for quick ad-hoc analysis\n",
    "# early_project_df.to_csv('earlyProjectReport.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program</th>\n",
       "      <th>api_source_search</th>\n",
       "      <th>core_project_num</th>\n",
       "      <th>project_title</th>\n",
       "      <th>project_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADataResourceforBloodandMarrowTransplantsandAd...</td>\n",
       "      <td>nofo_CA-12-503</td>\n",
       "      <td>U24CA076518</td>\n",
       "      <td>A Data Resource for Analyzing Blood and Marrow...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADataResourceforBloodandMarrowTransplantsandAd...</td>\n",
       "      <td>nofo_CA-17-031</td>\n",
       "      <td>U24CA076518</td>\n",
       "      <td>A Data Resource for Analyzing Blood and Marrow...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADataResourceforBloodandMarrowTransplantsandAd...</td>\n",
       "      <td>nofo_CA-22-026</td>\n",
       "      <td>U24CA076518</td>\n",
       "      <td>A Data Resource for Blood and Marrow Transplan...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>nofo_PA20-185</td>\n",
       "      <td>R01CA031845</td>\n",
       "      <td>Synthetic Studies Related to Cancer Research/T...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>nofo_PA20-185</td>\n",
       "      <td>R01CA047296</td>\n",
       "      <td>A Pathway of Tumor Suppression</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>UniversityofTexasSPOREinLungCancer</td>\n",
       "      <td>award_P50CA070907</td>\n",
       "      <td>P50CA070907</td>\n",
       "      <td>SPORE in Lung Cancer</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>UniversityofTexasSPOREinLungCancer</td>\n",
       "      <td>award_P50CA070907</td>\n",
       "      <td>P50CA070907</td>\n",
       "      <td>SPORE: Developing New Rationale, Personalized ...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>UniversityofTexasSPOREinLungCancer</td>\n",
       "      <td>award_P50CA070907</td>\n",
       "      <td>P50CA070907</td>\n",
       "      <td>Targeting Lung Cancer Vulnerabilities</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>UniversityofTexasSPOREinLungCancer</td>\n",
       "      <td>award_P50CA070907</td>\n",
       "      <td>P50CA070907</td>\n",
       "      <td>UNIVERSITY OF TEXAS SPORE IN LUNG CANCER</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>UniversityofTexasSPOREinLungCancer</td>\n",
       "      <td>award_P50CA070907</td>\n",
       "      <td>P50CA070907</td>\n",
       "      <td>University of Texas SPORE in Lung Cancer</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               program  api_source_search  \\\n",
       "0    ADataResourceforBloodandMarrowTransplantsandAd...     nofo_CA-12-503   \n",
       "1    ADataResourceforBloodandMarrowTransplantsandAd...     nofo_CA-17-031   \n",
       "2    ADataResourceforBloodandMarrowTransplantsandAd...     nofo_CA-22-026   \n",
       "3                                              AllofUs      nofo_PA20-185   \n",
       "4                                              AllofUs      nofo_PA20-185   \n",
       "..                                                 ...                ...   \n",
       "115                 UniversityofTexasSPOREinLungCancer  award_P50CA070907   \n",
       "116                 UniversityofTexasSPOREinLungCancer  award_P50CA070907   \n",
       "117                 UniversityofTexasSPOREinLungCancer  award_P50CA070907   \n",
       "118                 UniversityofTexasSPOREinLungCancer  award_P50CA070907   \n",
       "119                 UniversityofTexasSPOREinLungCancer  award_P50CA070907   \n",
       "\n",
       "    core_project_num                                      project_title  \\\n",
       "0        U24CA076518  A Data Resource for Analyzing Blood and Marrow...   \n",
       "1        U24CA076518  A Data Resource for Analyzing Blood and Marrow...   \n",
       "2        U24CA076518  A Data Resource for Blood and Marrow Transplan...   \n",
       "3        R01CA031845  Synthetic Studies Related to Cancer Research/T...   \n",
       "4        R01CA047296                     A Pathway of Tumor Suppression   \n",
       "..               ...                                                ...   \n",
       "115      P50CA070907                               SPORE in Lung Cancer   \n",
       "116      P50CA070907  SPORE: Developing New Rationale, Personalized ...   \n",
       "117      P50CA070907              Targeting Lung Cancer Vulnerabilities   \n",
       "118      P50CA070907           UNIVERSITY OF TEXAS SPORE IN LUNG CANCER   \n",
       "119      P50CA070907           University of Texas SPORE in Lung Cancer   \n",
       "\n",
       "     project_num  \n",
       "0              6  \n",
       "1              5  \n",
       "2              1  \n",
       "3              2  \n",
       "4              2  \n",
       "..           ...  \n",
       "115            8  \n",
       "116            7  \n",
       "117            3  \n",
       "118           10  \n",
       "119            6  \n",
       "\n",
       "[120 rows x 5 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group to find patterns and common sources for the projects with very early start dates\n",
    "early_project_df.groupby(['program', 'api_source_search', 'core_project_num', 'project_title'])['project_num'].nunique().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check for \"Bottleneck Effect\" grants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_rows_with_different_values(df, shared_column, compare_column):\n",
    "    \"\"\"Find rows that share a value in a specified column\n",
    "    but have different values in another specified column.\n",
    "    \"\"\"\n",
    "\n",
    "    grouped = df.groupby(shared_column)[compare_column].transform('nunique')\n",
    "    selected_row_df = df[grouped > 1]\n",
    "\n",
    "    return selected_row_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for Core Projects found in multiple programs and compare sources\n",
    "\n",
    "shared_column = 'core_project_num'\n",
    "compare_column = 'program'\n",
    "\n",
    "# Get grant-level rows with same project but different program\n",
    "df_shared = find_rows_with_different_values(df, shared_column, compare_column)\n",
    "\n",
    "# Group with the provided search value and count unique grants \n",
    "df_shared_projects = (df_shared.groupby(\n",
    "                        ['api_source_search', shared_column, compare_column])\n",
    "                        .size().reset_index()\n",
    "                        .rename(columns={0:'grant_count'}))\n",
    "\n",
    "# Store shared programs\n",
    "shared_projects_filename = reports_dir + '/' + 'sharedProjects.csv'\n",
    "df_shared_projects.to_csv(shared_projects_filename, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For any Core Project gathered for more than one program (e.g. shared NOFOs or a NOFO in one program and an Award in another), the grants and all downstream outputs will need to be associated with BOTH programs. It's important to identify these. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>api_source_search</th>\n",
       "      <th>core_project_num</th>\n",
       "      <th>program</th>\n",
       "      <th>grant_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>award_P50CA098131</td>\n",
       "      <td>P50CA098131</td>\n",
       "      <td>SPOREinBreastCancer</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>award_R01CA239701</td>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>award_U10CA180821</td>\n",
       "      <td>U10CA180821</td>\n",
       "      <td>TheAdjuvantLungCancerEnrichmentMarkerIdentific...</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nofo_CA-17-015</td>\n",
       "      <td>U01CA224145</td>\n",
       "      <td>CONSORTIUMFORPANCREATICDUCTALADENOCARCINOMAPDA...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nofo_CA-17-015</td>\n",
       "      <td>U01CA224146</td>\n",
       "      <td>CONSORTIUMFORPANCREATICDUCTALADENOCARCINOMAPDA...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>nofo_RFA-CA-22-038</td>\n",
       "      <td>U24CA224319</td>\n",
       "      <td>CANCERIMMUNEMONITORINGANDANALYSISCENTERS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>nofo_RFA-CA-22-038</td>\n",
       "      <td>U24CA224319</td>\n",
       "      <td>CancerImmunologicDataCommons</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>nofo_RFA-CA-22-038</td>\n",
       "      <td>U24CA224331</td>\n",
       "      <td>CANCERIMMUNEMONITORINGANDANALYSISCENTERS</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>nofo_RFA-CA-22-038</td>\n",
       "      <td>U24CA224331</td>\n",
       "      <td>CancerImmunologicDataCommons</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>nofo_RM21-010</td>\n",
       "      <td>U24CA268108</td>\n",
       "      <td>CellularSenescenceNetwork</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      api_source_search core_project_num  \\\n",
       "0     award_P50CA098131      P50CA098131   \n",
       "1     award_R01CA239701      R01CA239701   \n",
       "2     award_U10CA180821      U10CA180821   \n",
       "3        nofo_CA-17-015      U01CA224145   \n",
       "4        nofo_CA-17-015      U01CA224146   \n",
       "..                  ...              ...   \n",
       "116  nofo_RFA-CA-22-038      U24CA224319   \n",
       "117  nofo_RFA-CA-22-038      U24CA224319   \n",
       "118  nofo_RFA-CA-22-038      U24CA224331   \n",
       "119  nofo_RFA-CA-22-038      U24CA224331   \n",
       "120       nofo_RM21-010      U24CA268108   \n",
       "\n",
       "                                               program  grant_count  \n",
       "0                                  SPOREinBreastCancer           26  \n",
       "1    ADMIRALStudyAdmixtureanalysisofacutelymphoblas...            4  \n",
       "2    TheAdjuvantLungCancerEnrichmentMarkerIdentific...           35  \n",
       "3    CONSORTIUMFORPANCREATICDUCTALADENOCARCINOMAPDA...            3  \n",
       "4    CONSORTIUMFORPANCREATICDUCTALADENOCARCINOMAPDA...            3  \n",
       "..                                                 ...          ...  \n",
       "116           CANCERIMMUNEMONITORINGANDANALYSISCENTERS            1  \n",
       "117                       CancerImmunologicDataCommons            1  \n",
       "118           CANCERIMMUNEMONITORINGANDANALYSISCENTERS            1  \n",
       "119                       CancerImmunologicDataCommons            1  \n",
       "120                          CellularSenescenceNetwork            2  \n",
       "\n",
       "[121 rows x 4 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show core projects found in multiple programs\n",
    "df_shared_projects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some Key Programs specify Awards for grants rather than the entire NOFO. Only the Award and downstream grants should be associated with that Program, rather than everything dowstream of the larger NOFO.  \n",
    "However, all grants gathered from NIH RePORTER necessarily have a 'full_foa' (aka NOFO) value within the RePORTER data. This is a potential for complications.  \n",
    "\n",
    "We need to check for two things:  \n",
    "1. Did we search NIH RePORTER with any NOFOs that do not match the full_foa within grants returned? \n",
    "    - That would be unexpected and indicate an API gathering problem.\n",
    "    - Would appear in the table below as a rows with \"nofo_\"... in the api_source_search column\n",
    "2. Did we search NIH RePORTER with any Awards with differing full_foa values within grants returned? \n",
    "    - That's acceptable - it could indicate that the particular Core Project received funding from many sources (e.g. data sharing NOFO effort)\n",
    "    - It's still important to note. We need to be careful not to accidentally make \"upstream\" connections from grants back to full_foas back to programs for projects derived from specified Award gathering.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program_1</th>\n",
       "      <th>program_2</th>\n",
       "      <th>shared_project_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OncologyModelsForumU24</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CONSORTIUMFORPANCREATICDUCTALADENOCARCINOMAPDA...</td>\n",
       "      <td>PancreaticCancerMicroenvironmentNetwork</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CancerSystemsBiologyConsortiumCSBC</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>PediatricPreclinicalinVivoTestingPIVOT</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FusionOncoproteinsinChildhoodCancersFusOnc2</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HumanTumorAtlasNetwork</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CancerImmunologicDataCommons</td>\n",
       "      <td>CANCERIMMUNEMONITORINGANDANALYSISCENTERS</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>BarrettsEsophagusTranslationalResearchNetworkB...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HIVAssociatedMalignancyResearchCenters</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CellularCancerBiologyImagingResearch</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ChildhoodCancerSurvivorStudy</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>TheAdjuvantLungCancerEnrichmentMarkerIdentific...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CancerImmunologicDataCommons</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CellularSenescenceNetwork</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>PancreaticCancerMicroenvironmentNetwork</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AcquiredResistancetoTherapyNetworkARTNet</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CONSORTIUMFORPANCREATICDUCTALADENOCARCINOMAPDA...</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SPOREinBreastCancer</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            program_1  \\\n",
       "0                              OncologyModelsForumU24   \n",
       "1   CONSORTIUMFORPANCREATICDUCTALADENOCARCINOMAPDA...   \n",
       "2                  CancerSystemsBiologyConsortiumCSBC   \n",
       "3                                             AllofUs   \n",
       "4         FusionOncoproteinsinChildhoodCancersFusOnc2   \n",
       "5                              HumanTumorAtlasNetwork   \n",
       "6                        CancerImmunologicDataCommons   \n",
       "7                                             AllofUs   \n",
       "8              HIVAssociatedMalignancyResearchCenters   \n",
       "9                                             AllofUs   \n",
       "10               CellularCancerBiologyImagingResearch   \n",
       "11                       ChildhoodCancerSurvivorStudy   \n",
       "12                                            AllofUs   \n",
       "13                       CancerImmunologicDataCommons   \n",
       "14                          CellularSenescenceNetwork   \n",
       "15            PancreaticCancerMicroenvironmentNetwork   \n",
       "16           AcquiredResistancetoTherapyNetworkARTNet   \n",
       "17  CONSORTIUMFORPANCREATICDUCTALADENOCARCINOMAPDA...   \n",
       "18                                            AllofUs   \n",
       "19                                SPOREinBreastCancer   \n",
       "\n",
       "                                            program_2  shared_project_count  \n",
       "0                                             AllofUs                     6  \n",
       "1             PancreaticCancerMicroenvironmentNetwork                     6  \n",
       "2                                             AllofUs                     6  \n",
       "3              PediatricPreclinicalinVivoTestingPIVOT                     5  \n",
       "4                                             AllofUs                     5  \n",
       "5                                             AllofUs                     5  \n",
       "6            CANCERIMMUNEMONITORINGANDANALYSISCENTERS                     4  \n",
       "7   BarrettsEsophagusTranslationalResearchNetworkB...                     4  \n",
       "8                                             AllofUs                     3  \n",
       "9   ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...                     2  \n",
       "10                                            AllofUs                     2  \n",
       "11                                            AllofUs                     1  \n",
       "12  TheAdjuvantLungCancerEnrichmentMarkerIdentific...                     1  \n",
       "13                                            AllofUs                     1  \n",
       "14                                            AllofUs                     1  \n",
       "15                                            AllofUs                     1  \n",
       "16                                            AllofUs                     1  \n",
       "17                                            AllofUs                     1  \n",
       "18  ADMIRALStudyAdmixtureanalysisofacutelymphoblas...                     1  \n",
       "19                                            AllofUs                     1  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by core project and combine programs into unordered set\n",
    "grouped = df.groupby(\"core_project_num\")[\"program\"].apply(set)\n",
    "\n",
    "# Create combos of programs for each core project with itertools combinations\n",
    "df_shared_programs = grouped.apply(lambda x: list(combinations(x, 2)))\n",
    "\n",
    "# Flatten the program combos and count occurrences\n",
    "df_shared_programs = df_shared_programs.explode().value_counts().reset_index()\n",
    "\n",
    "# Split program combo column into two separate columns\n",
    "# NOTE this will need to be reworked if a project has more than 2 programs\n",
    "df_shared_programs[['program_1','program_2']] = df_shared_programs['index'].apply(pd.Series)\n",
    "\n",
    "# Reorder and rename\n",
    "df_shared_programs.rename(columns={'program':'shared_project_count'}, inplace=True)\n",
    "df_shared_programs = df_shared_programs[['program_1','program_2','shared_project_count']]\n",
    "\n",
    "# Export as report\n",
    "shared_programs_filename = reports_dir + '/' + 'sharedProjectsByProgramPair.csv'\n",
    "df_shared_programs.to_csv(shared_programs_filename, index=False)\n",
    "\n",
    "# Show table of program combos and number of shared projects between them\n",
    "df_shared_programs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>api_source_search</th>\n",
       "      <th>full_foa</th>\n",
       "      <th>grant_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>award_P50CA058223</td>\n",
       "      <td>PAR-00-087</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>award_P50CA058223</td>\n",
       "      <td>PAR-05-156</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>award_P50CA058223</td>\n",
       "      <td>PAR-10-003</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>award_P50CA058223</td>\n",
       "      <td>PAR-14-353</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>award_P50CA058223</td>\n",
       "      <td>RFA-CA-94-027</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>award_R01CA239701</td>\n",
       "      <td>PA-20-272</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>award_R01CA239701</td>\n",
       "      <td>PA-21-071</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>award_U10CA180821</td>\n",
       "      <td>PA-20-272</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>award_U10CA180821</td>\n",
       "      <td>RFA-CA-12-010</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>award_U10CA180821</td>\n",
       "      <td>RFA-CA-17-056</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>66 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    api_source_search       full_foa  grant_count\n",
       "0   award_P50CA058223     PAR-00-087           10\n",
       "1   award_P50CA058223     PAR-05-156            8\n",
       "2   award_P50CA058223     PAR-10-003            7\n",
       "3   award_P50CA058223     PAR-14-353            5\n",
       "4   award_P50CA058223  RFA-CA-94-027            2\n",
       "..                ...            ...          ...\n",
       "61  award_R01CA239701      PA-20-272            2\n",
       "62  award_R01CA239701      PA-21-071            1\n",
       "63  award_U10CA180821      PA-20-272            8\n",
       "64  award_U10CA180821  RFA-CA-12-010           16\n",
       "65  award_U10CA180821  RFA-CA-17-056           11\n",
       "\n",
       "[66 rows x 3 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for rows with the same ODS-provided NOFO but different FOA/NOFO accordign to RePORTER \n",
    "shared_column = 'api_source_search'\n",
    "compare_column = 'full_foa'\n",
    "\n",
    "# Select only rows with different values\n",
    "df_shared = find_rows_with_different_values(df, shared_column, compare_column)\n",
    "# Group and summarize\n",
    "df_shared.groupby([shared_column, compare_column]).size().reset_index().rename(columns={0:'grant_count'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build visualizations in Plotly\n",
    "The goal is to communicate patterns and highlight limitations or oddities within the grants data in order to provide feedback and capabilities to ODS.  \n",
    "Visualizations may be a better route than spreadsheets to summarize this. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Program Funding"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pie Chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Grouping data to calculate the sum of total_cost for each program\n",
    "# program_agency_funding = df.groupby('program')['agency_ic_fundings'].sum().reset_index()\n",
    "\n",
    "# # Creating the Pie Chart\n",
    "# fig = px.pie(program_agency_funding, \n",
    "#              names='program', \n",
    "#              values='agency_ic_fundings', \n",
    "#              title='Distribution of Agency IC Fundings Across Programs',\n",
    "#              height=600,\n",
    "#              width=1200,\n",
    "#              )\n",
    "\n",
    "# # Show the Pie Chart\n",
    "# fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### TreeMap Block Chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Grouping data to calculate the sum of total_cost for each program\n",
    "# # program_core_funding = df.groupby(['program', 'core_project_num'])['agency_ic_fundings'].sum().reset_index()\n",
    "\n",
    "# # Creating the Treemap\n",
    "# fig = px.treemap(df, \n",
    "#                  path=['program', 'project_title', 'project_num'], \n",
    "#                  values='agency_ic_fundings',\n",
    "#                  title='Distribution of NCI Funding by Program, Project, and Award',\n",
    "#                  height=1200,\n",
    "#                  width=1800)\n",
    "\n",
    "# # Show the Treemap\n",
    "# fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>org_state</th>\n",
       "      <th>agency_ic_fundings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL</td>\n",
       "      <td>12158908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AR</td>\n",
       "      <td>10446810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AZ</td>\n",
       "      <td>41317184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BC</td>\n",
       "      <td>3288053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CA</td>\n",
       "      <td>450768605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CO</td>\n",
       "      <td>16094052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CT</td>\n",
       "      <td>68695568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DC</td>\n",
       "      <td>40861789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DE</td>\n",
       "      <td>823703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FL</td>\n",
       "      <td>54670242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GA</td>\n",
       "      <td>41607253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HI</td>\n",
       "      <td>6867343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>IA</td>\n",
       "      <td>10109777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>IL</td>\n",
       "      <td>191251636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>IN</td>\n",
       "      <td>23529821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>KS</td>\n",
       "      <td>3344900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>KY</td>\n",
       "      <td>17994887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>LA</td>\n",
       "      <td>8611648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MA</td>\n",
       "      <td>658684274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MD</td>\n",
       "      <td>112046834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ME</td>\n",
       "      <td>7830625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MI</td>\n",
       "      <td>86894852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MN</td>\n",
       "      <td>72520719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MO</td>\n",
       "      <td>82824135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NC</td>\n",
       "      <td>144156239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NE</td>\n",
       "      <td>11461006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NH</td>\n",
       "      <td>8251391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NJ</td>\n",
       "      <td>26676017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NM</td>\n",
       "      <td>9322618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NY</td>\n",
       "      <td>372256324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>OH</td>\n",
       "      <td>82129734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>OK</td>\n",
       "      <td>7928144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ON</td>\n",
       "      <td>2448767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>OR</td>\n",
       "      <td>54290618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>PA</td>\n",
       "      <td>151413530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>PQ</td>\n",
       "      <td>1123806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>PR</td>\n",
       "      <td>627180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>RI</td>\n",
       "      <td>3078495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>SC</td>\n",
       "      <td>12564191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>TN</td>\n",
       "      <td>156308730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>TX</td>\n",
       "      <td>404592165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>UT</td>\n",
       "      <td>31497021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>VA</td>\n",
       "      <td>35766164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>VT</td>\n",
       "      <td>554911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>WA</td>\n",
       "      <td>144747358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>WI</td>\n",
       "      <td>72570283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>WV</td>\n",
       "      <td>714780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10332742</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   org_state  agency_ic_fundings\n",
       "0         AL            12158908\n",
       "1         AR            10446810\n",
       "2         AZ            41317184\n",
       "3         BC             3288053\n",
       "4         CA           450768605\n",
       "5         CO            16094052\n",
       "6         CT            68695568\n",
       "7         DC            40861789\n",
       "8         DE              823703\n",
       "9         FL            54670242\n",
       "10        GA            41607253\n",
       "11        HI             6867343\n",
       "12        IA            10109777\n",
       "13        IL           191251636\n",
       "14        IN            23529821\n",
       "15        KS             3344900\n",
       "16        KY            17994887\n",
       "17        LA             8611648\n",
       "18        MA           658684274\n",
       "19        MD           112046834\n",
       "20        ME             7830625\n",
       "21        MI            86894852\n",
       "22        MN            72520719\n",
       "23        MO            82824135\n",
       "24        NC           144156239\n",
       "25        NE            11461006\n",
       "26        NH             8251391\n",
       "27        NJ            26676017\n",
       "28        NM             9322618\n",
       "29        NY           372256324\n",
       "30        OH            82129734\n",
       "31        OK             7928144\n",
       "32        ON             2448767\n",
       "33        OR            54290618\n",
       "34        PA           151413530\n",
       "35        PQ             1123806\n",
       "36        PR              627180\n",
       "37        RI             3078495\n",
       "38        SC            12564191\n",
       "39        TN           156308730\n",
       "40        TX           404592165\n",
       "41        UT            31497021\n",
       "42        VA            35766164\n",
       "43        VT              554911\n",
       "44        WA           144747358\n",
       "45        WI            72570283\n",
       "46        WV              714780\n",
       "47       NaN            10332742"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grouping data to calculate the sum of agency_ic_funding for each org_state\n",
    "state_funding = df.groupby('org_state', dropna=False)['agency_ic_fundings'].sum().reset_index()\n",
    "state_funding"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Chloropleth (US Map with funding by state)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Grouping data to calculate the sum of agency_ic_funding for each org_state\n",
    "# state_funding = df.groupby('org_state',dropna=False)['agency_ic_fundings'].sum().reset_index()\n",
    "\n",
    "# # Creating the Chloropleth\n",
    "# fig = px.choropleth(state_funding, \n",
    "#                     locations='org_state', \n",
    "#                     locationmode='USA-states', \n",
    "#                     color='agency_ic_fundings',\n",
    "#                     scope='usa',\n",
    "#                     color_continuous_scale='Blues',\n",
    "#                     title='Total NCI Funding by State since 2000',\n",
    "#                     height=600,\n",
    "#                     width=1200)\n",
    "\n",
    "# # Show the Chloropleth\n",
    "# fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program</th>\n",
       "      <th>fiscal_year</th>\n",
       "      <th>agency_ic_fundings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>2020</td>\n",
       "      <td>1898254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>2021</td>\n",
       "      <td>11554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>2022</td>\n",
       "      <td>197648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...</td>\n",
       "      <td>2016</td>\n",
       "      <td>3167336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...</td>\n",
       "      <td>2017</td>\n",
       "      <td>12039602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>2019</td>\n",
       "      <td>2134771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>2020</td>\n",
       "      <td>2207890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>472</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>2021</td>\n",
       "      <td>2003668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>2022</td>\n",
       "      <td>2046999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>2023</td>\n",
       "      <td>2069846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>475 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               program  fiscal_year  \\\n",
       "0    ADMIRALStudyAdmixtureanalysisofacutelymphoblas...         2020   \n",
       "1    ADMIRALStudyAdmixtureanalysisofacutelymphoblas...         2021   \n",
       "2    ADMIRALStudyAdmixtureanalysisofacutelymphoblas...         2022   \n",
       "3    ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...         2016   \n",
       "4    ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...         2017   \n",
       "..                                                 ...          ...   \n",
       "470  YaleSPOREinLungCancerYSILCTheBiologyandPersona...         2019   \n",
       "471  YaleSPOREinLungCancerYSILCTheBiologyandPersona...         2020   \n",
       "472  YaleSPOREinLungCancerYSILCTheBiologyandPersona...         2021   \n",
       "473  YaleSPOREinLungCancerYSILCTheBiologyandPersona...         2022   \n",
       "474  YaleSPOREinLungCancerYSILCTheBiologyandPersona...         2023   \n",
       "\n",
       "     agency_ic_fundings  \n",
       "0               1898254  \n",
       "1                 11554  \n",
       "2                197648  \n",
       "3               3167336  \n",
       "4              12039602  \n",
       "..                  ...  \n",
       "470             2134771  \n",
       "471             2207890  \n",
       "472             2003668  \n",
       "473             2046999  \n",
       "474             2069846  \n",
       "\n",
       "[475 rows x 3 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "program_fy_funding = df.groupby(['program','fiscal_year'])['agency_ic_fundings'].sum().reset_index()\n",
    "program_fy_funding"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Stacked Bar Chart (Program funding per year)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Group data\n",
    "# program_fy_funding = df.groupby(['program','fiscal_year'])['agency_ic_fundings'].sum().reset_index()\n",
    "\n",
    "# # Create the line chart\n",
    "# fig = px.bar(program_fy_funding, \n",
    "#               x='fiscal_year', \n",
    "#               y='agency_ic_fundings', \n",
    "#               color='program', \n",
    "#               title='NCI Funding Since Fiscal Year 2000 by Key Program',\n",
    "#               labels={'fiscal_year': 'Fiscal Year', 'agency_ic_fundings': 'NCI Funding', 'program':'Key Program'},\n",
    "#               height=600,\n",
    "#               width=1800\n",
    "#             )\n",
    "# # Show the line chart\n",
    "# fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sankey Diagram\n",
    "There was a lot of trial and error not fully shown here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program</th>\n",
       "      <th>full_foa</th>\n",
       "      <th>core_project_num</th>\n",
       "      <th>agency_ic_fundings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>PA-19-056</td>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>1898254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>PA-20-272</td>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>197648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>PA-21-071</td>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>11554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...</td>\n",
       "      <td>PAR-15-331</td>\n",
       "      <td>U24CA180922</td>\n",
       "      <td>3997961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...</td>\n",
       "      <td>PAR-15-331</td>\n",
       "      <td>U24CA180996</td>\n",
       "      <td>2099900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2408</th>\n",
       "      <td>UniversityofTexasSPOREinLungCancer</td>\n",
       "      <td>PAR-18-313</td>\n",
       "      <td>P50CA070907</td>\n",
       "      <td>6449069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2409</th>\n",
       "      <td>VanderbiltIngramCancerCenterSPOREinGastrointes...</td>\n",
       "      <td>PAR-18-313</td>\n",
       "      <td>P50CA236733</td>\n",
       "      <td>11590546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2410</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>PAR-14-031</td>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>12004456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2411</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>PAR-18-313</td>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>8328403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2412</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>RFA-CA-16-012</td>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>207600</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2413 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                program       full_foa  \\\n",
       "0     ADMIRALStudyAdmixtureanalysisofacutelymphoblas...      PA-19-056   \n",
       "1     ADMIRALStudyAdmixtureanalysisofacutelymphoblas...      PA-20-272   \n",
       "2     ADMIRALStudyAdmixtureanalysisofacutelymphoblas...      PA-21-071   \n",
       "3     ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...     PAR-15-331   \n",
       "4     ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...     PAR-15-331   \n",
       "...                                                 ...            ...   \n",
       "2408                 UniversityofTexasSPOREinLungCancer     PAR-18-313   \n",
       "2409  VanderbiltIngramCancerCenterSPOREinGastrointes...     PAR-18-313   \n",
       "2410  YaleSPOREinLungCancerYSILCTheBiologyandPersona...     PAR-14-031   \n",
       "2411  YaleSPOREinLungCancerYSILCTheBiologyandPersona...     PAR-18-313   \n",
       "2412  YaleSPOREinLungCancerYSILCTheBiologyandPersona...  RFA-CA-16-012   \n",
       "\n",
       "     core_project_num  agency_ic_fundings  \n",
       "0         R01CA239701             1898254  \n",
       "1         R01CA239701              197648  \n",
       "2         R01CA239701               11554  \n",
       "3         U24CA180922             3997961  \n",
       "4         U24CA180996             2099900  \n",
       "...               ...                 ...  \n",
       "2408      P50CA070907             6449069  \n",
       "2409      P50CA236733            11590546  \n",
       "2410      P50CA196530            12004456  \n",
       "2411      P50CA196530             8328403  \n",
       "2412      P50CA196530              207600  \n",
       "\n",
       "[2413 rows x 4 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sum NCI funding for each program, NOFO, and project combo\n",
    "sankey_summary = df.groupby(['program', 'full_foa', 'core_project_num'])['agency_ic_fundings'].sum().reset_index()\n",
    "sankey_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>PA-19-056</td>\n",
       "      <td>1898254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>PA-20-272</td>\n",
       "      <td>197648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>PA-21-071</td>\n",
       "      <td>11554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...</td>\n",
       "      <td>PAR-15-331</td>\n",
       "      <td>49498005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...</td>\n",
       "      <td>PAR-15-332</td>\n",
       "      <td>19027032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2377</th>\n",
       "      <td>RFA-CA-22-051</td>\n",
       "      <td>UG1CA284884</td>\n",
       "      <td>324621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2378</th>\n",
       "      <td>RFA-CA-22-051</td>\n",
       "      <td>UG1CA284918</td>\n",
       "      <td>275910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379</th>\n",
       "      <td>RFA-CA-94-027</td>\n",
       "      <td>P50CA058223</td>\n",
       "      <td>1011600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2380</th>\n",
       "      <td>RFA-CA-95-021</td>\n",
       "      <td>P50CA062924</td>\n",
       "      <td>6538584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2399</th>\n",
       "      <td>RFA-RM-21-016</td>\n",
       "      <td>DP2CA281605</td>\n",
       "      <td>188769</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2495 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 source       target     value\n",
       "0     ADMIRALStudyAdmixtureanalysisofacutelymphoblas...    PA-19-056   1898254\n",
       "1     ADMIRALStudyAdmixtureanalysisofacutelymphoblas...    PA-20-272    197648\n",
       "2     ADMIRALStudyAdmixtureanalysisofacutelymphoblas...    PA-21-071     11554\n",
       "3     ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...   PAR-15-331  49498005\n",
       "4     ADVANCEDDEVELOPMENTOFINFORMATICSTECHNOLOGIESFO...   PAR-15-332  19027032\n",
       "...                                                 ...          ...       ...\n",
       "2377                                      RFA-CA-22-051  UG1CA284884    324621\n",
       "2378                                      RFA-CA-22-051  UG1CA284918    275910\n",
       "2379                                      RFA-CA-94-027  P50CA058223   1011600\n",
       "2380                                      RFA-CA-95-021  P50CA062924   6538584\n",
       "2399                                      RFA-RM-21-016  DP2CA281605    188769\n",
       "\n",
       "[2495 rows x 3 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Build connections from programs to full_foas\n",
    "links_program_nofo = df.groupby(['program','full_foa'])['agency_ic_fundings'].sum().reset_index()\n",
    "# Rename with standard cols\n",
    "links_program_nofo.columns = ['source','target','value']\n",
    "\n",
    "\n",
    "# Build connections from full_foas to project_nums\n",
    "links_nofo_project = df.groupby(['full_foa','core_project_num'])['agency_ic_fundings'].sum().reset_index()\n",
    "# Rename with standard cols\n",
    "links_nofo_project.columns = ['source','target','value']\n",
    "\n",
    "# Combine different links dataframes\n",
    "links = pd.concat([links_program_nofo, links_nofo_project])\n",
    "\n",
    "# Remove links with no funding\n",
    "links = links[links['value'] > 0]\n",
    "links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>71</td>\n",
       "      <td>1898254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>73</td>\n",
       "      <td>197648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>74</td>\n",
       "      <td>11554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>88</td>\n",
       "      <td>49498005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>89</td>\n",
       "      <td>19027032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2377</th>\n",
       "      <td>181</td>\n",
       "      <td>2384</td>\n",
       "      <td>324621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2378</th>\n",
       "      <td>181</td>\n",
       "      <td>2385</td>\n",
       "      <td>275910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379</th>\n",
       "      <td>182</td>\n",
       "      <td>1999</td>\n",
       "      <td>1011600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2380</th>\n",
       "      <td>183</td>\n",
       "      <td>2000</td>\n",
       "      <td>6538584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2399</th>\n",
       "      <td>184</td>\n",
       "      <td>2386</td>\n",
       "      <td>188769</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2495 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      source  target     value\n",
       "0          0      71   1898254\n",
       "1          0      73    197648\n",
       "2          0      74     11554\n",
       "3          1      88  49498005\n",
       "4          1      89  19027032\n",
       "...      ...     ...       ...\n",
       "2377     181    2384    324621\n",
       "2378     181    2385    275910\n",
       "2379     182    1999   1011600\n",
       "2380     183    2000   6538584\n",
       "2399     184    2386    188769\n",
       "\n",
       "[2495 rows x 3 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create nodes dataframe from links dataframe\n",
    "nodes_data = pd.concat([links['source'], links['target']]).unique()\n",
    "nodes = pd.DataFrame({\n",
    "    'node': nodes_data,\n",
    "    'node_id': range(len(nodes_data))\n",
    "})\n",
    "\n",
    "# Create a dictionary to map node names to node IDs\n",
    "node_id_mapping = dict(zip(nodes['node'], nodes['node_id']))\n",
    "\n",
    "# Map the node names to their corresponding node IDs in the links dataframe\n",
    "links['source'] = links['source'].map(node_id_mapping)\n",
    "links['target'] = links['target'].map(node_id_mapping)\n",
    "\n",
    "links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Create Sankey diagram\n",
    "# fig = go.Figure(go.Sankey(\n",
    "#     node=dict(\n",
    "#         pad=15,\n",
    "#         thickness=20,\n",
    "#         line=dict(color=\"black\", width=0.5),\n",
    "#         label=nodes['node'],  # Use the node names as labels\n",
    "\n",
    "#     ),\n",
    "#     link=dict(\n",
    "#         source=links['source'],\n",
    "#         target=links['target'],\n",
    "#         value=links['value']\n",
    "#     )\n",
    "# ))\n",
    "\n",
    "# # Customize layout\n",
    "# fig.update_layout(\n",
    "#     title_text=\"NCI Key Program Funding Flow\",\n",
    "#     font_size=14,\n",
    "#     height=1800,\n",
    "#     width=1200\n",
    "# )\n",
    "\n",
    "# # Display the figure\n",
    "# fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Program Keywords (from Grants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pref_terms</th>\n",
       "      <th>keyword_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5259</th>\n",
       "      <td>Malignant Neoplasms</td>\n",
       "      <td>3826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6558</th>\n",
       "      <td>Patients</td>\n",
       "      <td>3717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2413</th>\n",
       "      <td>Data</td>\n",
       "      <td>3498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13785</th>\n",
       "      <td>tumor</td>\n",
       "      <td>3320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12252</th>\n",
       "      <td>novel</td>\n",
       "      <td>3283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8541</th>\n",
       "      <td>Testing</td>\n",
       "      <td>3231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3775</th>\n",
       "      <td>Goals</td>\n",
       "      <td>3089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1851</th>\n",
       "      <td>Clinical</td>\n",
       "      <td>2996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2531</th>\n",
       "      <td>Development</td>\n",
       "      <td>2923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11337</th>\n",
       "      <td>improved</td>\n",
       "      <td>2909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1577</th>\n",
       "      <td>Cells</td>\n",
       "      <td>2772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4194</th>\n",
       "      <td>Human</td>\n",
       "      <td>2537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5642</th>\n",
       "      <td>Modeling</td>\n",
       "      <td>2536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7664</th>\n",
       "      <td>Role</td>\n",
       "      <td>2449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13077</th>\n",
       "      <td>response</td>\n",
       "      <td>2435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7537</th>\n",
       "      <td>Research</td>\n",
       "      <td>2427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2648</th>\n",
       "      <td>Disease</td>\n",
       "      <td>2369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9557</th>\n",
       "      <td>base</td>\n",
       "      <td>2359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5653</th>\n",
       "      <td>Molecular</td>\n",
       "      <td>2305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8568</th>\n",
       "      <td>Therapeutic</td>\n",
       "      <td>2200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9209</th>\n",
       "      <td>Work</td>\n",
       "      <td>2140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>Address</td>\n",
       "      <td>2035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6529</th>\n",
       "      <td>Pathway interactions</td>\n",
       "      <td>1967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11350</th>\n",
       "      <td>in vivo</td>\n",
       "      <td>1854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5762</th>\n",
       "      <td>Mus</td>\n",
       "      <td>1823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7956</th>\n",
       "      <td>Signal Transduction</td>\n",
       "      <td>1702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5376</th>\n",
       "      <td>Mediating</td>\n",
       "      <td>1700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6324</th>\n",
       "      <td>Outcome</td>\n",
       "      <td>1667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8634</th>\n",
       "      <td>Time</td>\n",
       "      <td>1645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11422</th>\n",
       "      <td>innovation</td>\n",
       "      <td>1644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12003</th>\n",
       "      <td>mouse model</td>\n",
       "      <td>1628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5783</th>\n",
       "      <td>Mutation</td>\n",
       "      <td>1605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3616</th>\n",
       "      <td>Genes</td>\n",
       "      <td>1557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5527</th>\n",
       "      <td>Methods</td>\n",
       "      <td>1553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1885</th>\n",
       "      <td>Clinical Trials</td>\n",
       "      <td>1531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7566</th>\n",
       "      <td>Resistance</td>\n",
       "      <td>1517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12800</th>\n",
       "      <td>programs</td>\n",
       "      <td>1516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7141</th>\n",
       "      <td>Proteins</td>\n",
       "      <td>1492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7785</th>\n",
       "      <td>Sampling</td>\n",
       "      <td>1477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3617</th>\n",
       "      <td>Genetic</td>\n",
       "      <td>1377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8668</th>\n",
       "      <td>Tissues</td>\n",
       "      <td>1373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7006</th>\n",
       "      <td>Process</td>\n",
       "      <td>1352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7547</th>\n",
       "      <td>Research Personnel</td>\n",
       "      <td>1347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10373</th>\n",
       "      <td>design</td>\n",
       "      <td>1345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9752</th>\n",
       "      <td>cancer cell</td>\n",
       "      <td>1342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11427</th>\n",
       "      <td>insight</td>\n",
       "      <td>1337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4840</th>\n",
       "      <td>Knowledge</td>\n",
       "      <td>1323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12153</th>\n",
       "      <td>neoplastic cell</td>\n",
       "      <td>1299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4466</th>\n",
       "      <td>In Vitro</td>\n",
       "      <td>1275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8365</th>\n",
       "      <td>System</td>\n",
       "      <td>1234</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 pref_terms  keyword_count\n",
       "5259    Malignant Neoplasms           3826\n",
       "6558               Patients           3717\n",
       "2413                   Data           3498\n",
       "13785                 tumor           3320\n",
       "12252                 novel           3283\n",
       "8541                Testing           3231\n",
       "3775                  Goals           3089\n",
       "1851               Clinical           2996\n",
       "2531            Development           2923\n",
       "11337              improved           2909\n",
       "1577                  Cells           2772\n",
       "4194                  Human           2537\n",
       "5642               Modeling           2536\n",
       "7664                   Role           2449\n",
       "13077              response           2435\n",
       "7537               Research           2427\n",
       "2648                Disease           2369\n",
       "9557                   base           2359\n",
       "5653              Molecular           2305\n",
       "8568            Therapeutic           2200\n",
       "9209                   Work           2140\n",
       "300                 Address           2035\n",
       "6529   Pathway interactions           1967\n",
       "11350               in vivo           1854\n",
       "5762                    Mus           1823\n",
       "7956    Signal Transduction           1702\n",
       "5376              Mediating           1700\n",
       "6324                Outcome           1667\n",
       "8634                   Time           1645\n",
       "11422            innovation           1644\n",
       "12003           mouse model           1628\n",
       "5783               Mutation           1605\n",
       "3616                  Genes           1557\n",
       "5527                Methods           1553\n",
       "1885        Clinical Trials           1531\n",
       "7566             Resistance           1517\n",
       "12800              programs           1516\n",
       "7141               Proteins           1492\n",
       "7785               Sampling           1477\n",
       "3617                Genetic           1377\n",
       "8668                Tissues           1373\n",
       "7006                Process           1352\n",
       "7547     Research Personnel           1347\n",
       "10373                design           1345\n",
       "9752            cancer cell           1342\n",
       "11427               insight           1337\n",
       "4840              Knowledge           1323\n",
       "12153       neoplastic cell           1299\n",
       "4466               In Vitro           1275\n",
       "8365                 System           1234"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get top keywords across all grants gathered\n",
    "df['pref_terms'].str.split(';').explode().reset_index().groupby('pref_terms').size().reset_index().sort_values(by=0, ascending=False).rename(columns={0:'keyword_count'}).head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program</th>\n",
       "      <th>pref_terms</th>\n",
       "      <th>keyword_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>12 year old</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>Accounting</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>Acute Lymphocytic Leukemia</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>Administrative Supplement</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>Admixture</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40416</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>tumor</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40417</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>tumor initiation</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40418</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>tumor microenvironment</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40419</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>tumor progression</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40420</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>tumorigenesis</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40421 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 program  \\\n",
       "0      ADMIRALStudyAdmixtureanalysisofacutelymphoblas...   \n",
       "1      ADMIRALStudyAdmixtureanalysisofacutelymphoblas...   \n",
       "2      ADMIRALStudyAdmixtureanalysisofacutelymphoblas...   \n",
       "3      ADMIRALStudyAdmixtureanalysisofacutelymphoblas...   \n",
       "4      ADMIRALStudyAdmixtureanalysisofacutelymphoblas...   \n",
       "...                                                  ...   \n",
       "40416  YaleSPOREinLungCancerYSILCTheBiologyandPersona...   \n",
       "40417  YaleSPOREinLungCancerYSILCTheBiologyandPersona...   \n",
       "40418  YaleSPOREinLungCancerYSILCTheBiologyandPersona...   \n",
       "40419  YaleSPOREinLungCancerYSILCTheBiologyandPersona...   \n",
       "40420  YaleSPOREinLungCancerYSILCTheBiologyandPersona...   \n",
       "\n",
       "                       pref_terms  keyword_count  \n",
       "0                     12 year old              1  \n",
       "1                      Accounting              1  \n",
       "2      Acute Lymphocytic Leukemia              4  \n",
       "3       Administrative Supplement              2  \n",
       "4                       Admixture              4  \n",
       "...                           ...            ...  \n",
       "40416                       tumor             11  \n",
       "40417            tumor initiation              1  \n",
       "40418      tumor microenvironment              4  \n",
       "40419           tumor progression             11  \n",
       "40420               tumorigenesis              1  \n",
       "\n",
       "[40421 rows x 3 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Split keywords and explode\n",
    "keywords_df = df.assign(pref_terms=df['pref_terms'].str.split(';')).explode('pref_terms')\n",
    "\n",
    "# Count keyword frequencies per program\n",
    "keyword_counts = keywords_df.groupby(['program', 'pref_terms']).size().reset_index(name='keyword_count')\n",
    "\n",
    "keyword_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Keywords seem valuable but it's difficult to find an appropriate visualization. I'd rather not rely on WordCloud. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ins_data_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
