{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary Statistics \n",
    "2023-08-08 ZD  \n",
    "\n",
    "This notebook will explore options to gather summary statistics and other reporting data calculated from data within processed grants data.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "import numpy as np\n",
    "import re\n",
    "from itertools import combinations\n",
    "\n",
    "\n",
    "# Method to import from parent directory\n",
    "import os\n",
    "import sys\n",
    "root_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "sys.path.append(root_dir)\n",
    "\n",
    "import config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load data from all Key Program grants output CSVs into a single dataframe for stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Project data pulled from: ../data/processed/2023-07-19/api-gathered-2023-08-30\n",
      "Reports will be output to: ../reports/2023-07-19/api-gathered-2023-08-30\n"
     ]
    }
   ],
   "source": [
    "# Define directory containing processed grant data\n",
    "processed_dir = '../'+config.PROCESSED_DIR\n",
    "\n",
    "# Define directory to store reports. Create if doesn't already exist\n",
    "reports_dir = '../'+config.REPORTS_DIR\n",
    "if not os.path.exists(reports_dir):\n",
    "    os.makedirs(reports_dir)\n",
    "\n",
    "# # Backup paths to allow for using same set of gathered data during notebook development\n",
    "# processed_dir = '../' + 'data/processed/2023-07-19/api-gathered-2023-08-25'\n",
    "# reports_dir = '../' + 'reports/2023-07-19/api-gathered-2023-08-25'\n",
    "if not os.path.exists(reports_dir):\n",
    "    os.makedirs(reports_dir)\n",
    "\n",
    "print(f\"Project data pulled from: {processed_dir}\")\n",
    "print(f\"Reports will be output to: {reports_dir}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load grants data\n",
    "grants_filename = os.path.join(processed_dir, 'project.tsv')\n",
    "df = pd.read_csv(grants_filename, sep='\\t')\n",
    "\n",
    "# Rename program.program_id column\n",
    "df.rename(columns={'program.program_id':'program'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quick detour to handle abstract text cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'IMMUNE MONITORING AND ANALYSIS OF CANCER AT STANFORD (IMACS) Abstract The Center for Immune Monitoring and Analysis of Cancer at Stanford (IMACS) will perform highly comprehensive assays of immune phenotype and function for NCI-identified clinical trials. These will include standardized assays already developed on CyTOF, high-dimensional flow cytometry, Luminex, TCRseq, and RNAseq platforms. As part of the program we will also standardize and offer as assays Stanford-invented technologies under development, including Multiplexed Ion Beam Imaging (MIBI) and Assays of Transposon- Accessible Chromatin (ATAC-seq). We have designed our center structure to work with investigators to define the assays best suited to the immunological questions being posed, and match these with the required sample types. We will perform quality control measures on all assays, as well as generate a standard report for each assay and project. Data will be organized via our online database, Stanford Data Miner, to ensure data longevity and transferability, as well as access to both raw data files and analyzed results. Finally, we will work with investigators on novel bioinformatics approaches to mining these high-dimensional data sets. These will include approaches designed for a single data type (e.g., viSNE and Citrus for CyTOF and flow cytometry data), as well as approaches for integrating data across assays, using appropriate machine learning algorithms to aid NCI researchers in identifying immune hallmarks central to their trials. Relevance: The IMACS center will provide access to a suite of state-of-the-art immune assays, many of them developed or refined at Stanford. This unmatched set of technologies will facilitate the discovery of new biomarkers for predicting cancer outcome or therapeutic response, as well as defining potential new mechanisms of immune control of cancer.'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pull one abstract to check for odd character encoding\n",
    "text = df.abstract_text[1]\n",
    "text"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lots of odd characters in here like \"\\xad\" and \"\\xa0\". Double-spaces were an issue before as well.  \n",
    "Make a function that can clean the column with regex and space replacement.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_abstract(text):\n",
    "    # Define a list of characters to be removed or replaced\n",
    "    chars_to_remove = ['\\xad']\n",
    "    chars_to_space = ['  ']\n",
    "\n",
    "    if isinstance(text, str):\n",
    "        # Remove unwanted characters (no space added)\n",
    "        for char in chars_to_remove:\n",
    "            text = text.replace(char, '')\n",
    "        # Replace unwanted characters with a space\n",
    "        for char in chars_to_space:\n",
    "            text = text.replace(char, ' ')\n",
    "\n",
    "        # Remove non-breaking spaces, newlines, and other unwanted characters\n",
    "        cleaned_text = re.sub(r'[\\s\\xa0]+', ' ', text).strip()\n",
    "        return cleaned_text\n",
    "        \n",
    "    # Return original value if it is NaN or float\n",
    "    else: \n",
    "        return text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'IMMUNE MONITORING AND ANALYSIS OF CANCER AT STANFORD (IMACS) Abstract The Center for Immune Monitoring and Analysis of Cancer at Stanford (IMACS) will perform highly comprehensive assays of immune phenotype and function for NCI-identified clinical trials. These will include standardized assays already developed on CyTOF, high-dimensional flow cytometry, Luminex, TCRseq, and RNAseq platforms. As part of the program we will also standardize and offer as assays Stanford-invented technologies under development, including Multiplexed Ion Beam Imaging (MIBI) and Assays of Transposon- Accessible Chromatin (ATAC-seq). We have designed our center structure to work with investigators to define the assays best suited to the immunological questions being posed, and match these with the required sample types. We will perform quality control measures on all assays, as well as generate a standard report for each assay and project. Data will be organized via our online database, Stanford Data Miner, to ensure data longevity and transferability, as well as access to both raw data files and analyzed results. Finally, we will work with investigators on novel bioinformatics approaches to mining these high-dimensional data sets. These will include approaches designed for a single data type (e.g., viSNE and Citrus for CyTOF and flow cytometry data), as well as approaches for integrating data across assays, using appropriate machine learning algorithms to aid NCI researchers in identifying immune hallmarks central to their trials. Relevance: The IMACS center will provide access to a suite of state-of-the-art immune assays, many of them developed or refined at Stanford. This unmatched set of technologies will facilitate the discovery of new biomarkers for predicting cancer outcome or therapeutic response, as well as defining potential new mechanisms of immune control of cancer.'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Try it on the example abstract from before\n",
    "clean = clean_abstract(text)\n",
    "clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean abstract column\n",
    "df['abstract_text'] = df['abstract_text'].apply(clean_abstract)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ABSTRACT Advances in immunotherapy have shown efficacy in various cancers. Immunotherapeutic approaches are becoming the new treatment modality in some cancer types. However, clinical efficacy is still limited to a marginal number of patients due to a newly developing understanding of the complex tumor microenvironment and a limited ability to appropriately select patients who have specific biomarker signatures and have the potential to optimally respond to specific immunotherapy or combination strategies. An additional challenge in assessment of biomarkers that show clinical utility to predict patient benefit includes the use of different methodologies and platforms that make it difficult to make accurate conclusions on the biomarkers in question. Thus, optimized biomarker strategies that can overcome immune barriers will allow tailoring of the therapeutic approaches and result in bold interventions that would be most beneficial to individual patients. The National Cancer Moonshot was launched in 2016 with the goal of boldly combating cancer beyond incremental advances leading to the creation of a “Cancer Immunotherapy Translational Science Network” that would facilitate translational research and clinical advances in immunotherapy. Our proposal to create a resource center, named Translational Cancer Immune Monitoring and Analysis Center (TCIMAC) that will be a critical resource capable of serving any of the NCI-supported clinical trial networks. The TCIMAC will be co-led by Drs. Ignacio Wistuba, renowned cancer surgical and molecular pathologist, Elizabeth Mittendorf, well-known breast cancer surgeon and immunologist, and Chantale Bernatchez, cancer immunologist. They will be supported by a team that encompasses a multidisciplinary group of world-class and highly collaborative experts on cancer and immunotherapy. The main goals of TCIMAC are to: 1) provide centralized and harmonized services for sample collection, processing and quality assurance, and 2) use analytically-validated and standardized assays and platforms to offer state-of-the-art analyses for patients enrolled on NCI- sponsored early phase immunotherapy clinical trials. In Aim 1, we will assist in standardizing bio-banking activities and share/develop standard operation procedures (SOPs) for biospecimen collection, processing, quality assessment, and storage of biospecimens collected from NCI-sponsored early phase immunotherapy clinical trials and to link the specimens to relevant clinical, pathological, immune and molecular data within the TCIMAC-CIDC Network. In Aim 2, we will provide services to conduct both routine and innovative pathological, immunological and molecular analyses using standardized and validated assays to aid the completion of NCI- sponsored trials. In Aim 3, we will provide comprehensive biostatistics and computational services for data collection and analysis for each analytical platform, and to perform analyses, interpretation and predictive modeling of high dimensional (`omic”) data in the context of clinical outcomes. We envision that the TCIMAC will be an indispensable element as we make meaningful progress in cancer immunotherapeutic approaches.\n",
      "IMMUNE MONITORING AND ANALYSIS OF CANCER AT STANFORD (IMACS) Abstract The Center for Immune Monitoring and Analysis of Cancer at Stanford (IMACS) will perform highly comprehensive assays of immune phenotype and function for NCI-identified clinical trials. These will include standardized assays already developed on CyTOF, high-dimensional flow cytometry, Luminex, TCRseq, and RNAseq platforms. As part of the program we will also standardize and offer as assays Stanford-invented technologies under development, including Multiplexed Ion Beam Imaging (MIBI) and Assays of Transposon- Accessible Chromatin (ATAC-seq). We have designed our center structure to work with investigators to define the assays best suited to the immunological questions being posed, and match these with the required sample types. We will perform quality control measures on all assays, as well as generate a standard report for each assay and project. Data will be organized via our online database, Stanford Data Miner, to ensure data longevity and transferability, as well as access to both raw data files and analyzed results. Finally, we will work with investigators on novel bioinformatics approaches to mining these high-dimensional data sets. These will include approaches designed for a single data type (e.g., viSNE and Citrus for CyTOF and flow cytometry data), as well as approaches for integrating data across assays, using appropriate machine learning algorithms to aid NCI researchers in identifying immune hallmarks central to their trials. Relevance: The IMACS center will provide access to a suite of state-of-the-art immune assays, many of them developed or refined at Stanford. This unmatched set of technologies will facilitate the discovery of new biomarkers for predicting cancer outcome or therapeutic response, as well as defining potential new mechanisms of immune control of cancer.\n",
      "Summary The critical role of the immune system in effecting anti-tumor responses to traditional and molecularly-targeted therapies is increasingly recognized, such that there is now an explosion of activity in the clinical investigation of immuno-oncology. In parallel, we have gained great advances in our ability to extract useful biologic information from patient material through the development of: (i) high-throughput DNA/RNA sequencing, that has uncovered major genetic determinants of clinically-evident immune responses; (ii) highly multiplexed biological assays, that have revealed the complex interplay between immunity and cancer in relation to therapy; and (iii) advanced computational methods, that have enabled large-scale integrative data analysis. Underlying this activity is a need to identify the critical biomarkers of response and resistance to therapy in each disease and across diseases. With the rapidly evolving clinical and experimental developments in immunotherapy occurring across all major academic centers, the need to build a coordinated infrastructure for data collection, processing and analysis has emerged as a high priority. At Dana-Farber Cancer Institute (DFCI), the Center of Immuno-Oncology (CIO) was established several years ago and is a local hub of translational immuno-oncology activity. As a part of its mission, the CIO supports dedicated translational biomarker laboratories to provide expertise in the analysis of immune correlatives in the service of immuno-oncology trials. Collectively, these resources provide multidisciplinary capabilities supporting both investigator-initiated and externally-sponsored studies ranging from clinical trial design, the performance of well-validated immune monitoring capabilities on patient samples during trials, the interpretation of data from immune monitoring testing, and the development of novel assays based on the scientific questions motivated by individual clinical studies. We propose to expand the laboratory and biostatistical capabilities of the CIO along with those of additional DFCI units as a Cancer Immune Monitoring and Analysis Center (CIMAC). We want to coordinate the efforts of the proposed DFCI-CIMAC with other CIMACs and with the Cancer Immunologic Data Commons (CIDC) to develop the standards for designing studies rich with immune correlatives and informative for candidate biomarkers, to be tested in future trials. Aim 1 focuses on harmonizing the technical aspects of assays offered by the CIMACs; Aim 2 focuses on the assay-specific scoring analyses and reporting of CIMAC offerings. Modeled on our long history of implementing standard and innovative assays to study DFCI phase I/II trials, we describe a process for using `Tier 1' `shovel-ready' assays that we expect as standard components of an immune monitoring tool-kit and `Tier 2' assays that are tailored to the specific needs of a study (Aim 3). Altogether, these efforts fall within the concept and vision of personalized medicine for immuno-oncology because only by the analyses of large well-annotated datasets can we identify links between biology, molecular markers and clinical response.\n",
      "PROJECT SUMMARY Immunotherapy is transforming decades of clinical practice in cancer care, but it also comes with new questions about understanding mechanisms of action contributing to both antitumor activity and potential associated toxicity. Most importantly, identifying why only a fraction of patients derives clinical benefit is at the forefront of future developments, with the ever-elusive validation of useful clinical biomarkers as the ultimate goal. As a consortium of immunologists, technologists, clinicians, computational biologists, data specialists, and biostatisticians, the Mount Sinai CIMAC (MS-CIMAC) is uniquely positioned to generate immune profiling datasets at an unprecedented level of granularity to identify biomarker signatures of disease course and response to immunotherapy in cancer patients. MS-CIMAC will take full advantage of the smart cancer immunotherapy trial designs championed by the NCI that mandate collection of baseline and on-treatment biospecimens. Key strengths of MS-CIMAC are expected to be in unique high-dimensional yet sample-sparing approaches, including CyTOF mass cytometry, seromics, microbiome, and multiplex chromogen IHC, thanks to the implementation of cutting-edge yet validated protocols and analysis pipelines. Through a comprehensive array of assays and analytical tools that bridge innovation and standardization, MS-CIMAC intends to pursue the following three aims: a) help characterize immunocompetence at baseline and assess global immune changes during treatment, b) drill down the specificity and quality of immune responses for mechanistic evaluation of drugs, and c) automate, optimize, and integrate analyses of resulting datasets to facilitate sharing, and to ultimately discover composite immune biomarkers that will impact clinical cancer care.\n",
      "ABSTRACT Immunotherapy (IMT) has emerged as a promising treatment strategy across a broad spectrum of human cancers, with multiple agents, particularly immune check point inhibitors, showing promising results in various types of cancers. Its full potential has yet to be realized, due in part, to a lack of biomarkers predicting response to treatment. Multiple immune and genomic biomarkers of response based on analysis of pre-treatment specimens have been described but most are not very robust with significant overlap between responders and non-responders. Therefore, there is a critical unmet need to perform comprehensive characterization of candidate biomarkers in early phase IMT trials using standardized assays and novel methodologies. Recognizing the need for comprehensive immune monitoring for IMT clinical trials, in 2017 NCI developed the Cancer Immune Monitoring and Analysis Center (CIMAC) and Cancer Immunologic Data Commons (CIDC) Network, with the goal of identifying biomarkers of response, resistance, and adverse events to optimize immunotherapy approaches for patients with cancer. The MD Anderson Cancer Immune Monitoring and Analysis Center (MDA-CIMAC) is one of four CIMAC sites established five years ago that has been standardizing genomic, pathology and immunology assays and supporting profiling of tissue and blood specimens from patients treated in IMT trials. The MDA-CIMAC will be co-led by Drs. Ignacio Wistuba, renowned cancer surgical and molecular pathologist, Gheath Al-Atrash, well-known medical oncologist with expertise in stem cell transplantation and immunotherapy, and Cara Haymaker, cancer immunologist with expertise in immune biomarker analysis. They will be supported by a multidisciplinary team of world-class and highly collaborative experts on cancer and immunotherapy. The main goals of MDA-CIMAC are to: 1) provide a centralized and harmonized platform for sample collection, processing and quality assurance, and 2) use analytically-validated and standardized (Tiers 1 and 2) and highly innovative (Tier 3) assays to offer analyses for phenotypic, genomic, and functional characterization of responses of patients enrolled on IMT clinical trials. In Aim 1, we will utilize Standard Operating Procedures (SOPs) following the developed CIMAC umbrella protocol to provide services for processing and distribution of annotated biospecimens from the NCI-sponsored early phase immunotherapy clinical trials and to link the specimens to relevant clinical, pathological, immune and molecular data within the CIMAC-CIDC Network. In Aim 2, we will perform routine and innovative pathological, immunological and molecular analyses using standardized and validated and highly innovative assays to aid the completion of NCI- sponsored early phase clinical trials and the development of novel predictive IMT biomarkers. In Aim 3, in conjunction with the CIDC team, we will provide biostatistics and computational services for data collection and analysis, and will perform, interpret and predict modeling of high dimensional (‘omic”) data. We envision that the MDA-CIMAC will be indispensable as we make meaningful progress in cancer immunotherapeutic approaches.\n",
      "Summary Because cancer immunotherapy constitutes one of the most transformative therapeutic movements over the history of cancer treatment, there are now hundreds, if not thousands, of ongoing clinical trials in this field. Despite the remarkable success for a subset of patients, there are substantial gaps in our knowledge. No more than 30% of patients respond to immune checkpoint blockade and related therapies, many responding patients develop therapeutic resistance, and many responding and non-responding patients suffer from immune-related toxicities. There is a clear need to identify critical biomarkers of response, resistance, and toxicity. To improve our ability to meet this challenge, the NCI-sponsored Cancer Immune Monitoring and Analysis Center (CIMAC) program was created to enable consistent and comparable interrogation of patient biospecimens across multiple clinical trials. The program has successfully established a common set of assays harmonized across the four CIMACs, augmented this basic set with more specialized tests that may be CIMAC-specific but are still accessible across institutions, deployed these tools across dozens of phase I/II immunotherapy trials, and generated a pool of standardized data to facilitate inter-trial comparisons. The CIMAC created at Dana-Farber Cancer Institute (DFCI) built on the existing Center of Immuno-Oncology (CIO) and the long-standing collaboration with the Broad Institute of MIT and Harvard. The DFCI/Broad CIMAC, joining with the other CIMACS, now proposes to continue the network mission to compare tumor and immune response features of individual cancer types treated by multiple immunotherapeutic modalities and multiple cancer types treated by a single immunotherapeutic modality in order to dissect common and individualized mechanisms of action. One driving motivation for this renewal application is that the power to identify biomarkers increases with the number of clinical trials analyzed. Thus, Aim 1 focuses on deploying the suite of assays established over the past five years to biospecimens collected from additional innovative immunotherapy clinical trials to standardize biomarker analysis and interpretation. To broaden the scope of parameters being examined, Aim 2 proposes to develop and verify performance for novel and more complex assays to expand the network capabilities and advance discovery. Of particular value will be methods that improve multiomic single cell profiling, multiomic spatial analysis, and immunopeptidome characterization. Aim 3 addresses the challenge of integrating multiomic data across an increasing number of clinical trials by building novel computational tools that can efficiently process massive amounts of data and generate interpretable representations of the results. Altogether, these efforts will support the role of immuno-oncology in the personalized treatment of cancer because only by the analyses of large well-annotated datasets can we identify links between biology, molecular markers and clinical response.\n",
      "Abstract For this CIMAC renewal, the Stanford Cancer Immune Monitoring and Analysis Center (CIMAC) will continue to collaborate with NCI and the CIMAC/CIDC network to identify and, where appropriate, lead correlative studies for trials testing novel immunotherapy regimens. We will participate in working group calls, network meetings, and coordination with clinical teams. The Stanford CIMAC performs highly comprehensive assays of immune phenotype and function for NCI-identified clinical trials. These will include already validated and harmonized Tier 1 assays, validated Tier 2 assays, and newly proposed exploratory Tier 3 assays. For Tier 1 assays, we propose CyTOF, singleplex IHC, Olink, TCRseq and RNAseq. For Tier 2, we propose single-cell TCRseq, MIBI, ATACseq, and CyTOF proteomics. For Tier 3, we propose single-cell glycan imaging (by MALDI-ToF), spatial transcriptomics (Nanostring DSP platform), and single-cell genomics/proteomics (Mission Bio Tapestri and BD Rhapsody platforms). We will use longitudinal reference materials for tracking inter-batch and inter-project consistency. We will assess quality control measures on all assays before uploading data to the Cancer Immune Data Commons (CIDC) according to their specifications. We will also perform biostatistical analysis of results for all assays performed, in relation to clinical outcome data. For those trials where Stanford is the lead CIMAC, we will perform integrative analysis across assays, using appropriate machine learning techniques and multivariate regression algorithms such as LASSO or Elastic Net. We will work closely with the clinical teams to obtain standardized clinical data, and to disseminate and publish results in accordance with NCI guidelines.\n",
      "SUMMARY Defining who is likely to respond to cancer immunotherapy and who may not benefit, due to primary/secondary resistance or adverse events, is of critical interest for improving immuno-oncology (IO) treatments. Most clinical trials of novel combinations either fail to include any support for complex correlative biomarker discovery, often due to costs of specimen collection and assays, or are performed in an idiosyncratic manner, making non- standardized data difficult to integrate as part of a larger repository. There is a need to go beyond just PD-L1 tissue expression and tumor mutation burden as FDA-approved yet imperfect predictors of immune checkpoint blockade, by generating high-dimensional datasets that characterize the tumor immune microenvironment, systemic circulating immune and proteomic markers from blood, and host-level immune fitness information (microbiome). The Cancer Immune Monitoring and Analysis Centers, Cancer Immunologic Data Commons, and Partnership for Accelerating Cancer Therapies (CIMAC-CIDC-PACT) is a Cancer Moonshot-funded network that brings together laboratory, clinical, and computational scientists, administrators, and regulatory specialists from four institutions, Icahn School of Medicine at Mount Sinai (ISMMS), MD Anderson Cancer Center (MDACC), Dana-Farber Cancer Institute (DFCI), and Stanford University. Our unprecedented, coordinated effort to conduct IO clinical trial correlative studies and define new candidate biomarkers to improve immunotherapy had resulted in 37 clinical trials selected in the first funding period, with ongoing datasets resulted for 17 of them. This was achieved using rigorously assessed SOPs and processes from trial selection to data sharing. We propose to continue and expand these efforts as follows: A set of well-established, harmonized, high-dimensional tumor and blood-based assays will be applied to all newly selected trials, with the aim of evaluating and validating novel biomarker of treatment, clinical response, and/or adverse events (Aim 1). To spur innovation and increase granularity of our understanding of mechanism of action, high-tech yet validated single cell and spatial transcriptomic technologies as well as other omics will be proposed in exceptional patients for unbiased phenotyping, tissue composition, architectural organization, and immune profiling (Aim 2). The resulting unified data repository will be mined to define multi-omic and cross-trial markers, through biostatistical and computational innovative design to integrate datasets from various assays into higher analytical archetypes across studies (Aim 3). Together, our network is at the forefront of immune monitoring standards, by balancing innovation with reproducibility, to comprehensively assess biomarkers and strive for precision cancer treatment. We are poised to create an unparalleled resource of harmonized, clinically annotated datasets shared with the public that will establish best practice guidelines with the research community. Through rigorous implementation of state-of-the-art, cross-compared platforms, the CIMAC-CIDC initiative aims to accelerate identification of clinically actionable biomarkers of response, resistance, and adverse events, and define mechanisms at play.\n",
      "This project proposes an integrated suite of microscopy and data analysis advances that would enable quantitative, mechanistic analysis of immune-microenvironment dynamics in poor prognosis solid tumors. While immunotherapies are showing remarkable clinical responses in some advanced cancers, to date, their impact on many solid tumors has been modest. This is due, in part, to solid tumor microenvironments limiting the effectiveness of natural immune responses and immunotherapies. Yet, our understanding of the physical and molecular mechanisms governing T cell infiltration, distribution, and function in native tumor microenvironments remains extremely limited. As such, defining key T cell behaviors as a function of complex tumor microenvironments will identify design criteria that can be used to develop novel cell engineering strategies that optimize T cell-centric therapies for solid tumors. To this end the research theme of the U54 Center for Multiparametric Imaging of Tumor Immune Microenvironments (C-MITIE) is to define physical and molecular barriers to effective anti-tumor immunity and immunotherapies through advancement and development of state-of-the-art live cell and tissue optical imaging platforms and quantitative analyses. To achieve our goals, our framework brings together advanced optical imaging platforms, nano- and micro- fabrication, genome engineering, cancer immunology, and biophysical modeling. Thus, from this integrated effort we seek to define mechanisms of immune suppression that guide the development of next-generation cell-based immunotherapies.\n",
      "Project Summary - Overall The mortality rates of breast and pancreatic cancers are intrinsically tied to metastasis. In pancreatic cancer, the 5-year survival rate is only 9% and in ~70% pathological evaluations of the resected tumor, instances of venous invasion are found. Metastasis is a complex multi-step process involving cancer cells, local vasculature, and the surrounding microenvironment at multiple sites. Venous invasion in pancreatic cancer, in which cancer cells gain often invade the portal vein, is an early step in this process and provides the cells a direct path to the liver, the most common site of pancreatic cancer metastasis. As in pancreatic cancer, invasion past normal breast tissue barriers is critically tied to breast cancer outcomes. Most breast tumors can be surgically removed, and so mortality is closely tied to the extent of distant metastasis through lymphovascular invasion. The detection of lymphovascular invasion in a breast tumor correlates with poor prognosis and is not captured in current molecular analyses. The spatial organization of cancer cells, and cellular and stromal components of the tumor microenvironment near and far away from blood vessels is intrinsically three-dimensional, non-symmetric, and highly heterogeneous. In the TECH units of the Johns Hopkins Center for 3D Multiscale Cancer Imaging, we will develop a versatile 3D multiscale imaging method, CODA, which will allow us to probe the phenotypic heterogeneity of tumors from the multi-cm to the micron scale via multiplexing serial imaging. CODA can readily incorporate other imaging modalities to extract high cellular/molecular content from 3D samples. These include immunocytochemistry (CODA+IHC), immunofluorescence (CODA+IF), imaging mass cytometry (CODA+IMC) and spatial transcriptomics/proteomics (CODA+DBiT-seq). These proposed expanded versions of CODA offer a unique opportunity to produce new 3D multi-omic maps of human PDAC and breast tumors near and far from blood vessels. CODA and its integrated versions CODA+X will be tested in the RTB units of the Center in both human/mouse tissue samples and organoids, in breast and pancreatic cancer. Results from these test beds will provide novel mechanistic insights into venous invasion in breast and pancreatic cancer. Exploiting our extensive experience in data dissemination, we will make the large datasets and software produced by our CODA+X platforms and software widely available to the larger cancer research community. The units of the Center will be co-led by Johns Hopkins/Yale engineers, scientists and physicians. Substantial additional support will be provided by the Johns Hopkins U., the JH School of Medicine, the Institute for Convergence, the Department of Pathology, and the Whiting School of Engineering.\n"
     ]
    }
   ],
   "source": [
    "# Check top 10 abstracts \n",
    "for abstract in df['abstract_text'][0:10]: print(abstract)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No sign of the problem characters. There is likely a better way to clean this in a less explicitly defined way, but this works well enough to get rid of the egregious encoding oddities and can be expanded as others are identified."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Start exploring for patterns and stats to report out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "project_num                                                  1U24CA224285-01\n",
       "core_project_num                                                 U24CA224285\n",
       "appl_id                                                              9455402\n",
       "fiscal_year                                                             2017\n",
       "project_title              Translational Cancer Immune Monitoring and Ana...\n",
       "abstract_text              ABSTRACT Advances in immunotherapy have shown ...\n",
       "pref_terms                 Affect;Antibodies;Basic Science;Bioinformatics...\n",
       "org_name                                UNIVERSITY OF TX MD ANDERSON CAN CTR\n",
       "org_city                                                             HOUSTON\n",
       "org_state                                                                 TX\n",
       "org_country                                                    UNITED STATES\n",
       "principal_investigators    Gheath Al-Atrash, Cara L Haymaker, Ignacio I. ...\n",
       "program_officers                                            Magdalena Thurin\n",
       "award_amount                                                        11005230\n",
       "agency_ic_fundings                                                  11005230\n",
       "award_notice_date                                       2017-09-25T12:09:00Z\n",
       "project_start_date                                      2017-09-30T12:09:00Z\n",
       "project_end_date                                        2023-06-30T12:06:00Z\n",
       "full_foa                                                       RFA-CA-17-005\n",
       "api_source_search                                         nofo_RFA-CA-17-005\n",
       "program                             CANCERIMMUNEMONITORINGANDANALYSISCENTERS\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Look at a single row in detail\n",
    "df.loc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "project_num                object\n",
       "core_project_num           object\n",
       "appl_id                     int64\n",
       "fiscal_year                 int64\n",
       "project_title              object\n",
       "abstract_text              object\n",
       "pref_terms                 object\n",
       "org_name                   object\n",
       "org_city                   object\n",
       "org_state                  object\n",
       "org_country                object\n",
       "principal_investigators    object\n",
       "program_officers           object\n",
       "award_amount                int64\n",
       "agency_ic_fundings          int64\n",
       "award_notice_date          object\n",
       "project_start_date         object\n",
       "project_end_date           object\n",
       "full_foa                   object\n",
       "api_source_search          object\n",
       "program                    object\n",
       "dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>fiscal_year</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2001</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2002</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2003</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2004</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2005</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>2006</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2007</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>2008</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2009</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2010</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2011</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2012</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2013</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2014</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2015</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2016</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2017</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2018</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2019</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2020</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2021</td>\n",
       "      <td>697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2022</td>\n",
       "      <td>1250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2023</td>\n",
       "      <td>1462</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index  fiscal_year\n",
       "23   2000            1\n",
       "22   2001            2\n",
       "19   2002            3\n",
       "18   2003            3\n",
       "17   2004            3\n",
       "12   2005            6\n",
       "21   2006            2\n",
       "15   2007            5\n",
       "20   2008            2\n",
       "14   2009            5\n",
       "16   2010            4\n",
       "13   2011            6\n",
       "11   2012            7\n",
       "10   2013            7\n",
       "9    2014           10\n",
       "8    2015           11\n",
       "7    2016           36\n",
       "6    2017           44\n",
       "5    2018           53\n",
       "4    2019           57\n",
       "3    2020           61\n",
       "2    2021          697\n",
       "1    2022         1250\n",
       "0    2023         1462"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check fiscal years of all grants\n",
    "df['fiscal_year'].value_counts().reset_index().sort_values(by='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['project_num',\n",
       " 'core_project_num',\n",
       " 'appl_id',\n",
       " 'fiscal_year',\n",
       " 'project_title',\n",
       " 'abstract_text',\n",
       " 'pref_terms',\n",
       " 'org_name',\n",
       " 'org_city',\n",
       " 'org_state',\n",
       " 'org_country',\n",
       " 'principal_investigators',\n",
       " 'program_officers',\n",
       " 'award_amount',\n",
       " 'agency_ic_fundings',\n",
       " 'award_notice_date',\n",
       " 'project_start_date',\n",
       " 'project_end_date',\n",
       " 'full_foa',\n",
       " 'api_source_search',\n",
       " 'program']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check columns\n",
    "df.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program</th>\n",
       "      <th>core_project_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AcquiredResistancetoTherapyNetworkARTNet</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BarrettsEsophagusTranslationalResearchNetworkB...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BrainTumorSPOREGrant</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CANCERIMMUNEMONITORINGANDANALYSISCENTERS</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CancerPreventionInterceptionTargetedAgentDisco...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CellularCancerBiologyImagingResearch</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FredHutchinsonCancerResearchCenterLungSPORE</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FusionOncoproteinsinChildhoodCancersFusOnc2</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SPORETargetedTherapiesforGlioma</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SPOREinBladderCancer</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SPOREinMyeloidMalignancies</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SmallCellLungCancerSCLCConsortium</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TheAdjuvantLungCancerEnrichmentMarkerIdentific...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TheEarlyDetectionResearchNetworkEDRN</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TheMemorialSloanKetteringCancerCenterSPOREinLe...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TheUniversityofTexasMDAndersonCancerCenterSPOR...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>VanderbiltIngramCancerCenterSPOREinGastrointes...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              program  core_project_num\n",
       "0   ADMIRALStudyAdmixtureanalysisofacutelymphoblas...                 1\n",
       "1            AcquiredResistancetoTherapyNetworkARTNet                 6\n",
       "2                                             AllofUs              1866\n",
       "3   BarrettsEsophagusTranslationalResearchNetworkB...                 4\n",
       "4                                BrainTumorSPOREGrant                 1\n",
       "5            CANCERIMMUNEMONITORINGANDANALYSISCENTERS                 4\n",
       "6   CancerPreventionInterceptionTargetedAgentDisco...                 2\n",
       "7                CellularCancerBiologyImagingResearch                 4\n",
       "8         FredHutchinsonCancerResearchCenterLungSPORE                 1\n",
       "9         FusionOncoproteinsinChildhoodCancersFusOnc2                 9\n",
       "10                    SPORETargetedTherapiesforGlioma                 1\n",
       "11                               SPOREinBladderCancer                 1\n",
       "12                         SPOREinMyeloidMalignancies                 1\n",
       "13                  SmallCellLungCancerSCLCConsortium                21\n",
       "14  TheAdjuvantLungCancerEnrichmentMarkerIdentific...                 3\n",
       "15               TheEarlyDetectionResearchNetworkEDRN                32\n",
       "16  TheMemorialSloanKetteringCancerCenterSPOREinLe...                 1\n",
       "17  TheUniversityofTexasMDAndersonCancerCenterSPOR...                 1\n",
       "18  VanderbiltIngramCancerCenterSPOREinGastrointes...                 1\n",
       "19  YaleSPOREinLungCancerYSILCTheBiologyandPersona...                 1"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get number of core projects for each program\n",
    "df.groupby('program')['core_project_num'].nunique().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build a summary of all Programs and the number of projects for each"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the year from the date objects\n",
    "# Lose some resolution but make the data easier to read for stats\n",
    "df['project_start_date_year'] = df['project_start_date'].apply(lambda x: int(x[:4]))\n",
    "df['project_end_date_year'] = df['project_end_date'].apply(lambda x: int(x[:4]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy fiscal year column for later min and max stats\n",
    "df['fiscal_year_copy'] = df['fiscal_year']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define functions to apply to each column\n",
    "agg_funcs = {\n",
    "    'api_source_search': 'nunique',\n",
    "    'core_project_num': 'nunique',\n",
    "    'project_num': 'nunique',\n",
    "    'agency_ic_fundings': 'sum', # This will be a little off due to duplicates!\n",
    "    'project_start_date_year': 'min',\n",
    "    'project_end_date_year': 'max',\n",
    "    'fiscal_year': 'min',\n",
    "    'fiscal_year_copy': 'max'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Define column titles better suited for reporting\n",
    "# rename_dict = {\n",
    "#     \"program\": \"Program\",\n",
    "#     \"core_project_num\": \"Core Project Count\",\n",
    "#     \"project_num\": \"Grant/Award Count\",\n",
    "#     'agency_ic_fundings': \"Total NCI Funding (since 2000)\", # This will be a little off due to duplicates! \n",
    "#     \"api_source_search\": \"Provided NOFOs/Awards with Associated Grants\",\n",
    "#     \"project_start_date_year\": \"Earliest Project Start Date\",\n",
    "#     \"project_end_date_year\": \"Latest Project End Date\",\n",
    "#     \"fiscal_year\": \"Earliest Fiscal Year\",\n",
    "#     \"fiscal_year_copy\": \"Latest Fiscal Year\"\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group by 'program' and apply aggregation functions defined above\n",
    "summary_stat_df = df.groupby('program').agg(agg_funcs).reset_index()\n",
    "\n",
    "# # Rename columns for better presentation as defined above\n",
    "# summary_stat_df.rename(columns=rename_dict, inplace=True)\n",
    "\n",
    "# # Store program summary \n",
    "# program_summary_filename = reports_dir + '/' + 'programSummaryStats.csv'\n",
    "# summary_stat_df.to_csv(program_summary_filename, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program</th>\n",
       "      <th>api_source_search</th>\n",
       "      <th>core_project_num</th>\n",
       "      <th>project_num</th>\n",
       "      <th>agency_ic_fundings</th>\n",
       "      <th>project_start_date_year</th>\n",
       "      <th>project_end_date_year</th>\n",
       "      <th>fiscal_year</th>\n",
       "      <th>fiscal_year_copy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2107456</td>\n",
       "      <td>2020</td>\n",
       "      <td>2023</td>\n",
       "      <td>2020</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AcquiredResistancetoTherapyNetworkARTNet</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>7357128</td>\n",
       "      <td>2017</td>\n",
       "      <td>2027</td>\n",
       "      <td>2022</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>6</td>\n",
       "      <td>1866</td>\n",
       "      <td>3208</td>\n",
       "      <td>1440658962</td>\n",
       "      <td>1981</td>\n",
       "      <td>2028</td>\n",
       "      <td>2018</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BarrettsEsophagusTranslationalResearchNetworkB...</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>40</td>\n",
       "      <td>38637870</td>\n",
       "      <td>2011</td>\n",
       "      <td>2024</td>\n",
       "      <td>2011</td>\n",
       "      <td>2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BrainTumorSPOREGrant</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "      <td>44541554</td>\n",
       "      <td>2002</td>\n",
       "      <td>2023</td>\n",
       "      <td>2002</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CANCERIMMUNEMONITORINGANDANALYSISCENTERS</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>55551765</td>\n",
       "      <td>2017</td>\n",
       "      <td>2028</td>\n",
       "      <td>2017</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CancerPreventionInterceptionTargetedAgentDisco...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2412236</td>\n",
       "      <td>2022</td>\n",
       "      <td>2027</td>\n",
       "      <td>2022</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CellularCancerBiologyImagingResearch</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>12810400</td>\n",
       "      <td>2021</td>\n",
       "      <td>2026</td>\n",
       "      <td>2021</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FredHutchinsonCancerResearchCenterLungSPORE</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>12339863</td>\n",
       "      <td>2019</td>\n",
       "      <td>2024</td>\n",
       "      <td>2019</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FusionOncoproteinsinChildhoodCancersFusOnc2</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>16</td>\n",
       "      <td>73063983</td>\n",
       "      <td>2018</td>\n",
       "      <td>2024</td>\n",
       "      <td>2018</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SPORETargetedTherapiesforGlioma</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>21143737</td>\n",
       "      <td>2013</td>\n",
       "      <td>2024</td>\n",
       "      <td>2013</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SPOREinBladderCancer</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>10900913</td>\n",
       "      <td>2018</td>\n",
       "      <td>2024</td>\n",
       "      <td>2018</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SPOREinMyeloidMalignancies</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>10644690</td>\n",
       "      <td>2017</td>\n",
       "      <td>2023</td>\n",
       "      <td>2017</td>\n",
       "      <td>2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SmallCellLungCancerSCLCConsortium</td>\n",
       "      <td>5</td>\n",
       "      <td>21</td>\n",
       "      <td>93</td>\n",
       "      <td>53617109</td>\n",
       "      <td>2017</td>\n",
       "      <td>2027</td>\n",
       "      <td>2017</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TheAdjuvantLungCancerEnrichmentMarkerIdentific...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>69</td>\n",
       "      <td>294483788</td>\n",
       "      <td>1983</td>\n",
       "      <td>2025</td>\n",
       "      <td>2000</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TheEarlyDetectionResearchNetworkEDRN</td>\n",
       "      <td>7</td>\n",
       "      <td>32</td>\n",
       "      <td>149</td>\n",
       "      <td>130748148</td>\n",
       "      <td>2000</td>\n",
       "      <td>2027</td>\n",
       "      <td>2016</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TheMemorialSloanKetteringCancerCenterSPOREinLe...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>6158913</td>\n",
       "      <td>2021</td>\n",
       "      <td>2026</td>\n",
       "      <td>2021</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TheUniversityofTexasMDAndersonCancerCenterSPOR...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>9063325</td>\n",
       "      <td>2019</td>\n",
       "      <td>2024</td>\n",
       "      <td>2019</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>VanderbiltIngramCancerCenterSPOREinGastrointes...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>11590546</td>\n",
       "      <td>2019</td>\n",
       "      <td>2024</td>\n",
       "      <td>2019</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>20540459</td>\n",
       "      <td>2015</td>\n",
       "      <td>2025</td>\n",
       "      <td>2015</td>\n",
       "      <td>2023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              program  api_source_search  \\\n",
       "0   ADMIRALStudyAdmixtureanalysisofacutelymphoblas...                  1   \n",
       "1            AcquiredResistancetoTherapyNetworkARTNet                  2   \n",
       "2                                             AllofUs                  6   \n",
       "3   BarrettsEsophagusTranslationalResearchNetworkB...                  3   \n",
       "4                                BrainTumorSPOREGrant                  1   \n",
       "5            CANCERIMMUNEMONITORINGANDANALYSISCENTERS                  2   \n",
       "6   CancerPreventionInterceptionTargetedAgentDisco...                  1   \n",
       "7                CellularCancerBiologyImagingResearch                  1   \n",
       "8         FredHutchinsonCancerResearchCenterLungSPORE                  1   \n",
       "9         FusionOncoproteinsinChildhoodCancersFusOnc2                  2   \n",
       "10                    SPORETargetedTherapiesforGlioma                  1   \n",
       "11                               SPOREinBladderCancer                  1   \n",
       "12                         SPOREinMyeloidMalignancies                  1   \n",
       "13                  SmallCellLungCancerSCLCConsortium                  5   \n",
       "14  TheAdjuvantLungCancerEnrichmentMarkerIdentific...                  3   \n",
       "15               TheEarlyDetectionResearchNetworkEDRN                  7   \n",
       "16  TheMemorialSloanKetteringCancerCenterSPOREinLe...                  1   \n",
       "17  TheUniversityofTexasMDAndersonCancerCenterSPOR...                  1   \n",
       "18  VanderbiltIngramCancerCenterSPOREinGastrointes...                  1   \n",
       "19  YaleSPOREinLungCancerYSILCTheBiologyandPersona...                  1   \n",
       "\n",
       "    core_project_num  project_num  agency_ic_fundings  \\\n",
       "0                  1            4             2107456   \n",
       "1                  6            6             7357128   \n",
       "2               1866         3208          1440658962   \n",
       "3                  4           40            38637870   \n",
       "4                  1           26            44541554   \n",
       "5                  4            8            55551765   \n",
       "6                  2            2             2412236   \n",
       "7                  4            8            12810400   \n",
       "8                  1            6            12339863   \n",
       "9                  9           16            73063983   \n",
       "10                 1           13            21143737   \n",
       "11                 1            5            10900913   \n",
       "12                 1            6            10644690   \n",
       "13                21           93            53617109   \n",
       "14                 3           69           294483788   \n",
       "15                32          149           130748148   \n",
       "16                 1            3             6158913   \n",
       "17                 1            4             9063325   \n",
       "18                 1            5            11590546   \n",
       "19                 1           11            20540459   \n",
       "\n",
       "    project_start_date_year  project_end_date_year  fiscal_year  \\\n",
       "0                      2020                   2023         2020   \n",
       "1                      2017                   2027         2022   \n",
       "2                      1981                   2028         2018   \n",
       "3                      2011                   2024         2011   \n",
       "4                      2002                   2023         2002   \n",
       "5                      2017                   2028         2017   \n",
       "6                      2022                   2027         2022   \n",
       "7                      2021                   2026         2021   \n",
       "8                      2019                   2024         2019   \n",
       "9                      2018                   2024         2018   \n",
       "10                     2013                   2024         2013   \n",
       "11                     2018                   2024         2018   \n",
       "12                     2017                   2023         2017   \n",
       "13                     2017                   2027         2017   \n",
       "14                     1983                   2025         2000   \n",
       "15                     2000                   2027         2016   \n",
       "16                     2021                   2026         2021   \n",
       "17                     2019                   2024         2019   \n",
       "18                     2019                   2024         2019   \n",
       "19                     2015                   2025         2015   \n",
       "\n",
       "    fiscal_year_copy  \n",
       "0               2022  \n",
       "1               2022  \n",
       "2               2023  \n",
       "3               2021  \n",
       "4               2022  \n",
       "5               2023  \n",
       "6               2022  \n",
       "7               2023  \n",
       "8               2023  \n",
       "9               2022  \n",
       "10              2022  \n",
       "11              2022  \n",
       "12              2021  \n",
       "13              2023  \n",
       "14              2023  \n",
       "15              2023  \n",
       "16              2023  \n",
       "17              2022  \n",
       "18              2023  \n",
       "19              2023  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_stat_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the most part, Earliest Grant Date aligns with earliest Fiscal Year. Outliers are the very early start dates for All of Us, ALCHEMIST, and EDRN programs compared to the earlist fiscal year. This might indicate a project that received additional funds or supplements many years after the proejct began. Could this be data deposition or a similar modernization effort?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Enrich Key Program CSV with Grants data\n",
    "This will begin to mock up the INS Key Programs page as requested by ODS. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use config to get path and load clean key programs csv \n",
    "key_programs_filename = '../' + config.CLEANED_KEY_PROGRAMS_CSV\n",
    "key_programs_df = pd.read_csv(key_programs_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a temporary program ID column to use for mapping stats to key programs\n",
    "key_programs_df['program'] = key_programs_df['program_name'].apply(lambda name: ''.join(filter(str.isalnum, name)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['program_name',\n",
       " 'program_acronym',\n",
       " 'focus_area',\n",
       " 'doc',\n",
       " 'contact_pi',\n",
       " 'contact_pi_email',\n",
       " 'contact_nih',\n",
       " 'contact_nih_email',\n",
       " 'nofo',\n",
       " 'award',\n",
       " 'program_link',\n",
       " 'data_link',\n",
       " 'cancer_type',\n",
       " 'program']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check columns in key programs df\n",
    "key_programs_df.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['program',\n",
       " 'api_source_search',\n",
       " 'core_project_num',\n",
       " 'project_num',\n",
       " 'agency_ic_fundings',\n",
       " 'project_start_date_year',\n",
       " 'project_end_date_year',\n",
       " 'fiscal_year',\n",
       " 'fiscal_year_copy']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check columns in program summary stats df\n",
    "summary_stat_df.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add stats to key programs df using temp program ID as connector\n",
    "key_programs_enriched = pd.merge(left=key_programs_df, \n",
    "                                right=summary_stat_df[['program',\n",
    "                                                       'core_project_num', \n",
    "                                                       'project_num',\n",
    "                                                       'agency_ic_fundings']], \n",
    "                                on='program')\n",
    "# Drop temp program ID column\n",
    "key_programs_enriched.drop(columns='program', inplace=True)\n",
    "\n",
    "# Rename stat columns \n",
    "stat_renamer_dict = {\n",
    "    'core_project_num': 'core_project_count',\n",
    "    'project_num': 'grant_count',\n",
    "    'agency_ic_fundings': 'total_nci_funding',\n",
    "}\n",
    "key_programs_enriched.rename(columns=stat_renamer_dict, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['program_name',\n",
       " 'program_acronym',\n",
       " 'focus_area',\n",
       " 'doc',\n",
       " 'contact_pi',\n",
       " 'contact_pi_email',\n",
       " 'contact_nih',\n",
       " 'contact_nih_email',\n",
       " 'nofo',\n",
       " 'award',\n",
       " 'program_link',\n",
       " 'data_link',\n",
       " 'cancer_type',\n",
       " 'core_project_count',\n",
       " 'grant_count',\n",
       " 'total_nci_funding']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "key_programs_enriched.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store enriched Key Programs Stats Table\n",
    "key_program_stats_filename = reports_dir + '/' + 'keyProgramStats.csv'\n",
    "key_programs_enriched.to_csv(key_program_stats_filename, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Continue exploring data report options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a df of all grants with project start dates before year 2000\n",
    "early_project_df = df[df['project_start_date_year'] < 2000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Export to csv for quick ad-hoc analysis\n",
    "# early_project_df.to_csv('earlyProjectReport.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program</th>\n",
       "      <th>api_source_search</th>\n",
       "      <th>core_project_num</th>\n",
       "      <th>project_title</th>\n",
       "      <th>project_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>nofo_PA20-185</td>\n",
       "      <td>R01CA031845</td>\n",
       "      <td>Synthetic Studies Related to Cancer Research/T...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>nofo_PA20-185</td>\n",
       "      <td>R01CA047296</td>\n",
       "      <td>A Pathway of Tumor Suppression</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>nofo_PA20-185</td>\n",
       "      <td>R01CA053840</td>\n",
       "      <td>Protein Tyrosine Dephosphorylation &amp; Signal Tr...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>nofo_PA20-185</td>\n",
       "      <td>R01CA067007</td>\n",
       "      <td>Mismatch Repair and Carcinogenesis</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>nofo_PA20-185</td>\n",
       "      <td>R01CA067985</td>\n",
       "      <td>DNA Damage Repair by MUTYH and MUTYH Variants ...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>P30CA082103</td>\n",
       "      <td>Project HOPE: The Pediatric/AYA Omics Project</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>P30CA082709</td>\n",
       "      <td>The Big Ten Electronic Health Record Consortiu...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U24CA055727</td>\n",
       "      <td>CHILDHOOD CANCER SURVIVOR STUDY: Somatic and G...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>TheAdjuvantLungCancerEnrichmentMarkerIdentific...</td>\n",
       "      <td>award_U10CA031946</td>\n",
       "      <td>U10CA031946</td>\n",
       "      <td>CANCER AND LEUKEMIA GROUP B</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>TheAdjuvantLungCancerEnrichmentMarkerIdentific...</td>\n",
       "      <td>award_U10CA031946</td>\n",
       "      <td>U10CA031946</td>\n",
       "      <td>Cancer and Leukemia Group B</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>105 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               program  api_source_search  \\\n",
       "0                                              AllofUs      nofo_PA20-185   \n",
       "1                                              AllofUs      nofo_PA20-185   \n",
       "2                                              AllofUs      nofo_PA20-185   \n",
       "3                                              AllofUs      nofo_PA20-185   \n",
       "4                                              AllofUs      nofo_PA20-185   \n",
       "..                                                 ...                ...   \n",
       "100                                            AllofUs      nofo_PA20-272   \n",
       "101                                            AllofUs      nofo_PA20-272   \n",
       "102                                            AllofUs      nofo_PA20-272   \n",
       "103  TheAdjuvantLungCancerEnrichmentMarkerIdentific...  award_U10CA031946   \n",
       "104  TheAdjuvantLungCancerEnrichmentMarkerIdentific...  award_U10CA031946   \n",
       "\n",
       "    core_project_num                                      project_title  \\\n",
       "0        R01CA031845  Synthetic Studies Related to Cancer Research/T...   \n",
       "1        R01CA047296                     A Pathway of Tumor Suppression   \n",
       "2        R01CA053840  Protein Tyrosine Dephosphorylation & Signal Tr...   \n",
       "3        R01CA067007                 Mismatch Repair and Carcinogenesis   \n",
       "4        R01CA067985  DNA Damage Repair by MUTYH and MUTYH Variants ...   \n",
       "..               ...                                                ...   \n",
       "100      P30CA082103      Project HOPE: The Pediatric/AYA Omics Project   \n",
       "101      P30CA082709  The Big Ten Electronic Health Record Consortiu...   \n",
       "102      U24CA055727  CHILDHOOD CANCER SURVIVOR STUDY: Somatic and G...   \n",
       "103      U10CA031946                        CANCER AND LEUKEMIA GROUP B   \n",
       "104      U10CA031946                        Cancer and Leukemia Group B   \n",
       "\n",
       "     project_num  \n",
       "0              2  \n",
       "1              2  \n",
       "2              2  \n",
       "3              2  \n",
       "4              3  \n",
       "..           ...  \n",
       "100            1  \n",
       "101            1  \n",
       "102            1  \n",
       "103            5  \n",
       "104           21  \n",
       "\n",
       "[105 rows x 5 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group to find patterns and common sources for the projects with very early start dates\n",
    "early_project_df.groupby(['program', 'api_source_search', 'core_project_num', 'project_title'])['project_num'].nunique().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check for \"Bottleneck Effect\" grants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_rows_with_different_values(df, shared_column, compare_column):\n",
    "    \"\"\"Find rows that share a value in a specified column\n",
    "    but have different values in another specified column.\n",
    "    \"\"\"\n",
    "\n",
    "    grouped = df.groupby(shared_column)[compare_column].transform('nunique')\n",
    "    selected_row_df = df[grouped > 1]\n",
    "\n",
    "    return selected_row_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for Core Projects found in multiple programs and compare sources\n",
    "\n",
    "shared_column = 'core_project_num'\n",
    "compare_column = 'program'\n",
    "\n",
    "# Get grant-level rows with same project but different program\n",
    "df_shared = find_rows_with_different_values(df, shared_column, compare_column)\n",
    "\n",
    "# Group with the provided search value and count unique grants \n",
    "df_shared_projects = (df_shared.groupby(\n",
    "                        ['api_source_search', shared_column, compare_column])\n",
    "                        .size().reset_index()\n",
    "                        .rename(columns={0:'grant_count'}))\n",
    "\n",
    "# Store shared programs\n",
    "shared_projects_filename = reports_dir + '/' + 'sharedProjects.csv'\n",
    "df_shared_projects.to_csv(shared_projects_filename, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For any Core Project gathered for more than one program (e.g. shared NOFOs or a NOFO in one program and an Award in another), the grants and all downstream outputs will need to be associated with BOTH programs. It's important to identify these. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>api_source_search</th>\n",
       "      <th>core_project_num</th>\n",
       "      <th>program</th>\n",
       "      <th>grant_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>award_R01CA239701</td>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>award_U10CA180821</td>\n",
       "      <td>U10CA180821</td>\n",
       "      <td>TheAdjuvantLungCancerEnrichmentMarkerIdentific...</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U10CA180821</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U24CA163056</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U54CA163004</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U54CA163059</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U54CA163060</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U54CA224019</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U54CA231630</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U54CA231637</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U54CA231638</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U54CA231652</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U54CA243125</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U54CA268083</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>nofo_PA20-272</td>\n",
       "      <td>U54CA268084</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>nofo_RFA-CA-10-014</td>\n",
       "      <td>U54CA163004</td>\n",
       "      <td>BarrettsEsophagusTranslationalResearchNetworkB...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>nofo_RFA-CA-10-014</td>\n",
       "      <td>U54CA163059</td>\n",
       "      <td>BarrettsEsophagusTranslationalResearchNetworkB...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>nofo_RFA-CA-10-014</td>\n",
       "      <td>U54CA163060</td>\n",
       "      <td>BarrettsEsophagusTranslationalResearchNetworkB...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>nofo_RFA-CA-16-006</td>\n",
       "      <td>U54CA163004</td>\n",
       "      <td>BarrettsEsophagusTranslationalResearchNetworkB...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>nofo_RFA-CA-16-006</td>\n",
       "      <td>U54CA163059</td>\n",
       "      <td>BarrettsEsophagusTranslationalResearchNetworkB...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>nofo_RFA-CA-16-006</td>\n",
       "      <td>U54CA163060</td>\n",
       "      <td>BarrettsEsophagusTranslationalResearchNetworkB...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>nofo_RFA-CA-16-007</td>\n",
       "      <td>U24CA163056</td>\n",
       "      <td>BarrettsEsophagusTranslationalResearchNetworkB...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>nofo_RFA-CA-17-049</td>\n",
       "      <td>U54CA231637</td>\n",
       "      <td>FusionOncoproteinsinChildhoodCancersFusOnc2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>nofo_RFA-CA-17-049</td>\n",
       "      <td>U54CA231638</td>\n",
       "      <td>FusionOncoproteinsinChildhoodCancersFusOnc2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>nofo_RFA-CA-17-049</td>\n",
       "      <td>U54CA231652</td>\n",
       "      <td>FusionOncoproteinsinChildhoodCancersFusOnc2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>nofo_RFA-CA-19-016</td>\n",
       "      <td>U54CA231630</td>\n",
       "      <td>FusionOncoproteinsinChildhoodCancersFusOnc2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>nofo_RFA-CA-19-016</td>\n",
       "      <td>U54CA243125</td>\n",
       "      <td>FusionOncoproteinsinChildhoodCancersFusOnc2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>nofo_RFA-CA-21-002</td>\n",
       "      <td>U54CA268083</td>\n",
       "      <td>CellularCancerBiologyImagingResearch</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>nofo_RFA-CA-21-002</td>\n",
       "      <td>U54CA268084</td>\n",
       "      <td>CellularCancerBiologyImagingResearch</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>nofo_RFA-CA-21-052</td>\n",
       "      <td>U54CA224019</td>\n",
       "      <td>AcquiredResistancetoTherapyNetworkARTNet</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     api_source_search core_project_num  \\\n",
       "0    award_R01CA239701      R01CA239701   \n",
       "1    award_U10CA180821      U10CA180821   \n",
       "2        nofo_PA20-272      R01CA239701   \n",
       "3        nofo_PA20-272      U10CA180821   \n",
       "4        nofo_PA20-272      U24CA163056   \n",
       "5        nofo_PA20-272      U54CA163004   \n",
       "6        nofo_PA20-272      U54CA163059   \n",
       "7        nofo_PA20-272      U54CA163060   \n",
       "8        nofo_PA20-272      U54CA224019   \n",
       "9        nofo_PA20-272      U54CA231630   \n",
       "10       nofo_PA20-272      U54CA231637   \n",
       "11       nofo_PA20-272      U54CA231638   \n",
       "12       nofo_PA20-272      U54CA231652   \n",
       "13       nofo_PA20-272      U54CA243125   \n",
       "14       nofo_PA20-272      U54CA268083   \n",
       "15       nofo_PA20-272      U54CA268084   \n",
       "16  nofo_RFA-CA-10-014      U54CA163004   \n",
       "17  nofo_RFA-CA-10-014      U54CA163059   \n",
       "18  nofo_RFA-CA-10-014      U54CA163060   \n",
       "19  nofo_RFA-CA-16-006      U54CA163004   \n",
       "20  nofo_RFA-CA-16-006      U54CA163059   \n",
       "21  nofo_RFA-CA-16-006      U54CA163060   \n",
       "22  nofo_RFA-CA-16-007      U24CA163056   \n",
       "23  nofo_RFA-CA-17-049      U54CA231637   \n",
       "24  nofo_RFA-CA-17-049      U54CA231638   \n",
       "25  nofo_RFA-CA-17-049      U54CA231652   \n",
       "26  nofo_RFA-CA-19-016      U54CA231630   \n",
       "27  nofo_RFA-CA-19-016      U54CA243125   \n",
       "28  nofo_RFA-CA-21-002      U54CA268083   \n",
       "29  nofo_RFA-CA-21-002      U54CA268084   \n",
       "30  nofo_RFA-CA-21-052      U54CA224019   \n",
       "\n",
       "                                              program  grant_count  \n",
       "0   ADMIRALStudyAdmixtureanalysisofacutelymphoblas...            4  \n",
       "1   TheAdjuvantLungCancerEnrichmentMarkerIdentific...           35  \n",
       "2                                             AllofUs            2  \n",
       "3                                             AllofUs            8  \n",
       "4                                             AllofUs            1  \n",
       "5                                             AllofUs            1  \n",
       "6                                             AllofUs            1  \n",
       "7                                             AllofUs            1  \n",
       "8                                             AllofUs            1  \n",
       "9                                             AllofUs            2  \n",
       "10                                            AllofUs            2  \n",
       "11                                            AllofUs            3  \n",
       "12                                            AllofUs            1  \n",
       "13                                            AllofUs            1  \n",
       "14                                            AllofUs            1  \n",
       "15                                            AllofUs            1  \n",
       "16  BarrettsEsophagusTranslationalResearchNetworkB...            7  \n",
       "17  BarrettsEsophagusTranslationalResearchNetworkB...            6  \n",
       "18  BarrettsEsophagusTranslationalResearchNetworkB...            6  \n",
       "19  BarrettsEsophagusTranslationalResearchNetworkB...            6  \n",
       "20  BarrettsEsophagusTranslationalResearchNetworkB...            5  \n",
       "21  BarrettsEsophagusTranslationalResearchNetworkB...            5  \n",
       "22  BarrettsEsophagusTranslationalResearchNetworkB...            5  \n",
       "23        FusionOncoproteinsinChildhoodCancersFusOnc2            1  \n",
       "24        FusionOncoproteinsinChildhoodCancersFusOnc2            1  \n",
       "25        FusionOncoproteinsinChildhoodCancersFusOnc2            1  \n",
       "26        FusionOncoproteinsinChildhoodCancersFusOnc2            3  \n",
       "27        FusionOncoproteinsinChildhoodCancersFusOnc2            3  \n",
       "28               CellularCancerBiologyImagingResearch            2  \n",
       "29               CellularCancerBiologyImagingResearch            2  \n",
       "30           AcquiredResistancetoTherapyNetworkARTNet            1  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show core projects found in multiple programs\n",
    "df_shared_projects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some Key Programs specify Awards for grants rather than the entire NOFO. Only the Award and downstream grants should be associated with that Program, rather than everything dowstream of the larger NOFO.  \n",
    "However, all grants gathered from NIH RePORTER necessarily have a 'full_foa' (aka NOFO) value within the RePORTER data. This is a potential for complications.  \n",
    "\n",
    "We need to check for two things:  \n",
    "1. Did we search NIH RePORTER with any NOFOs that do not match the full_foa within grants returned? \n",
    "    - That would be unexpected and indicate an API gathering problem.\n",
    "    - Would appear in the table below as a rows with \"nofo_\"... in the api_source_search column\n",
    "2. Did we search NIH RePORTER with any Awards with differing full_foa values within grants returned? \n",
    "    - That's acceptable - it could indicate that the particular Core Project received funding from many sources (e.g. data sharing NOFO effort)\n",
    "    - It's still important to note. We need to be careful not to accidentally make \"upstream\" connections from grants back to full_foas back to programs for projects derived from specified Award gathering.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program_1</th>\n",
       "      <th>program_2</th>\n",
       "      <th>shared_project_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FusionOncoproteinsinChildhoodCancersFusOnc2</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>BarrettsEsophagusTranslationalResearchNetworkB...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CellularCancerBiologyImagingResearch</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>AllofUs</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>TheAdjuvantLungCancerEnrichmentMarkerIdentific...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>AcquiredResistancetoTherapyNetworkARTNet</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           program_1  \\\n",
       "0        FusionOncoproteinsinChildhoodCancersFusOnc2   \n",
       "1                                            AllofUs   \n",
       "2               CellularCancerBiologyImagingResearch   \n",
       "3  ADMIRALStudyAdmixtureanalysisofacutelymphoblas...   \n",
       "4                                            AllofUs   \n",
       "5                                            AllofUs   \n",
       "\n",
       "                                           program_2  shared_project_count  \n",
       "0                                            AllofUs                     5  \n",
       "1  BarrettsEsophagusTranslationalResearchNetworkB...                     4  \n",
       "2                                            AllofUs                     2  \n",
       "3                                            AllofUs                     1  \n",
       "4  TheAdjuvantLungCancerEnrichmentMarkerIdentific...                     1  \n",
       "5           AcquiredResistancetoTherapyNetworkARTNet                     1  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by core project and combine programs into unordered set\n",
    "grouped = df.groupby(\"core_project_num\")[\"program\"].apply(set)\n",
    "\n",
    "# Create combos of programs for each core project with itertools combinations\n",
    "df_shared_programs = grouped.apply(lambda x: list(combinations(x, 2)))\n",
    "\n",
    "# Flatten the program combos and count occurrences\n",
    "df_shared_programs = df_shared_programs.explode().value_counts().reset_index()\n",
    "\n",
    "# Split program combo column into two separate columns\n",
    "# NOTE this will need to be reworked if a project has more than 2 programs\n",
    "df_shared_programs[['program_1','program_2']] = df_shared_programs['index'].apply(pd.Series)\n",
    "\n",
    "# Reorder and rename\n",
    "df_shared_programs.rename(columns={'program':'shared_project_count'}, inplace=True)\n",
    "df_shared_programs = df_shared_programs[['program_1','program_2','shared_project_count']]\n",
    "\n",
    "# Export as report\n",
    "shared_programs_filename = reports_dir + '/' + 'sharedProjectsByProgramPair.csv'\n",
    "df_shared_programs.to_csv(shared_programs_filename, index=False)\n",
    "\n",
    "# Show table of program combos and number of shared projects between them\n",
    "df_shared_programs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>api_source_search</th>\n",
       "      <th>full_foa</th>\n",
       "      <th>grant_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>award_P50CA097257</td>\n",
       "      <td>PAR-00-087</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>award_P50CA097257</td>\n",
       "      <td>PAR-05-156</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>award_P50CA097257</td>\n",
       "      <td>PAR-10-003</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>award_P50CA097257</td>\n",
       "      <td>PAR-14-353</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>award_P50CA165962</td>\n",
       "      <td>PA-18-906</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>award_P50CA165962</td>\n",
       "      <td>PA-21-071</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>award_P50CA165962</td>\n",
       "      <td>PAR-10-003</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>award_P50CA165962</td>\n",
       "      <td>PAR-18-313</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>award_P50CA196530</td>\n",
       "      <td>PAR-14-031</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>award_P50CA196530</td>\n",
       "      <td>PAR-18-313</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>award_P50CA196530</td>\n",
       "      <td>RFA-CA-16-012</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>award_R01CA239701</td>\n",
       "      <td>PA-19-056</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>award_R01CA239701</td>\n",
       "      <td>PA-20-272</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>award_R01CA239701</td>\n",
       "      <td>PA-21-071</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>award_U10CA180821</td>\n",
       "      <td>PA-20-272</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>award_U10CA180821</td>\n",
       "      <td>RFA-CA-12-010</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>award_U10CA180821</td>\n",
       "      <td>RFA-CA-17-056</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    api_source_search       full_foa  grant_count\n",
       "0   award_P50CA097257     PAR-00-087            9\n",
       "1   award_P50CA097257     PAR-05-156            7\n",
       "2   award_P50CA097257     PAR-10-003            5\n",
       "3   award_P50CA097257     PAR-14-353            5\n",
       "4   award_P50CA165962      PA-18-906            1\n",
       "5   award_P50CA165962      PA-21-071            1\n",
       "6   award_P50CA165962     PAR-10-003            5\n",
       "7   award_P50CA165962     PAR-18-313            6\n",
       "8   award_P50CA196530     PAR-14-031            6\n",
       "9   award_P50CA196530     PAR-18-313            4\n",
       "10  award_P50CA196530  RFA-CA-16-012            1\n",
       "11  award_R01CA239701      PA-19-056            1\n",
       "12  award_R01CA239701      PA-20-272            2\n",
       "13  award_R01CA239701      PA-21-071            1\n",
       "14  award_U10CA180821      PA-20-272            8\n",
       "15  award_U10CA180821  RFA-CA-12-010           16\n",
       "16  award_U10CA180821  RFA-CA-17-056           11"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for rows with the same ODS-provided NOFO but different FOA/NOFO accordign to RePORTER \n",
    "shared_column = 'api_source_search'\n",
    "compare_column = 'full_foa'\n",
    "\n",
    "# Select only rows with different values\n",
    "df_shared = find_rows_with_different_values(df, shared_column, compare_column)\n",
    "# Group and summarize\n",
    "df_shared.groupby([shared_column, compare_column]).size().reset_index().rename(columns={0:'grant_count'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build visualizations in Plotly\n",
    "The goal is to communicate patterns and highlight limitations or oddities within the grants data in order to provide feedback and capabilities to ODS.  \n",
    "Visualizations may be a better route than spreadsheets to summarize this. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Program Funding"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pie Chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Grouping data to calculate the sum of total_cost for each program\n",
    "# program_agency_funding = df.groupby('program')['agency_ic_fundings'].sum().reset_index()\n",
    "\n",
    "# # Creating the Pie Chart\n",
    "# fig = px.pie(program_agency_funding, \n",
    "#              names='program', \n",
    "#              values='agency_ic_fundings', \n",
    "#              title='Distribution of Agency IC Fundings Across Programs',\n",
    "#              height=600,\n",
    "#              width=1200,\n",
    "#              )\n",
    "\n",
    "# # Show the Pie Chart\n",
    "# fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### TreeMap Block Chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Grouping data to calculate the sum of total_cost for each program\n",
    "# # program_core_funding = df.groupby(['program', 'core_project_num'])['agency_ic_fundings'].sum().reset_index()\n",
    "\n",
    "# # Creating the Treemap\n",
    "# fig = px.treemap(df, \n",
    "#                  path=['program', 'project_title', 'project_num'], \n",
    "#                  values='agency_ic_fundings',\n",
    "#                  title='Distribution of NCI Funding by Program, Project, and Award',\n",
    "#                  height=1200,\n",
    "#                  width=1800)\n",
    "\n",
    "# # Show the Treemap\n",
    "# fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>org_state</th>\n",
       "      <th>agency_ic_fundings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL</td>\n",
       "      <td>9095775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AR</td>\n",
       "      <td>2305873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AZ</td>\n",
       "      <td>19291308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BC</td>\n",
       "      <td>1313155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CA</td>\n",
       "      <td>276507528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CO</td>\n",
       "      <td>9434500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CT</td>\n",
       "      <td>46490803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DC</td>\n",
       "      <td>17706480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DE</td>\n",
       "      <td>823703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FL</td>\n",
       "      <td>47141005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GA</td>\n",
       "      <td>23413466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HI</td>\n",
       "      <td>6867343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>IA</td>\n",
       "      <td>7413559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>IL</td>\n",
       "      <td>148242112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>IN</td>\n",
       "      <td>19214283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>KS</td>\n",
       "      <td>2128229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>KY</td>\n",
       "      <td>12310190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>LA</td>\n",
       "      <td>6351656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MA</td>\n",
       "      <td>385381611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MD</td>\n",
       "      <td>38716389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ME</td>\n",
       "      <td>4043314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MI</td>\n",
       "      <td>58369519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MN</td>\n",
       "      <td>47962857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>MO</td>\n",
       "      <td>29368698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NC</td>\n",
       "      <td>51819254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NE</td>\n",
       "      <td>9716876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NH</td>\n",
       "      <td>8465229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NJ</td>\n",
       "      <td>17177724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NM</td>\n",
       "      <td>1943628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NY</td>\n",
       "      <td>267589842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>OH</td>\n",
       "      <td>70510115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>OK</td>\n",
       "      <td>7928144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ON</td>\n",
       "      <td>2448767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>OR</td>\n",
       "      <td>22473875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>PA</td>\n",
       "      <td>121206288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>PQ</td>\n",
       "      <td>1409518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>PR</td>\n",
       "      <td>627180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>RI</td>\n",
       "      <td>3078495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>SC</td>\n",
       "      <td>10582120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>TN</td>\n",
       "      <td>69343936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>TX</td>\n",
       "      <td>185164698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>UT</td>\n",
       "      <td>22617397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>VA</td>\n",
       "      <td>23311906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>VT</td>\n",
       "      <td>554911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>WA</td>\n",
       "      <td>113336002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>WI</td>\n",
       "      <td>24141576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>WV</td>\n",
       "      <td>714780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2317228</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   org_state  agency_ic_fundings\n",
       "0         AL             9095775\n",
       "1         AR             2305873\n",
       "2         AZ            19291308\n",
       "3         BC             1313155\n",
       "4         CA           276507528\n",
       "5         CO             9434500\n",
       "6         CT            46490803\n",
       "7         DC            17706480\n",
       "8         DE              823703\n",
       "9         FL            47141005\n",
       "10        GA            23413466\n",
       "11        HI             6867343\n",
       "12        IA             7413559\n",
       "13        IL           148242112\n",
       "14        IN            19214283\n",
       "15        KS             2128229\n",
       "16        KY            12310190\n",
       "17        LA             6351656\n",
       "18        MA           385381611\n",
       "19        MD            38716389\n",
       "20        ME             4043314\n",
       "21        MI            58369519\n",
       "22        MN            47962857\n",
       "23        MO            29368698\n",
       "24        NC            51819254\n",
       "25        NE             9716876\n",
       "26        NH             8465229\n",
       "27        NJ            17177724\n",
       "28        NM             1943628\n",
       "29        NY           267589842\n",
       "30        OH            70510115\n",
       "31        OK             7928144\n",
       "32        ON             2448767\n",
       "33        OR            22473875\n",
       "34        PA           121206288\n",
       "35        PQ             1409518\n",
       "36        PR              627180\n",
       "37        RI             3078495\n",
       "38        SC            10582120\n",
       "39        TN            69343936\n",
       "40        TX           185164698\n",
       "41        UT            22617397\n",
       "42        VA            23311906\n",
       "43        VT              554911\n",
       "44        WA           113336002\n",
       "45        WI            24141576\n",
       "46        WV              714780\n",
       "47       NaN             2317228"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grouping data to calculate the sum of agency_ic_funding for each org_state\n",
    "state_funding = df.groupby('org_state', dropna=False)['agency_ic_fundings'].sum().reset_index()\n",
    "state_funding"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Chloropleth (US Map with funding by state)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Grouping data to calculate the sum of agency_ic_funding for each org_state\n",
    "# state_funding = df.groupby('org_state',dropna=False)['agency_ic_fundings'].sum().reset_index()\n",
    "\n",
    "# # Creating the Chloropleth\n",
    "# fig = px.choropleth(state_funding, \n",
    "#                     locations='org_state', \n",
    "#                     locationmode='USA-states', \n",
    "#                     color='agency_ic_fundings',\n",
    "#                     scope='usa',\n",
    "#                     color_continuous_scale='Blues',\n",
    "#                     title='Total NCI Funding by State since 2000',\n",
    "#                     height=600,\n",
    "#                     width=1200)\n",
    "\n",
    "# # Show the Chloropleth\n",
    "# fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program</th>\n",
       "      <th>fiscal_year</th>\n",
       "      <th>agency_ic_fundings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>2020</td>\n",
       "      <td>1898254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>2021</td>\n",
       "      <td>11554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>2022</td>\n",
       "      <td>197648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AcquiredResistancetoTherapyNetworkARTNet</td>\n",
       "      <td>2022</td>\n",
       "      <td>7357128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AllofUs</td>\n",
       "      <td>2018</td>\n",
       "      <td>186724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>2019</td>\n",
       "      <td>2134771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>2020</td>\n",
       "      <td>2207890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>2021</td>\n",
       "      <td>2003668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>2022</td>\n",
       "      <td>2046999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>2023</td>\n",
       "      <td>2069846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>136 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               program  fiscal_year  \\\n",
       "0    ADMIRALStudyAdmixtureanalysisofacutelymphoblas...         2020   \n",
       "1    ADMIRALStudyAdmixtureanalysisofacutelymphoblas...         2021   \n",
       "2    ADMIRALStudyAdmixtureanalysisofacutelymphoblas...         2022   \n",
       "3             AcquiredResistancetoTherapyNetworkARTNet         2022   \n",
       "4                                              AllofUs         2018   \n",
       "..                                                 ...          ...   \n",
       "131  YaleSPOREinLungCancerYSILCTheBiologyandPersona...         2019   \n",
       "132  YaleSPOREinLungCancerYSILCTheBiologyandPersona...         2020   \n",
       "133  YaleSPOREinLungCancerYSILCTheBiologyandPersona...         2021   \n",
       "134  YaleSPOREinLungCancerYSILCTheBiologyandPersona...         2022   \n",
       "135  YaleSPOREinLungCancerYSILCTheBiologyandPersona...         2023   \n",
       "\n",
       "     agency_ic_fundings  \n",
       "0               1898254  \n",
       "1                 11554  \n",
       "2                197648  \n",
       "3               7357128  \n",
       "4                186724  \n",
       "..                  ...  \n",
       "131             2134771  \n",
       "132             2207890  \n",
       "133             2003668  \n",
       "134             2046999  \n",
       "135             2069846  \n",
       "\n",
       "[136 rows x 3 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "program_fy_funding = df.groupby(['program','fiscal_year'])['agency_ic_fundings'].sum().reset_index()\n",
    "program_fy_funding"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Stacked Bar Chart (Program funding per year)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Group data\n",
    "# program_fy_funding = df.groupby(['program','fiscal_year'])['agency_ic_fundings'].sum().reset_index()\n",
    "\n",
    "# # Create the line chart\n",
    "# fig = px.bar(program_fy_funding, \n",
    "#               x='fiscal_year', \n",
    "#               y='agency_ic_fundings', \n",
    "#               color='program', \n",
    "#               title='NCI Funding Since Fiscal Year 2000 by Key Program',\n",
    "#               labels={'fiscal_year': 'Fiscal Year', 'agency_ic_fundings': 'NCI Funding', 'program':'Key Program'},\n",
    "#               height=600,\n",
    "#               width=1800\n",
    "#             )\n",
    "# # Show the line chart\n",
    "# fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sankey Diagram\n",
    "There was a lot of trial and error not fully shown here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program</th>\n",
       "      <th>full_foa</th>\n",
       "      <th>core_project_num</th>\n",
       "      <th>agency_ic_fundings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>PA-19-056</td>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>1898254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>PA-20-272</td>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>197648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>PA-21-071</td>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>11554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AcquiredResistancetoTherapyNetworkARTNet</td>\n",
       "      <td>RFA-CA-21-052</td>\n",
       "      <td>U54CA224019</td>\n",
       "      <td>1308998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AcquiredResistancetoTherapyNetworkARTNet</td>\n",
       "      <td>RFA-CA-21-052</td>\n",
       "      <td>U54CA224081</td>\n",
       "      <td>1119180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1997</th>\n",
       "      <td>TheUniversityofTexasMDAndersonCancerCenterSPOR...</td>\n",
       "      <td>PAR-18-313</td>\n",
       "      <td>P50CA217674</td>\n",
       "      <td>9063325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998</th>\n",
       "      <td>VanderbiltIngramCancerCenterSPOREinGastrointes...</td>\n",
       "      <td>PAR-18-313</td>\n",
       "      <td>P50CA236733</td>\n",
       "      <td>11590546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>PAR-14-031</td>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>12004456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2000</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>PAR-18-313</td>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>8328403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2001</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>RFA-CA-16-012</td>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>207600</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2002 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                program       full_foa  \\\n",
       "0     ADMIRALStudyAdmixtureanalysisofacutelymphoblas...      PA-19-056   \n",
       "1     ADMIRALStudyAdmixtureanalysisofacutelymphoblas...      PA-20-272   \n",
       "2     ADMIRALStudyAdmixtureanalysisofacutelymphoblas...      PA-21-071   \n",
       "3              AcquiredResistancetoTherapyNetworkARTNet  RFA-CA-21-052   \n",
       "4              AcquiredResistancetoTherapyNetworkARTNet  RFA-CA-21-052   \n",
       "...                                                 ...            ...   \n",
       "1997  TheUniversityofTexasMDAndersonCancerCenterSPOR...     PAR-18-313   \n",
       "1998  VanderbiltIngramCancerCenterSPOREinGastrointes...     PAR-18-313   \n",
       "1999  YaleSPOREinLungCancerYSILCTheBiologyandPersona...     PAR-14-031   \n",
       "2000  YaleSPOREinLungCancerYSILCTheBiologyandPersona...     PAR-18-313   \n",
       "2001  YaleSPOREinLungCancerYSILCTheBiologyandPersona...  RFA-CA-16-012   \n",
       "\n",
       "     core_project_num  agency_ic_fundings  \n",
       "0         R01CA239701             1898254  \n",
       "1         R01CA239701              197648  \n",
       "2         R01CA239701               11554  \n",
       "3         U54CA224019             1308998  \n",
       "4         U54CA224081             1119180  \n",
       "...               ...                 ...  \n",
       "1997      P50CA217674             9063325  \n",
       "1998      P50CA236733            11590546  \n",
       "1999      P50CA196530            12004456  \n",
       "2000      P50CA196530             8328403  \n",
       "2001      P50CA196530              207600  \n",
       "\n",
       "[2002 rows x 4 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sum NCI funding for each program, NOFO, and project combo\n",
    "sankey_summary = df.groupby(['program', 'full_foa', 'core_project_num'])['agency_ic_fundings'].sum().reset_index()\n",
    "sankey_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>PA-19-056</td>\n",
       "      <td>1898254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>PA-20-272</td>\n",
       "      <td>197648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>PA-21-071</td>\n",
       "      <td>11554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AcquiredResistancetoTherapyNetworkARTNet</td>\n",
       "      <td>RFA-CA-21-052</td>\n",
       "      <td>6306724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AcquiredResistancetoTherapyNetworkARTNet</td>\n",
       "      <td>RFA-CA-21-053</td>\n",
       "      <td>1050404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1987</th>\n",
       "      <td>RFA-CA-22-038</td>\n",
       "      <td>U24CA224285</td>\n",
       "      <td>1924560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1988</th>\n",
       "      <td>RFA-CA-22-038</td>\n",
       "      <td>U24CA224309</td>\n",
       "      <td>1836054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1989</th>\n",
       "      <td>RFA-CA-22-038</td>\n",
       "      <td>U24CA224319</td>\n",
       "      <td>1983570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1990</th>\n",
       "      <td>RFA-CA-22-038</td>\n",
       "      <td>U24CA224331</td>\n",
       "      <td>1973873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999</th>\n",
       "      <td>RFA-RM-21-016</td>\n",
       "      <td>DP2CA281605</td>\n",
       "      <td>188769</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1982 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 source         target  \\\n",
       "0     ADMIRALStudyAdmixtureanalysisofacutelymphoblas...      PA-19-056   \n",
       "1     ADMIRALStudyAdmixtureanalysisofacutelymphoblas...      PA-20-272   \n",
       "2     ADMIRALStudyAdmixtureanalysisofacutelymphoblas...      PA-21-071   \n",
       "3              AcquiredResistancetoTherapyNetworkARTNet  RFA-CA-21-052   \n",
       "4              AcquiredResistancetoTherapyNetworkARTNet  RFA-CA-21-053   \n",
       "...                                                 ...            ...   \n",
       "1987                                      RFA-CA-22-038    U24CA224285   \n",
       "1988                                      RFA-CA-22-038    U24CA224309   \n",
       "1989                                      RFA-CA-22-038    U24CA224319   \n",
       "1990                                      RFA-CA-22-038    U24CA224331   \n",
       "1999                                      RFA-RM-21-016    DP2CA281605   \n",
       "\n",
       "        value  \n",
       "0     1898254  \n",
       "1      197648  \n",
       "2       11554  \n",
       "3     6306724  \n",
       "4     1050404  \n",
       "...       ...  \n",
       "1987  1924560  \n",
       "1988  1836054  \n",
       "1989  1983570  \n",
       "1990  1973873  \n",
       "1999   188769  \n",
       "\n",
       "[1982 rows x 3 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Build connections from programs to full_foas\n",
    "links_program_nofo = df.groupby(['program','full_foa'])['agency_ic_fundings'].sum().reset_index()\n",
    "# Rename with standard cols\n",
    "links_program_nofo.columns = ['source','target','value']\n",
    "\n",
    "\n",
    "# Build connections from full_foas to project_nums\n",
    "links_nofo_project = df.groupby(['full_foa','core_project_num'])['agency_ic_fundings'].sum().reset_index()\n",
    "# Rename with standard cols\n",
    "links_nofo_project.columns = ['source','target','value']\n",
    "\n",
    "# Combine different links dataframes\n",
    "links = pd.concat([links_program_nofo, links_nofo_project])\n",
    "\n",
    "# Remove links with no funding\n",
    "links = links[links['value'] > 0]\n",
    "links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>1898254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>197648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>11554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>6306724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>1050404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1987</th>\n",
       "      <td>60</td>\n",
       "      <td>1916</td>\n",
       "      <td>1924560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1988</th>\n",
       "      <td>60</td>\n",
       "      <td>1917</td>\n",
       "      <td>1836054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1989</th>\n",
       "      <td>60</td>\n",
       "      <td>1918</td>\n",
       "      <td>1983570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1990</th>\n",
       "      <td>60</td>\n",
       "      <td>1919</td>\n",
       "      <td>1973873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999</th>\n",
       "      <td>61</td>\n",
       "      <td>1945</td>\n",
       "      <td>188769</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1982 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      source  target    value\n",
       "0          0      23  1898254\n",
       "1          0      25   197648\n",
       "2          0      26    11554\n",
       "3          1      58  6306724\n",
       "4          1      59  1050404\n",
       "...      ...     ...      ...\n",
       "1987      60    1916  1924560\n",
       "1988      60    1917  1836054\n",
       "1989      60    1918  1983570\n",
       "1990      60    1919  1973873\n",
       "1999      61    1945   188769\n",
       "\n",
       "[1982 rows x 3 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create nodes dataframe from links dataframe\n",
    "nodes_data = pd.concat([links['source'], links['target']]).unique()\n",
    "nodes = pd.DataFrame({\n",
    "    'node': nodes_data,\n",
    "    'node_id': range(len(nodes_data))\n",
    "})\n",
    "\n",
    "# Create a dictionary to map node names to node IDs\n",
    "node_id_mapping = dict(zip(nodes['node'], nodes['node_id']))\n",
    "\n",
    "# Map the node names to their corresponding node IDs in the links dataframe\n",
    "links['source'] = links['source'].map(node_id_mapping)\n",
    "links['target'] = links['target'].map(node_id_mapping)\n",
    "\n",
    "links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Create Sankey diagram\n",
    "# fig = go.Figure(go.Sankey(\n",
    "#     node=dict(\n",
    "#         pad=15,\n",
    "#         thickness=20,\n",
    "#         line=dict(color=\"black\", width=0.5),\n",
    "#         label=nodes['node'],  # Use the node names as labels\n",
    "\n",
    "#     ),\n",
    "#     link=dict(\n",
    "#         source=links['source'],\n",
    "#         target=links['target'],\n",
    "#         value=links['value']\n",
    "#     )\n",
    "# ))\n",
    "\n",
    "# # Customize layout\n",
    "# fig.update_layout(\n",
    "#     title_text=\"NCI Key Program Funding Flow\",\n",
    "#     font_size=14,\n",
    "#     height=1800,\n",
    "#     width=1200\n",
    "# )\n",
    "\n",
    "# # Display the figure\n",
    "# fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Program Keywords (from Grants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pref_terms</th>\n",
       "      <th>keyword_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5006</th>\n",
       "      <td>Malignant Neoplasms</td>\n",
       "      <td>2813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6241</th>\n",
       "      <td>Patients</td>\n",
       "      <td>2722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2301</th>\n",
       "      <td>Data</td>\n",
       "      <td>2504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11625</th>\n",
       "      <td>novel</td>\n",
       "      <td>2483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8129</th>\n",
       "      <td>Testing</td>\n",
       "      <td>2482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13064</th>\n",
       "      <td>tumor</td>\n",
       "      <td>2437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590</th>\n",
       "      <td>Goals</td>\n",
       "      <td>2186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1510</th>\n",
       "      <td>Cells</td>\n",
       "      <td>2158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10764</th>\n",
       "      <td>improved</td>\n",
       "      <td>2075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2413</th>\n",
       "      <td>Development</td>\n",
       "      <td>2068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1772</th>\n",
       "      <td>Clinical</td>\n",
       "      <td>2052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7299</th>\n",
       "      <td>Role</td>\n",
       "      <td>2005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3990</th>\n",
       "      <td>Human</td>\n",
       "      <td>1860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5372</th>\n",
       "      <td>Modeling</td>\n",
       "      <td>1817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2525</th>\n",
       "      <td>Disease</td>\n",
       "      <td>1766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12394</th>\n",
       "      <td>response</td>\n",
       "      <td>1721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8155</th>\n",
       "      <td>Therapeutic</td>\n",
       "      <td>1687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5383</th>\n",
       "      <td>Molecular</td>\n",
       "      <td>1671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7181</th>\n",
       "      <td>Research</td>\n",
       "      <td>1605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10776</th>\n",
       "      <td>in vivo</td>\n",
       "      <td>1561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8759</th>\n",
       "      <td>Work</td>\n",
       "      <td>1537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9092</th>\n",
       "      <td>base</td>\n",
       "      <td>1510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6213</th>\n",
       "      <td>Pathway interactions</td>\n",
       "      <td>1510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5118</th>\n",
       "      <td>Mediating</td>\n",
       "      <td>1470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5484</th>\n",
       "      <td>Mus</td>\n",
       "      <td>1445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7573</th>\n",
       "      <td>Signal Transduction</td>\n",
       "      <td>1385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>Address</td>\n",
       "      <td>1373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11390</th>\n",
       "      <td>mouse model</td>\n",
       "      <td>1315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6016</th>\n",
       "      <td>Outcome</td>\n",
       "      <td>1241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3444</th>\n",
       "      <td>Genes</td>\n",
       "      <td>1238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6798</th>\n",
       "      <td>Proteins</td>\n",
       "      <td>1200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7209</th>\n",
       "      <td>Resistance</td>\n",
       "      <td>1143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10843</th>\n",
       "      <td>innovation</td>\n",
       "      <td>1138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8218</th>\n",
       "      <td>Time</td>\n",
       "      <td>1124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4250</th>\n",
       "      <td>In Vitro</td>\n",
       "      <td>1104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5502</th>\n",
       "      <td>Mutation</td>\n",
       "      <td>1091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9278</th>\n",
       "      <td>cancer cell</td>\n",
       "      <td>1070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10848</th>\n",
       "      <td>insight</td>\n",
       "      <td>1061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1805</th>\n",
       "      <td>Clinical Trials</td>\n",
       "      <td>1030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3445</th>\n",
       "      <td>Genetic</td>\n",
       "      <td>1013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7416</th>\n",
       "      <td>Sampling</td>\n",
       "      <td>997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5264</th>\n",
       "      <td>Methods</td>\n",
       "      <td>974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11529</th>\n",
       "      <td>neoplastic cell</td>\n",
       "      <td>967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8248</th>\n",
       "      <td>Tissues</td>\n",
       "      <td>951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6667</th>\n",
       "      <td>Process</td>\n",
       "      <td>947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12133</th>\n",
       "      <td>programs</td>\n",
       "      <td>936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9861</th>\n",
       "      <td>design</td>\n",
       "      <td>927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4604</th>\n",
       "      <td>Knowledge</td>\n",
       "      <td>920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10830</th>\n",
       "      <td>inhibitor</td>\n",
       "      <td>912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9320</th>\n",
       "      <td>cancer therapy</td>\n",
       "      <td>902</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 pref_terms  keyword_count\n",
       "5006    Malignant Neoplasms           2813\n",
       "6241               Patients           2722\n",
       "2301                   Data           2504\n",
       "11625                 novel           2483\n",
       "8129                Testing           2482\n",
       "13064                 tumor           2437\n",
       "3590                  Goals           2186\n",
       "1510                  Cells           2158\n",
       "10764              improved           2075\n",
       "2413            Development           2068\n",
       "1772               Clinical           2052\n",
       "7299                   Role           2005\n",
       "3990                  Human           1860\n",
       "5372               Modeling           1817\n",
       "2525                Disease           1766\n",
       "12394              response           1721\n",
       "8155            Therapeutic           1687\n",
       "5383              Molecular           1671\n",
       "7181               Research           1605\n",
       "10776               in vivo           1561\n",
       "8759                   Work           1537\n",
       "9092                   base           1510\n",
       "6213   Pathway interactions           1510\n",
       "5118              Mediating           1470\n",
       "5484                    Mus           1445\n",
       "7573    Signal Transduction           1385\n",
       "283                 Address           1373\n",
       "11390           mouse model           1315\n",
       "6016                Outcome           1241\n",
       "3444                  Genes           1238\n",
       "6798               Proteins           1200\n",
       "7209             Resistance           1143\n",
       "10843            innovation           1138\n",
       "8218                   Time           1124\n",
       "4250               In Vitro           1104\n",
       "5502               Mutation           1091\n",
       "9278            cancer cell           1070\n",
       "10848               insight           1061\n",
       "1805        Clinical Trials           1030\n",
       "3445                Genetic           1013\n",
       "7416               Sampling            997\n",
       "5264                Methods            974\n",
       "11529       neoplastic cell            967\n",
       "8248                Tissues            951\n",
       "6667                Process            947\n",
       "12133              programs            936\n",
       "9861                 design            927\n",
       "4604              Knowledge            920\n",
       "10830             inhibitor            912\n",
       "9320         cancer therapy            902"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get top keywords across all grants gathered\n",
    "df['pref_terms'].str.split(';').explode().reset_index().groupby('pref_terms').size().reset_index().sort_values(by=0, ascending=False).rename(columns={0:'keyword_count'}).head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program</th>\n",
       "      <th>pref_terms</th>\n",
       "      <th>keyword_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>12 year old</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>Accounting</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>Acute Lymphocytic Leukemia</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>Administrative Supplement</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADMIRALStudyAdmixtureanalysisofacutelymphoblas...</td>\n",
       "      <td>Admixture</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20092</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>tumor</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20093</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>tumor initiation</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20094</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>tumor microenvironment</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20095</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>tumor progression</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20096</th>\n",
       "      <td>YaleSPOREinLungCancerYSILCTheBiologyandPersona...</td>\n",
       "      <td>tumorigenesis</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20097 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 program  \\\n",
       "0      ADMIRALStudyAdmixtureanalysisofacutelymphoblas...   \n",
       "1      ADMIRALStudyAdmixtureanalysisofacutelymphoblas...   \n",
       "2      ADMIRALStudyAdmixtureanalysisofacutelymphoblas...   \n",
       "3      ADMIRALStudyAdmixtureanalysisofacutelymphoblas...   \n",
       "4      ADMIRALStudyAdmixtureanalysisofacutelymphoblas...   \n",
       "...                                                  ...   \n",
       "20092  YaleSPOREinLungCancerYSILCTheBiologyandPersona...   \n",
       "20093  YaleSPOREinLungCancerYSILCTheBiologyandPersona...   \n",
       "20094  YaleSPOREinLungCancerYSILCTheBiologyandPersona...   \n",
       "20095  YaleSPOREinLungCancerYSILCTheBiologyandPersona...   \n",
       "20096  YaleSPOREinLungCancerYSILCTheBiologyandPersona...   \n",
       "\n",
       "                       pref_terms  keyword_count  \n",
       "0                     12 year old              1  \n",
       "1                      Accounting              1  \n",
       "2      Acute Lymphocytic Leukemia              4  \n",
       "3       Administrative Supplement              2  \n",
       "4                       Admixture              4  \n",
       "...                           ...            ...  \n",
       "20092                       tumor             11  \n",
       "20093            tumor initiation              1  \n",
       "20094      tumor microenvironment              4  \n",
       "20095           tumor progression             11  \n",
       "20096               tumorigenesis              1  \n",
       "\n",
       "[20097 rows x 3 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Split keywords and explode\n",
    "keywords_df = df.assign(pref_terms=df['pref_terms'].str.split(';')).explode('pref_terms')\n",
    "\n",
    "# Count keyword frequencies per program\n",
    "keyword_counts = keywords_df.groupby(['program', 'pref_terms']).size().reset_index(name='keyword_count')\n",
    "\n",
    "keyword_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Keywords seem valuable but it's difficult to find an appropriate visualization. I'd rather not rely on WordCloud. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ins_data_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
