{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DbGaP FHIR API Exploration\n",
    "2023-08-16 ZD  \n",
    "\n",
    "This notebook will explore options to gather data from the [Database of Genotypes and Phenotypes (DbGaP)](https://www.ncbi.nlm.nih.gov/gap/) using the [dbGaP FHIR API](https://github.com/ncbi/DbGaP-FHIR-API-Docs/tree/production). \n",
    "\n",
    "Much of the code used in this notebook is reused from the [DbGaP-FHIR-API_Docs repo](https://github.com/ncbi/DbGaP-FHIR-API-Docs/tree/production)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Copy `dbgap_fhir.py` from [DbGaP-FHIR-API_Docs repo](https://github.com/ncbi/DbGaP-FHIR-API-Docs/tree/production)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import json\n",
    "import requests\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "from datetime import datetime\n",
    "import time\n",
    "import pprint\n",
    "\n",
    "class DbGapFHIR:\n",
    "\n",
    "    def __init__(self, fhir_server, verify_ssl = True, api_key=None, passport=None, debug=False, show_stats=True):\n",
    "        \n",
    "        # Optional: Turn off SSL verification. Useful when dealing with a corporate proxy with self-signed certificates.\n",
    "        # This should be set to True unless you actually see certificate errors.\n",
    "\n",
    "        if not verify_ssl:\n",
    "            requests.packages.urllib3.disable_warnings()\n",
    "            \n",
    "        self.fhir_server = fhir_server\n",
    "        self.api_key = api_key\n",
    "        self.passport = passport\n",
    "        self.debug = debug\n",
    "        self.show_stats = show_stats\n",
    "\n",
    "        # We make a requests.Session to ensure consistent headers/cookie across all the requests we make\n",
    "        self.s = requests.Session()\n",
    "        self.s.headers.update({'Accept': 'application/fhir+json'})\n",
    "        # handle security needed for dbGaP\n",
    "        self.__add_passport()\n",
    "        self.s.verify = verify_ssl        \n",
    "        self.bytes_retrieved = 0\n",
    "        \n",
    "        # Test out the client by querying the server metadata#\n",
    "        r = self.s.get(f\"{self.fhir_server}/metadata\")\n",
    "\n",
    "        if \"<!DOCTYPE html>\" in r.text:\n",
    "            sys.stderr.write('ERROR: Could not get the server capability statement. ')\n",
    "\n",
    "    # Resolves all pages for the bundle. Returns an array with all Bundles, including the original Bundle.\n",
    "    def resolve_pages(self, bundle, debug=False, sleep=None):\n",
    "\n",
    "        max_tries = 10 # maximum number of tries to get next page\n",
    "        retry_sleep = 10 # after multiple failures, wait this number of seconds for a retry\n",
    "        try:\n",
    "            next_page_link = next(filter(lambda link: link['relation'] == 'next', bundle['link']), None)\n",
    "        except KeyError:\n",
    "            print('Key error link/next_page')\n",
    "            print(json.dumps(bundle, indent=3))\n",
    "            raise\n",
    "        n = 1\n",
    "        if next_page_link:\n",
    "            if sleep != None:\n",
    "                time.sleep(sleep)\n",
    "            fhir_query = next_page_link['url']\n",
    "            if self.api_key != None:\n",
    "                fhir_query += f\"&api_key={self.api_key}\"\n",
    "            if debug:\n",
    "                print('_'*80)\n",
    "                print(fhir_query)\n",
    "            tries = 1\n",
    "            r = self.s.get(fhir_query)\n",
    "            while r.status_code == 500 and tries < max_tries:\n",
    "            \ttries += 1\n",
    "            \tif tries > 6:\n",
    "            \t\ttime.sleep(retry_sleep)\n",
    "            \t\tprint (f\"trying again - waiting {retry_sleep}s\")\n",
    "            \telse:\n",
    "            \t\tprint (\"trying again\")\n",
    "            \tr = self.s.get(fhir_query)\n",
    "            if tries > 1:\n",
    "            \tprint(f'took {tries} tries')\n",
    "            next_page = r.json()\n",
    "            self.bytes_retrieved += len(r.content)\n",
    "            if 'link' not in next_page:\n",
    "                print(json.dumps(next_page, indent=3))\n",
    "            nl = [l for l in next_page['link'] if l['relation'] == 'next']\n",
    "            if debug:\n",
    "                if len(nl) < 1:\n",
    "                    print('Full last response')\n",
    "                    print(json.dumps(next_page, indent=3))\n",
    "            return [bundle] + self.resolve_pages(next_page, debug, sleep)\n",
    "        else:\n",
    "            return [bundle]\n",
    "\n",
    "    # NOTE: No cell output.\n",
    "\n",
    "    # Run a query, and get the whole set of results back as a list of resources\n",
    "    # Set limit to True if you want  to the first page if you like\n",
    "    def run_query(self, query, limit=None, debug=False, sleep=None, show_stats=None):\n",
    "    \n",
    "        if show_stats == None:\n",
    "            show_stats = self.show_stats\n",
    "            \n",
    "        t_start = time.perf_counter()\n",
    "\n",
    "        self.bytes_retrieved = 0\n",
    "        subset = False\n",
    "        \n",
    "        fhir_query = f\"{self.fhir_server}/{query}\"\n",
    "        if self.api_key != None:\n",
    "            fhir_query += f\"&api_key={self.api_key}\"\n",
    "        if debug:\n",
    "            print(fhir_query)\n",
    "        r = self.s.get(fhir_query)\n",
    "        first_bundle = r.json()\n",
    "        self.bytes_retrieved += len(r.content)\n",
    "        if debug:\n",
    "            print(json.dumps(first_bundle, indent=3))\n",
    "            print('got response')\n",
    "        # if it's just a summary\n",
    "        if 'meta' in first_bundle and 'tag' in first_bundle['meta'] and first_bundle['meta']['tag'][0]['code'] == 'SUBSETTED':\n",
    "                subset = True\n",
    "                all_bundles = [first_bundle]\n",
    "        elif limit == None:\n",
    "            all_bundles = self.resolve_pages(first_bundle, debug, sleep)\n",
    "        else:\n",
    "            all_bundles = [first_bundle]\n",
    "\n",
    "        t_end = time.perf_counter()\n",
    "        pagecount = len(all_bundles)\n",
    "        \n",
    "        resources = []\n",
    "        if subset:\n",
    "            resources = [first_bundle]\n",
    "        else:\n",
    "            resources = []\n",
    "            for bundle in all_bundles:\n",
    "                if 'entry' in bundle:\n",
    "                    resources.extend([entry['resource']  for entry in bundle['entry']])\n",
    "        \n",
    "        elapsed = t_end - t_start\n",
    "        if show_stats:\n",
    "            print(f\"Total  Resources: {len(resources)}\")\n",
    "            print(f\"Total  Bytes: {self.bytes_retrieved}\")\n",
    "            print(f\"Total  Pages: {pagecount}\")\n",
    "            print(f\"Time elapsed {elapsed:0.4f} seconds\")\n",
    "        return resources\n",
    "        \n",
    "        \n",
    "    def __add_passport(self, passport=None):\n",
    "        '''Adds Passport/TST to session header\n",
    "        '''\n",
    "        if passport == None:\n",
    "            passport = self.passport\n",
    "        \n",
    "        if passport != None:\n",
    "            full_key_path = os.path.expanduser(passport)\n",
    "            file_content = \"\"\n",
    "            if self.debug: print(f\"passport path {full_key_path}\")\n",
    "            try:\n",
    "                with open(full_key_path) as f:\n",
    "                    file_content = f.read()\n",
    "                if self.debug: print(f\"content of passport file {file_content}\")\n",
    "                self.s.headers.update({'Authorization': f'Bearer {file_content}'})\n",
    "            except:\n",
    "                print(\"Could not find passport file\")\n",
    "\n",
    "def obs_to_df(observations):\n",
    "    patient_observations_dict = {}\n",
    "    \n",
    "    for obs in observations:\n",
    "        subject_id = obs['subject']['reference']\n",
    "        obs_display_name = obs['code']['coding'][0]['display']\n",
    "        if obs_display_name in ['SUBJECT_ID','SAMPLE_ID']:\n",
    "            obs_code = obs['code']['coding'][0]['code']\n",
    "            obs_display_name = f'{obs_display_name}_{obs_code}'\n",
    "        if 'valueQuantity' in obs:\n",
    "            value_text = obs['valueQuantity']['value']\n",
    "            #value_unit = obs['valueQuantity']['unit']\n",
    "        elif 'valueCodeableConcept' in obs:\n",
    "             value_text = obs['valueCodeableConcept']['coding'][0]['display']\n",
    "        else:\n",
    "            value_text = 'unknown'\n",
    "\n",
    "        if subject_id not in patient_observations_dict:\n",
    "            patient_observations_dict[subject_id] = {obs_display_name: value_text}\n",
    "        else:\n",
    "            patient_observations_dict[subject_id][obs_display_name] = value_text\n",
    "\n",
    "    df = pd.DataFrame.from_dict(patient_observations_dict, orient='index')\n",
    "    return df\n",
    "    \n",
    "def prettyprint(some_json):\n",
    "    print(json.dumps(some_json, indent=3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for a given resource find the extension identified by a given url\n",
    "# The assumption is that there is only one such extension within a given resource\n",
    "# For the dbGaP ResearchStudy resource that is true\n",
    "def getExtension(resource, uri):\n",
    "    exts = [d for d in resource['extension'] if d['url'] == uri]\n",
    "    if len(exts) > 0 :\n",
    "        return exts[0]\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def studies_to_df(documents, verbose=False):\n",
    "    studies = []\n",
    "    for s in documents:\n",
    "\n",
    "        if verbose:\n",
    "            print (s['id'])\n",
    "            print (s['title'])\n",
    "        # use our function to find the \"study content\" extension\n",
    "        content = getExtension(s, \"https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-Content\")\n",
    "        # use our function again to find the \"number of subjects\" extension nested within the content extension\n",
    "        subject_ext = getExtension(content, \"https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-Content-NumSubjects\")\n",
    "        #print(subject_ext)\n",
    "        # Handle the fact that not all studies may have this extension\n",
    "        if subject_ext != None and 'value' in subject_ext['valueCount']:\n",
    "            subject_count = subject_ext['valueCount']['value']\n",
    "        else:\n",
    "            subject_count = 0\n",
    "\n",
    "        # Now find the extension containing the study consents\n",
    "        consent_ext = getExtension(s, \"https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-StudyConsents\")\n",
    "        # extract the display name for each consent group and print them\n",
    "        if consent_ext != None:\n",
    "            consents = [d['valueCoding']['display'] for d in consent_ext['extension'] ]\n",
    "            if verbose:\n",
    "                print(consents)\n",
    "        else:\n",
    "            consents = []\n",
    "            \n",
    "        # focus\n",
    "        if 'focus' in s:\n",
    "            focus = s['focus'][0]['text']\n",
    "            if 'coding' in s['focus'][0]:\n",
    "                focus_code = s['focus'][0]['coding'][0]['code']\n",
    "            else:\n",
    "                focus_code = ''\n",
    "        else:\n",
    "            focus = ''\n",
    "            focus_code = ''\n",
    "        # Add the relevant details to our list of studies\n",
    "        study = {\"id\":s['id'], \"title\":s[\"title\"], \"num_subjects\":subject_count,\n",
    "                 \"focus\":focus,\"focus_mesh\":focus_code,\"consents\":consents}\n",
    "        studies.append(study)\n",
    "        if verbose:\n",
    "            print('_'*40)\n",
    "        df = pd.DataFrame(studies)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Call API\n",
    "Again reusing code from https://github.com/ncbi/DbGaP-FHIR-API-Docs/blob/production/jupyter/Notebook02_studies.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "FHIR_SERVER = 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1'\n",
    "\n",
    "# No API Key in use\n",
    "    \n",
    "mf = DbGapFHIR(FHIR_SERVER)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total  Resources: 557\n",
      "Total  Bytes: 11296929\n",
      "Total  Pages: 6\n",
      "Time elapsed 34.8254 seconds\n"
     ]
    }
   ],
   "source": [
    "nci_studies = mf.run_query(\"ResearchStudy?sponsor=NCI\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_from_docs = studies_to_df(nci_studies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>num_subjects</th>\n",
       "      <th>focus</th>\n",
       "      <th>focus_mesh</th>\n",
       "      <th>consents</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>phs001287</td>\n",
       "      <td>CPTAC 3 Study</td>\n",
       "      <td>5495</td>\n",
       "      <td>Neoplasms</td>\n",
       "      <td>D009369</td>\n",
       "      <td>[GRU]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>phs000527</td>\n",
       "      <td>National Cancer Institute (NCI) Cancer Genome ...</td>\n",
       "      <td>0</td>\n",
       "      <td>Lymphoma, Non-Hodgkin</td>\n",
       "      <td>D008228</td>\n",
       "      <td>[DS-CA-MDS, GRU, GRU-IRB]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>phs000528</td>\n",
       "      <td>National Cancer Institute (NCI) Cancer Genome ...</td>\n",
       "      <td>0</td>\n",
       "      <td>Lymphoma, AIDS-Related</td>\n",
       "      <td>D016483</td>\n",
       "      <td>[DS-CA-MDS, GRU, GRU-IRB]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>phs000531</td>\n",
       "      <td>National Cancer Institute (NCI) Cancer Genome ...</td>\n",
       "      <td>0</td>\n",
       "      <td>Medulloblastoma</td>\n",
       "      <td>D008527</td>\n",
       "      <td>[DS-CA-MDS, GRU, GRU-IRB]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>phs000530</td>\n",
       "      <td>National Cancer Institute (NCI) Cancer Genome ...</td>\n",
       "      <td>16</td>\n",
       "      <td>Lymphoma, AIDS-Related</td>\n",
       "      <td>D016483</td>\n",
       "      <td>[DS-CA-MDS, GRU, GRU-IRB]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552</th>\n",
       "      <td>phs000336</td>\n",
       "      <td>DCEG Lung Cancer Study</td>\n",
       "      <td>11517</td>\n",
       "      <td>Lung Neoplasms</td>\n",
       "      <td>D008175</td>\n",
       "      <td>[CADM, SLD]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>553</th>\n",
       "      <td>phs000124</td>\n",
       "      <td>Genome-Wide Association Study of Neuroblastoma</td>\n",
       "      <td>1662</td>\n",
       "      <td>Neuroblastoma</td>\n",
       "      <td>D009447</td>\n",
       "      <td>[GRU]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>554</th>\n",
       "      <td>phs000093</td>\n",
       "      <td>A Genome Wide Scan of Lung Cancer and Smoking</td>\n",
       "      <td>5566</td>\n",
       "      <td>Lung Neoplasms</td>\n",
       "      <td>D008175</td>\n",
       "      <td>[NRUP, CADM, SLD]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>phs000207</td>\n",
       "      <td>CGEMS Prostate Cancer GWAS - Stage 1 - PLCO</td>\n",
       "      <td>2252</td>\n",
       "      <td>Prostatic Neoplasms</td>\n",
       "      <td>D011471</td>\n",
       "      <td>[CADM]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>556</th>\n",
       "      <td>phs000187</td>\n",
       "      <td>High Density SNP Association Analysis of Melanoma</td>\n",
       "      <td>3115</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>D008545</td>\n",
       "      <td>[NRUP, GRU]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>557 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            id                                              title  \\\n",
       "0    phs001287                                      CPTAC 3 Study   \n",
       "1    phs000527  National Cancer Institute (NCI) Cancer Genome ...   \n",
       "2    phs000528  National Cancer Institute (NCI) Cancer Genome ...   \n",
       "3    phs000531  National Cancer Institute (NCI) Cancer Genome ...   \n",
       "4    phs000530  National Cancer Institute (NCI) Cancer Genome ...   \n",
       "..         ...                                                ...   \n",
       "552  phs000336                             DCEG Lung Cancer Study   \n",
       "553  phs000124     Genome-Wide Association Study of Neuroblastoma   \n",
       "554  phs000093      A Genome Wide Scan of Lung Cancer and Smoking   \n",
       "555  phs000207        CGEMS Prostate Cancer GWAS - Stage 1 - PLCO   \n",
       "556  phs000187  High Density SNP Association Analysis of Melanoma   \n",
       "\n",
       "     num_subjects                   focus focus_mesh  \\\n",
       "0            5495               Neoplasms    D009369   \n",
       "1               0   Lymphoma, Non-Hodgkin    D008228   \n",
       "2               0  Lymphoma, AIDS-Related    D016483   \n",
       "3               0         Medulloblastoma    D008527   \n",
       "4              16  Lymphoma, AIDS-Related    D016483   \n",
       "..            ...                     ...        ...   \n",
       "552         11517          Lung Neoplasms    D008175   \n",
       "553          1662           Neuroblastoma    D009447   \n",
       "554          5566          Lung Neoplasms    D008175   \n",
       "555          2252     Prostatic Neoplasms    D011471   \n",
       "556          3115                Melanoma    D008545   \n",
       "\n",
       "                      consents  \n",
       "0                        [GRU]  \n",
       "1    [DS-CA-MDS, GRU, GRU-IRB]  \n",
       "2    [DS-CA-MDS, GRU, GRU-IRB]  \n",
       "3    [DS-CA-MDS, GRU, GRU-IRB]  \n",
       "4    [DS-CA-MDS, GRU, GRU-IRB]  \n",
       "..                         ...  \n",
       "552                [CADM, SLD]  \n",
       "553                      [GRU]  \n",
       "554          [NRUP, CADM, SLD]  \n",
       "555                     [CADM]  \n",
       "556                [NRUP, GRU]  \n",
       "\n",
       "[557 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_from_docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(nci_studies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>resourceType</th>\n",
       "      <th>id</th>\n",
       "      <th>meta</th>\n",
       "      <th>extension</th>\n",
       "      <th>identifier</th>\n",
       "      <th>title</th>\n",
       "      <th>status</th>\n",
       "      <th>category</th>\n",
       "      <th>focus</th>\n",
       "      <th>condition</th>\n",
       "      <th>keyword</th>\n",
       "      <th>description</th>\n",
       "      <th>enrollment</th>\n",
       "      <th>sponsor</th>\n",
       "      <th>partOf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ResearchStudy</td>\n",
       "      <td>phs001287</td>\n",
       "      <td>{'versionId': '1', 'lastUpdated': '2022-02-14T...</td>\n",
       "      <td>[{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...</td>\n",
       "      <td>[{'type': {'coding': [{'system': 'https://dbga...</td>\n",
       "      <td>CPTAC 3 Study</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'coding': [{'system': 'https://dbgap-api.ncb...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>\\nRecently, significant progress has been made...</td>\n",
       "      <td>[{'reference': 'Group/phs001287.v13.p5-all-sub...</td>\n",
       "      <td>{'reference': 'Organization/NCI', 'type': 'Org...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ResearchStudy</td>\n",
       "      <td>phs000527</td>\n",
       "      <td>{'versionId': '1', 'lastUpdated': '2022-02-14T...</td>\n",
       "      <td>[{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...</td>\n",
       "      <td>[{'type': {'coding': [{'system': 'https://dbga...</td>\n",
       "      <td>National Cancer Institute (NCI) Cancer Genome ...</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'coding': [{'system': 'https://dbgap-api.ncb...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>\\nThe Office of Cancer Genomics at the Nationa...</td>\n",
       "      <td>[{'reference': 'Group/phs000527.v15.p5-all-sub...</td>\n",
       "      <td>{'reference': 'Organization/NCI', 'type': 'Org...</td>\n",
       "      <td>[{'reference': 'ResearchStudy/phs000235'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ResearchStudy</td>\n",
       "      <td>phs000528</td>\n",
       "      <td>{'versionId': '1', 'lastUpdated': '2022-02-14T...</td>\n",
       "      <td>[{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...</td>\n",
       "      <td>[{'type': {'coding': [{'system': 'https://dbga...</td>\n",
       "      <td>National Cancer Institute (NCI) Cancer Genome ...</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'coding': [{'system': 'https://dbgap-api.ncb...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>\\nThe Office of Cancer Genomics at the Nationa...</td>\n",
       "      <td>[{'reference': 'Group/phs000528.v15.p5-all-sub...</td>\n",
       "      <td>{'reference': 'Organization/NCI', 'type': 'Org...</td>\n",
       "      <td>[{'reference': 'ResearchStudy/phs000235'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ResearchStudy</td>\n",
       "      <td>phs000531</td>\n",
       "      <td>{'versionId': '1', 'lastUpdated': '2022-02-14T...</td>\n",
       "      <td>[{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...</td>\n",
       "      <td>[{'type': {'coding': [{'system': 'https://dbga...</td>\n",
       "      <td>National Cancer Institute (NCI) Cancer Genome ...</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'coding': [{'system': 'https://dbgap-api.ncb...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>\\nThe Office of Cancer Genomics (OCG) at the N...</td>\n",
       "      <td>[{'reference': 'Group/phs000531.v15.p5-all-sub...</td>\n",
       "      <td>{'reference': 'Organization/NCI', 'type': 'Org...</td>\n",
       "      <td>[{'reference': 'ResearchStudy/phs000235'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ResearchStudy</td>\n",
       "      <td>phs000530</td>\n",
       "      <td>{'versionId': '1', 'lastUpdated': '2022-02-14T...</td>\n",
       "      <td>[{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...</td>\n",
       "      <td>[{'type': {'coding': [{'system': 'https://dbga...</td>\n",
       "      <td>National Cancer Institute (NCI) Cancer Genome ...</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'coding': [{'system': 'https://dbgap-api.ncb...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>\\nThe Office of Cancer Genomics at the Nationa...</td>\n",
       "      <td>[{'reference': 'Group/phs000530.v15.p5-all-sub...</td>\n",
       "      <td>{'reference': 'Organization/NCI', 'type': 'Org...</td>\n",
       "      <td>[{'reference': 'ResearchStudy/phs000235'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>552</th>\n",
       "      <td>ResearchStudy</td>\n",
       "      <td>phs000336</td>\n",
       "      <td>{'versionId': '1', 'lastUpdated': '2022-02-14T...</td>\n",
       "      <td>[{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...</td>\n",
       "      <td>[{'type': {'coding': [{'system': 'https://dbga...</td>\n",
       "      <td>DCEG Lung Cancer Study</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'coding': [{'system': 'https://dbgap-api.ncb...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>\\nThree genetic loci for lung cancer risk have...</td>\n",
       "      <td>[{'reference': 'Group/phs000336.v1.p1-all-subj...</td>\n",
       "      <td>{'reference': 'Organization/NCI', 'type': 'Org...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>553</th>\n",
       "      <td>ResearchStudy</td>\n",
       "      <td>phs000124</td>\n",
       "      <td>{'versionId': '1', 'lastUpdated': '2022-02-14T...</td>\n",
       "      <td>[{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...</td>\n",
       "      <td>[{'type': {'coding': [{'system': 'https://dbga...</td>\n",
       "      <td>Genome-Wide Association Study of Neuroblastoma</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'coding': [{'system': 'https://dbgap-api.ncb...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>\\nNeuroblastoma is a malignancy of the develop...</td>\n",
       "      <td>[{'reference': 'Group/phs000124.v2.p1-all-subj...</td>\n",
       "      <td>{'reference': 'Organization/NCI', 'type': 'Org...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>554</th>\n",
       "      <td>ResearchStudy</td>\n",
       "      <td>phs000093</td>\n",
       "      <td>{'versionId': '1', 'lastUpdated': '2022-02-14T...</td>\n",
       "      <td>[{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...</td>\n",
       "      <td>[{'type': {'coding': [{'system': 'https://dbga...</td>\n",
       "      <td>A Genome Wide Scan of Lung Cancer and Smoking</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'coding': [{'system': 'https://dbgap-api.ncb...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>\\nThe majority of cases of lung cancer are the...</td>\n",
       "      <td>[{'reference': 'Group/phs000093.v2.p2-all-subj...</td>\n",
       "      <td>{'reference': 'Organization/NCI', 'type': 'Org...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>ResearchStudy</td>\n",
       "      <td>phs000207</td>\n",
       "      <td>{'versionId': '1', 'lastUpdated': '2022-02-14T...</td>\n",
       "      <td>[{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...</td>\n",
       "      <td>[{'type': {'coding': [{'system': 'https://dbga...</td>\n",
       "      <td>CGEMS Prostate Cancer GWAS - Stage 1 - PLCO</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'coding': [{'system': 'https://dbgap-api.ncb...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>\\nThe Cancer Genetic Markers of Susceptibility...</td>\n",
       "      <td>[{'reference': 'Group/phs000207.v1.p1-all-subj...</td>\n",
       "      <td>{'reference': 'Organization/NCI', 'type': 'Org...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>556</th>\n",
       "      <td>ResearchStudy</td>\n",
       "      <td>phs000187</td>\n",
       "      <td>{'versionId': '1', 'lastUpdated': '2022-02-14T...</td>\n",
       "      <td>[{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...</td>\n",
       "      <td>[{'type': {'coding': [{'system': 'https://dbga...</td>\n",
       "      <td>High Density SNP Association Analysis of Melanoma</td>\n",
       "      <td>completed</td>\n",
       "      <td>[{'coding': [{'system': 'https://dbgap-api.ncb...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>[{'coding': [{'system': 'urn:oid:2.16.840.1.11...</td>\n",
       "      <td>\\nThis research builds upon an extensive resou...</td>\n",
       "      <td>[{'reference': 'Group/phs000187.v1.p1-all-subj...</td>\n",
       "      <td>{'reference': 'Organization/NCI', 'type': 'Org...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>557 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      resourceType         id  \\\n",
       "0    ResearchStudy  phs001287   \n",
       "1    ResearchStudy  phs000527   \n",
       "2    ResearchStudy  phs000528   \n",
       "3    ResearchStudy  phs000531   \n",
       "4    ResearchStudy  phs000530   \n",
       "..             ...        ...   \n",
       "552  ResearchStudy  phs000336   \n",
       "553  ResearchStudy  phs000124   \n",
       "554  ResearchStudy  phs000093   \n",
       "555  ResearchStudy  phs000207   \n",
       "556  ResearchStudy  phs000187   \n",
       "\n",
       "                                                  meta  \\\n",
       "0    {'versionId': '1', 'lastUpdated': '2022-02-14T...   \n",
       "1    {'versionId': '1', 'lastUpdated': '2022-02-14T...   \n",
       "2    {'versionId': '1', 'lastUpdated': '2022-02-14T...   \n",
       "3    {'versionId': '1', 'lastUpdated': '2022-02-14T...   \n",
       "4    {'versionId': '1', 'lastUpdated': '2022-02-14T...   \n",
       "..                                                 ...   \n",
       "552  {'versionId': '1', 'lastUpdated': '2022-02-14T...   \n",
       "553  {'versionId': '1', 'lastUpdated': '2022-02-14T...   \n",
       "554  {'versionId': '1', 'lastUpdated': '2022-02-14T...   \n",
       "555  {'versionId': '1', 'lastUpdated': '2022-02-14T...   \n",
       "556  {'versionId': '1', 'lastUpdated': '2022-02-14T...   \n",
       "\n",
       "                                             extension  \\\n",
       "0    [{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...   \n",
       "1    [{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...   \n",
       "2    [{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...   \n",
       "3    [{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...   \n",
       "4    [{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...   \n",
       "..                                                 ...   \n",
       "552  [{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...   \n",
       "553  [{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...   \n",
       "554  [{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...   \n",
       "555  [{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...   \n",
       "556  [{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/f...   \n",
       "\n",
       "                                            identifier  \\\n",
       "0    [{'type': {'coding': [{'system': 'https://dbga...   \n",
       "1    [{'type': {'coding': [{'system': 'https://dbga...   \n",
       "2    [{'type': {'coding': [{'system': 'https://dbga...   \n",
       "3    [{'type': {'coding': [{'system': 'https://dbga...   \n",
       "4    [{'type': {'coding': [{'system': 'https://dbga...   \n",
       "..                                                 ...   \n",
       "552  [{'type': {'coding': [{'system': 'https://dbga...   \n",
       "553  [{'type': {'coding': [{'system': 'https://dbga...   \n",
       "554  [{'type': {'coding': [{'system': 'https://dbga...   \n",
       "555  [{'type': {'coding': [{'system': 'https://dbga...   \n",
       "556  [{'type': {'coding': [{'system': 'https://dbga...   \n",
       "\n",
       "                                                 title     status  \\\n",
       "0                                        CPTAC 3 Study  completed   \n",
       "1    National Cancer Institute (NCI) Cancer Genome ...  completed   \n",
       "2    National Cancer Institute (NCI) Cancer Genome ...  completed   \n",
       "3    National Cancer Institute (NCI) Cancer Genome ...  completed   \n",
       "4    National Cancer Institute (NCI) Cancer Genome ...  completed   \n",
       "..                                                 ...        ...   \n",
       "552                             DCEG Lung Cancer Study  completed   \n",
       "553     Genome-Wide Association Study of Neuroblastoma  completed   \n",
       "554      A Genome Wide Scan of Lung Cancer and Smoking  completed   \n",
       "555        CGEMS Prostate Cancer GWAS - Stage 1 - PLCO  completed   \n",
       "556  High Density SNP Association Analysis of Melanoma  completed   \n",
       "\n",
       "                                              category  \\\n",
       "0    [{'coding': [{'system': 'https://dbgap-api.ncb...   \n",
       "1    [{'coding': [{'system': 'https://dbgap-api.ncb...   \n",
       "2    [{'coding': [{'system': 'https://dbgap-api.ncb...   \n",
       "3    [{'coding': [{'system': 'https://dbgap-api.ncb...   \n",
       "4    [{'coding': [{'system': 'https://dbgap-api.ncb...   \n",
       "..                                                 ...   \n",
       "552  [{'coding': [{'system': 'https://dbgap-api.ncb...   \n",
       "553  [{'coding': [{'system': 'https://dbgap-api.ncb...   \n",
       "554  [{'coding': [{'system': 'https://dbgap-api.ncb...   \n",
       "555  [{'coding': [{'system': 'https://dbgap-api.ncb...   \n",
       "556  [{'coding': [{'system': 'https://dbgap-api.ncb...   \n",
       "\n",
       "                                                 focus  \\\n",
       "0    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "1    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "2    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "3    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "4    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "..                                                 ...   \n",
       "552  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "553  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "554  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "555  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "556  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "\n",
       "                                             condition  \\\n",
       "0    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "1    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "2    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "3    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "4    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "..                                                 ...   \n",
       "552  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "553  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "554  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "555  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "556  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "\n",
       "                                               keyword  \\\n",
       "0    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "1    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "2    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "3    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "4    [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "..                                                 ...   \n",
       "552  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "553  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "554  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "555  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "556  [{'coding': [{'system': 'urn:oid:2.16.840.1.11...   \n",
       "\n",
       "                                           description  \\\n",
       "0    \\nRecently, significant progress has been made...   \n",
       "1    \\nThe Office of Cancer Genomics at the Nationa...   \n",
       "2    \\nThe Office of Cancer Genomics at the Nationa...   \n",
       "3    \\nThe Office of Cancer Genomics (OCG) at the N...   \n",
       "4    \\nThe Office of Cancer Genomics at the Nationa...   \n",
       "..                                                 ...   \n",
       "552  \\nThree genetic loci for lung cancer risk have...   \n",
       "553  \\nNeuroblastoma is a malignancy of the develop...   \n",
       "554  \\nThe majority of cases of lung cancer are the...   \n",
       "555  \\nThe Cancer Genetic Markers of Susceptibility...   \n",
       "556  \\nThis research builds upon an extensive resou...   \n",
       "\n",
       "                                            enrollment  \\\n",
       "0    [{'reference': 'Group/phs001287.v13.p5-all-sub...   \n",
       "1    [{'reference': 'Group/phs000527.v15.p5-all-sub...   \n",
       "2    [{'reference': 'Group/phs000528.v15.p5-all-sub...   \n",
       "3    [{'reference': 'Group/phs000531.v15.p5-all-sub...   \n",
       "4    [{'reference': 'Group/phs000530.v15.p5-all-sub...   \n",
       "..                                                 ...   \n",
       "552  [{'reference': 'Group/phs000336.v1.p1-all-subj...   \n",
       "553  [{'reference': 'Group/phs000124.v2.p1-all-subj...   \n",
       "554  [{'reference': 'Group/phs000093.v2.p2-all-subj...   \n",
       "555  [{'reference': 'Group/phs000207.v1.p1-all-subj...   \n",
       "556  [{'reference': 'Group/phs000187.v1.p1-all-subj...   \n",
       "\n",
       "                                               sponsor  \\\n",
       "0    {'reference': 'Organization/NCI', 'type': 'Org...   \n",
       "1    {'reference': 'Organization/NCI', 'type': 'Org...   \n",
       "2    {'reference': 'Organization/NCI', 'type': 'Org...   \n",
       "3    {'reference': 'Organization/NCI', 'type': 'Org...   \n",
       "4    {'reference': 'Organization/NCI', 'type': 'Org...   \n",
       "..                                                 ...   \n",
       "552  {'reference': 'Organization/NCI', 'type': 'Org...   \n",
       "553  {'reference': 'Organization/NCI', 'type': 'Org...   \n",
       "554  {'reference': 'Organization/NCI', 'type': 'Org...   \n",
       "555  {'reference': 'Organization/NCI', 'type': 'Org...   \n",
       "556  {'reference': 'Organization/NCI', 'type': 'Org...   \n",
       "\n",
       "                                         partOf  \n",
       "0                                           NaN  \n",
       "1    [{'reference': 'ResearchStudy/phs000235'}]  \n",
       "2    [{'reference': 'ResearchStudy/phs000235'}]  \n",
       "3    [{'reference': 'ResearchStudy/phs000235'}]  \n",
       "4    [{'reference': 'ResearchStudy/phs000235'}]  \n",
       "..                                          ...  \n",
       "552                                         NaN  \n",
       "553                                         NaN  \n",
       "554                                         NaN  \n",
       "555                                         NaN  \n",
       "556                                         NaN  \n",
       "\n",
       "[557 rows x 15 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['completed'], dtype=object)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.status.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      {'versionId': '1', 'lastUpdated': '2022-02-14T...\n",
       "1      {'versionId': '1', 'lastUpdated': '2022-02-14T...\n",
       "2      {'versionId': '1', 'lastUpdated': '2022-02-14T...\n",
       "3      {'versionId': '1', 'lastUpdated': '2022-02-14T...\n",
       "4      {'versionId': '1', 'lastUpdated': '2022-02-14T...\n",
       "                             ...                        \n",
       "552    {'versionId': '1', 'lastUpdated': '2022-02-14T...\n",
       "553    {'versionId': '1', 'lastUpdated': '2022-02-14T...\n",
       "554    {'versionId': '1', 'lastUpdated': '2022-02-14T...\n",
       "555    {'versionId': '1', 'lastUpdated': '2022-02-14T...\n",
       "556    {'versionId': '1', 'lastUpdated': '2022-02-14T...\n",
       "Name: meta, Length: 557, dtype: object"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['meta']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2022-02-14'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['meta'][0]['lastUpdated'][:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_last_updated_date(meta):\n",
    "    return meta['lastUpdated'][:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['last_updated'] = df['meta'].apply(lambda x: extract_last_updated_date(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['2022-02-14'], dtype=object)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.last_updated.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Issue: The response results for ALL studies show a \"Last Updated Date\" of Feb 2022.\n",
    "Does this mean the date at which the entire API db was last updated? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total  Resources: 1\n",
      "Total  Bytes: 13916\n",
      "Total  Pages: 1\n",
      "Time elapsed 0.6606 seconds\n"
     ]
    }
   ],
   "source": [
    "cptac3 = mf.run_query('ResearchStudy?_id=phs001287')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'resourceType': 'ResearchStudy',\n",
       "  'id': 'phs001287',\n",
       "  'meta': {'versionId': '1',\n",
       "   'lastUpdated': '2022-02-14T02:04:51.822-05:00',\n",
       "   'source': '#qRs8LVqZYunRZQh4',\n",
       "   'security': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/CodeSystem/DbGaPConcept-SecurityStudyConsent',\n",
       "     'code': 'public',\n",
       "     'display': 'public'}]},\n",
       "  'extension': [{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-StudyOverviewUrl',\n",
       "    'valueUrl': 'https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs001287.v13.p5'},\n",
       "   {'url': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-ReleaseDate',\n",
       "    'valueDate': '2021-12-13'},\n",
       "   {'url': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-StudyConsents',\n",
       "    'extension': [{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-StudyConsents-StudyConsent',\n",
       "      'valueCoding': {'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/CodeSystem/ResearchStudy-StudyConsents-StudyConsent',\n",
       "       'code': 'phs001287.v13.p5 - 1',\n",
       "       'display': 'GRU'}}]},\n",
       "   {'url': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-Content',\n",
       "    'extension': [{'url': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-Content-NumPhenotypeDatasets',\n",
       "      'valueCount': {'value': 2,\n",
       "       'system': 'http://unitsofmeasure.org',\n",
       "       'code': '1'}},\n",
       "     {'url': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-Content-NumMolecularDatasets',\n",
       "      'valueCount': {'system': 'http://unitsofmeasure.org', 'code': '1'}},\n",
       "     {'url': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-Content-NumVariables',\n",
       "      'valueCount': {'value': 5,\n",
       "       'system': 'http://unitsofmeasure.org',\n",
       "       'code': '1'}},\n",
       "     {'url': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-Content-NumSubjects',\n",
       "      'valueCount': {'value': 5495,\n",
       "       'system': 'http://unitsofmeasure.org',\n",
       "       'code': '1'}},\n",
       "     {'url': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-Content-NumSamples',\n",
       "      'valueCount': {'value': 5495,\n",
       "       'system': 'http://unitsofmeasure.org',\n",
       "       'code': '1'}},\n",
       "     {'url': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/StructureDefinition/ResearchStudy-Content-NumSubStudies',\n",
       "      'valueCount': {'system': 'http://unitsofmeasure.org', 'code': '1'}}]}],\n",
       "  'identifier': [{'type': {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/CodeSystem/DbGaPConcept-DbGaPStudyIdentifier',\n",
       "       'code': 'dbgap_study_id',\n",
       "       'display': 'dbgap_study_id'}]},\n",
       "    'value': 'phs001287.v13.p5'}],\n",
       "  'title': 'CPTAC 3 Study',\n",
       "  'status': 'completed',\n",
       "  'category': [{'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/CodeSystem/ResearchStudy-StudyDesign',\n",
       "      'code': 'Case Set',\n",
       "      'display': 'Case Set'}],\n",
       "    'text': 'Case Set'}],\n",
       "  'focus': [{'coding': [{'system': 'urn:oid:2.16.840.1.113883.6.177',\n",
       "      'code': 'D009369',\n",
       "      'display': 'Neoplasms'}],\n",
       "    'text': 'Neoplasms'}],\n",
       "  'condition': [{'coding': [{'system': 'urn:oid:2.16.840.1.113883.6.177',\n",
       "      'code': 'D009369',\n",
       "      'display': 'Neoplasms'}],\n",
       "    'text': 'Neoplasms'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Benign Neoplasms',\n",
       "      'display': 'Benign Neoplasms'}],\n",
       "    'text': 'Benign Neoplasms'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Cancer',\n",
       "      'display': 'Cancer'}],\n",
       "    'text': 'Cancer'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Malignancies',\n",
       "      'display': 'Malignancies'}],\n",
       "    'text': 'Malignancies'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Malignancy',\n",
       "      'display': 'Malignancy'}],\n",
       "    'text': 'Malignancy'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasia',\n",
       "      'display': 'Neoplasia'}],\n",
       "    'text': 'Neoplasia'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasias',\n",
       "      'display': 'Neoplasias'}],\n",
       "    'text': 'Neoplasias'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasm',\n",
       "      'display': 'Neoplasm'}],\n",
       "    'text': 'Neoplasm'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasm, Benign',\n",
       "      'display': 'Neoplasm, Benign'}],\n",
       "    'text': 'Neoplasm, Benign'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasm, Malignant',\n",
       "      'display': 'Neoplasm, Malignant'}],\n",
       "    'text': 'Neoplasm, Malignant'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasms, Benign',\n",
       "      'display': 'Neoplasms, Benign'}],\n",
       "    'text': 'Neoplasms, Benign'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasms, Malignant',\n",
       "      'display': 'Neoplasms, Malignant'}],\n",
       "    'text': 'Neoplasms, Malignant'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Tumors',\n",
       "      'display': 'Tumors'}],\n",
       "    'text': 'Tumors'},\n",
       "   {'coding': [{'system': 'https://uts.nlm.nih.gov/metathesaurus.html',\n",
       "      'code': 'C0027651',\n",
       "      'display': 'Cancer'}],\n",
       "    'text': 'Cancer'}],\n",
       "  'keyword': [{'coding': [{'system': 'urn:oid:2.16.840.1.113883.6.177',\n",
       "      'code': 'D009369',\n",
       "      'display': 'Neoplasms'}],\n",
       "    'text': 'Neoplasms'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Benign Neoplasms',\n",
       "      'display': 'Benign Neoplasms'}],\n",
       "    'text': 'Benign Neoplasms'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Cancer',\n",
       "      'display': 'Cancer'}],\n",
       "    'text': 'Cancer'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Malignancies',\n",
       "      'display': 'Malignancies'}],\n",
       "    'text': 'Malignancies'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Malignancy',\n",
       "      'display': 'Malignancy'}],\n",
       "    'text': 'Malignancy'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasia',\n",
       "      'display': 'Neoplasia'}],\n",
       "    'text': 'Neoplasia'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasias',\n",
       "      'display': 'Neoplasias'}],\n",
       "    'text': 'Neoplasias'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasm',\n",
       "      'display': 'Neoplasm'}],\n",
       "    'text': 'Neoplasm'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasm, Benign',\n",
       "      'display': 'Neoplasm, Benign'}],\n",
       "    'text': 'Neoplasm, Benign'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasm, Malignant',\n",
       "      'display': 'Neoplasm, Malignant'}],\n",
       "    'text': 'Neoplasm, Malignant'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasms, Benign',\n",
       "      'display': 'Neoplasms, Benign'}],\n",
       "    'text': 'Neoplasms, Benign'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Neoplasms, Malignant',\n",
       "      'display': 'Neoplasms, Malignant'}],\n",
       "    'text': 'Neoplasms, Malignant'},\n",
       "   {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/NamingSystem/MeshEntryTerm',\n",
       "      'code': 'D009369 entry term: Tumors',\n",
       "      'display': 'Tumors'}],\n",
       "    'text': 'Tumors'},\n",
       "   {'coding': [{'system': 'https://uts.nlm.nih.gov/metathesaurus.html',\n",
       "      'code': 'C0027651',\n",
       "      'display': 'Cancer'}],\n",
       "    'text': 'Cancer'}],\n",
       "  'description': '\\nRecently, significant progress has been made in characterizing and sequencing the genomic alterations in statistically robust numbers of samples from several types of cancer. For example, The Cancer Genome Atlas (TCGA), International Cancer Genome Consortium (ICGC) and other similar efforts are identifying genomic alterations associated with specific cancers (e.g., copy number aberrations, rearrangements, point mutations, epigenomic changes, etc.) The availability of these multi-dimensional data to the scientific community sets the stage for the development of new molecularly targeted cancer interventions. Understanding the comprehensive functional changes in cancer proteomes arising from genomic alterations and other factors is the next logical step in the development of high-value candidate protein biomarkers. Hence, proteomics can greatly advance the understanding of molecular mechanisms of disease pathology via the analysis of changes in protein expression, their modifications and variations, as well as protein&#61;protein interaction, signaling pathways and networks responsible for cellular functions such as apoptosis and oncogenesis. Realizing this great potential, the NCI launched the third phase of the CPTC initiative in September 2016. As the Clinical Proteomic Tumor Analysis Consortium, CPTAC continues to define cancer proteomes on genomically-characterized biospecimens. The purpose of this integrative approach was to provide the broad scientific community with knowledge that links genotype to proteotype and ultimately phenotype. In this third phase of CPTAC, the program aims to expand on CPTAC II and genomically and proteomically characterize over 2000 samples from 10 cancer types (Lung Adenocarcinoma, Pancreatic Ductal Adenocarcinoma, Glioblastoma Multiforme, Acute Myeloid Leukemia, Clear cell renal Carcinoma, Head and Neck Squamous Cell Carcinoma, Cutaneous Melanoma, Sarcoma, Lung Squamous Cell Carcinoma, Uterine Corpus Endometrial Carcinoma) .Germline DNA is obtained from blood and Normal control samples for proteomics varied by organ site. All cancer samples were derived from primary and untreated tumor.\\n',\n",
       "  'enrollment': [{'reference': 'Group/phs001287.v13.p5-all-subjects'}],\n",
       "  'sponsor': {'reference': 'Organization/NCI',\n",
       "   'type': 'Organization',\n",
       "   'display': 'National Cancer Institute'}}]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cptac3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Issue: Searching the phs ID for the CPTAC study returns outdated results compared to the DbGaP site\n",
    "https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs001287.v17.p6  \n",
    "\n",
    "Most recent version in API results: \n",
    "```\n",
    "'identifier': [{'type': {'coding': [{'system': 'https://dbgap-api.ncbi.nlm.nih.gov/fhir/x1/CodeSystem/DbGaPConcept-DbGaPStudyIdentifier',\n",
    "       'code': 'dbgap_study_id',\n",
    "       'display': 'dbgap_study_id'}]},\n",
    "    'value': 'phs001287.v13.p5'}],\n",
    "  'title': 'CPTAC 3 Study',\n",
    "  ```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ins_data_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
