{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explore the iCite API for gathering Publication data\n",
    "2023-11-22 ZD  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 1: Trying the iCite API\n",
    "Part 2 (below): Using bulk download snapshot  \n",
    "\n",
    "Relevant Jira Ticket: [INS-790](https://tracker.nci.nih.gov/browse/INS-790)  \n",
    "\n",
    "Exploratory notebook to investigate gathering Publications data for INS from [the iCite API](https://icite.od.nih.gov/api). This wiil build upon the work to in `notebooks/07_gather_publications.ipynb` and `modules/gather_publication_data.py`.  \n",
    "\n",
    "The primary goal is to gather metrics specific to iCite: Citation Count and Relative Citation Ratio. The secondary goal is to explore whether the iCite API could replace the Biopython Entrez PubMed API, which is very slow. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Method to import from parent directory\n",
    "import os\n",
    "import sys\n",
    "root_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "sys.path.append(root_dir)\n",
    "\n",
    "import requests\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "# Get all existing publication functions\n",
    "import modules.gather_publication_data as gpub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R01CA263500: (1/1): {'search_id': None, 'total': 7, 'offset': 0, 'limit': 500, 'sort_field': 'core_project_nums', 'sort_order': 'desc', 'sorted_by_relevance': False, 'properties': {}}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'coreproject': 'R01CA263500', 'pmid': 37138086, 'applid': 10679077},\n",
       " {'coreproject': 'R01CA263500', 'pmid': 36288726, 'applid': 10679077},\n",
       " {'coreproject': 'R01CA263500', 'pmid': 36917953, 'applid': 10679077},\n",
       " {'coreproject': 'R01CA263500', 'pmid': 36734849, 'applid': 10679077},\n",
       " {'coreproject': 'R01CA263500', 'pmid': 37059069, 'applid': 10679077},\n",
       " {'coreproject': 'R01CA263500', 'pmid': 35130560, 'applid': 10679077},\n",
       " {'coreproject': 'R01CA263500', 'pmid': 37024595, 'applid': 10679077}]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Test imported functions\n",
    "gpub.get_pmids_from_nih_reporter_api('R01CA263500', print_meta=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'publication_id': '37138086)',\n",
       " 'title': 'Glioblastoma remodelling of human neural circuits decreases survival.',\n",
       " 'authors': 'Saritha Krishna, Abrar Choudhury, Michael B Keough, Kyounghee Seo, Lijun Ni, Sofia Kakaizada, Anthony Lee, Alexander Aabedi, Galina Popova, Benjamin Lipkin, Caroline Cao, Cesar Nava Gonzales, Rasika Sudharshan, Andrew Egladyous, Nyle Almeida, Yalan Zhang, Annette M Molinaro, Humsa S Venkatesh, Andy G S Daniel, Kiarash Shamardani, Jeanette Hyer, Edward F Chang, Anne Findlay, Joanna J Phillips, Srikantan Nagarajan, David R Raleigh, David Brang, Michelle Monje, Shawn L Hervey-Jumper',\n",
       " 'publication_year': '2023'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gpub.get_publication_info_from_pmid('37138086)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'pmid': 23456789, 'year': 2013, 'title': 'Hospital volume is associated with survival but not multimodality therapy in Medicare patients with advanced head and neck cancer.', 'authors': 'Arun Sharma, Stephen M Schwartz, Eduardo Méndez', 'journal': 'Cancer', 'is_research_article': 'Yes', 'relative_citation_ratio': 1.77, 'nih_percentile': 70.8, 'human': 1.0, 'animal': 0.0, 'molecular_cellular': 0.0, 'apt': 0.75, 'is_clinical': 'No', 'citation_count': 45, 'citations_per_year': 4.5, 'expected_citations_per_year': 2.547166821310601, 'field_citation_rate': 5.361749145554551, 'provisional': 'No', 'x_coord': 0.0, 'y_coord': 1.0, 'cited_by_clin': [25488965, 29180076], 'cited_by': [30186960, 34399637, 30220318, 37564472, 34795020, 28606602, 24123512, 36746098, 29100787, 26777060, 26553389, 25488965, 30194691, 35792549, 33556919, 27061951, 24706437, 29794540, 25042524, 28079775, 35547406, 32600116, 24488549, 31334365, 30409307, 35868508, 26868285, 29079897, 33449369, 32191271, 30698823, 25681489, 37847431, 30173901, 27571233, 28640546, 33998094, 36826717, 30737961, 25983103, 29180076, 29112234, 33857781, 30447127, 25797461], 'references': [3558716, 16161069, 20168233, 19340866, 14645636, 14568854, 8323597, 21181983, 15609302, 9552032, 15128894, 11146273, 9842949, 11463014, 10613347, 18436289, 1607900, 20589761, 15128893, 21125633, 18559875, 8985019, 22314971, 9423262, 12187164, 12187163, 15161894], 'doi': '10.1002/cncr.27976', 'last_modified': '10/31/2023, 06:39:28'}\n"
     ]
    }
   ],
   "source": [
    "# Basic iCite python example\n",
    "\n",
    "response = requests.get(\n",
    "    \"/\".join([\n",
    "        \"https://icite.od.nih.gov/api\",\n",
    "        \"pubs\",\n",
    "        \"23456789\",\n",
    "    ]),\n",
    ")\n",
    "pub = response.json()\n",
    "print(pub)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Try a one-to-one replacement of the Entrez `get_publication_info_from_pmid` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_publication_info_from_pmid_icite(pmid):\n",
    "    \"\"\"\n",
    "    Get publication information for a given PMID using the iCite API.\n",
    "\n",
    "    :param pmid: PubMed ID (str)\n",
    "    :return: Dictionary containing publication information\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Use the iCite API to get publication data\n",
    "        response = requests.get(f\"https://icite.od.nih.gov/api/pubs/{pmid}\")\n",
    "        pub = response.json()\n",
    "\n",
    "        # Extract relevant information\n",
    "        publication_info = {\n",
    "            'publication_id': pub.get('pmid', ''),\n",
    "            'title': pub.get('title', ''),\n",
    "            'authors': pub.get('authors', ''),\n",
    "            'publication_year': pub.get('year', ''),\n",
    "            'doi':pub.get('doi', ''),\n",
    "            'citation_count': pub.get('citation_count', ''),\n",
    "            'relative_citation_ratio': pub.get('relative_citation_ratio', ''),\n",
    "        }\n",
    "\n",
    "        return publication_info\n",
    "\n",
    "    except Exception as e:\n",
    "        # Use tqdm.write() instead of print() for long processes\n",
    "        tqdm.write(f\"Error fetching information for PMID {pmid} from iCite API: {e}\")\n",
    "        #print(f\"Error fetching information for PMID {pmid} from iCite API: {e}\")\n",
    "        return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'publication_id': 37138086,\n",
       " 'title': 'Glioblastoma remodelling of human neural circuits decreases survival.',\n",
       " 'authors': 'Saritha Krishna, Abrar Choudhury, Michael B Keough, Kyounghee Seo, Lijun Ni, Sofia Kakaizada, Anthony Lee, Alexander Aabedi, Galina Popova, Benjamin Lipkin, Caroline Cao, Cesar Nava Gonzales, Rasika Sudharshan, Andrew Egladyous, Nyle Almeida, Yalan Zhang, Annette M Molinaro, Humsa S Venkatesh, Andy G S Daniel, Kiarash Shamardani, Jeanette Hyer, Edward F Chang, Anne Findlay, Joanna J Phillips, Srikantan Nagarajan, David R Raleigh, David Brang, Michelle Monje, Shawn L Hervey-Jumper',\n",
       " 'publication_year': 2023,\n",
       " 'doi': '10.1038/s41586-023-06036-1',\n",
       " 'citation_count': 21,\n",
       " 'relative_citation_ratio': 10.5}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_publication_info_from_pmid_icite('37138086')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Checkpoint loading instead of regathering data during development\n",
    "pmid_filename = 'gathered_pmids_20231110.csv'\n",
    "df_pmid = pd.read_csv(pmid_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate through each unique PMID with tqdm progress bar\n",
    "def get_pub_info_test_loop(df_pmid):\n",
    "\n",
    "    df_pmid_info = pd.DataFrame()\n",
    "\n",
    "    for pmid in tqdm(df_pmid['pmid'].unique(), \n",
    "                    #total=remaining_pmid_count, \n",
    "                    ncols=80):\n",
    "        try:\n",
    "            # Use PubMed API to get publication data\n",
    "            publication_info = get_publication_info_from_pmid_icite(pmid)\n",
    "\n",
    "            if publication_info:\n",
    "                # Combine the information with the original DataFrame\n",
    "                df_current = pd.DataFrame({\n",
    "                    'pmid': pmid,\n",
    "                    'title': publication_info['title'],\n",
    "                    'authors': publication_info['authors'],\n",
    "                    'publication_year': publication_info['publication_year'],\n",
    "                    'doi': publication_info['doi'],\n",
    "                    'citation_count': publication_info['citation_count'],\n",
    "                    'relative_citation_ratio': publication_info['relative_citation_ratio']\n",
    "                }, index=[0])\n",
    "\n",
    "                # Add the current DataFrame to df_pmid_info\n",
    "                df_pmid_info = pd.concat([df_pmid_info, df_current], ignore_index=True)\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"Error processing PMID {pmid}: {e}\")\n",
    "            # Fill in fields with NaN if not available\n",
    "            df_current = pd.DataFrame({\n",
    "                'pmid': pmid,\n",
    "                'title': pd.NA,\n",
    "                'authors': pd.NA,\n",
    "                'publication_year': pd.NA,\n",
    "                'doi': pd.NA,\n",
    "                'citation_count': pd.NA,\n",
    "                'relative_citation_ratio': pd.NA\n",
    "            }, index=[0])\n",
    "\n",
    "            # Add the current DataFrame to df_pmid_info\n",
    "            df_pmid_info = pd.concat([df_pmid_info, df_current], ignore_index=True)\n",
    "\n",
    "    return df_pmid_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                           | 1/914 [00:00<11:27,  1.33it/s]C:\\Users\\dormanza\\AppData\\Local\\Temp\\2\\ipykernel_13608\\659356853.py:26: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  df_pmid_info = pd.concat([df_pmid_info, df_current], ignore_index=True)\n",
      "100%|█████████████████████████████████████████| 914/914 [13:41<00:00,  1.11it/s]\n"
     ]
    }
   ],
   "source": [
    "df_pmid_info_icite = get_pub_info_test_loop(df_pmid.head(1000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>publication_year</th>\n",
       "      <th>doi</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>relative_citation_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36127808</td>\n",
       "      <td>Genetic ancestry, differential gene expression...</td>\n",
       "      <td>Freddy A Barragan, Lauren J Mills, Andrew R Ra...</td>\n",
       "      <td>2023</td>\n",
       "      <td>10.1002/cam4.5266</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29074302</td>\n",
       "      <td>Endogenous antibody responses to mucin 1 in a ...</td>\n",
       "      <td>Janardan P Pandey, Aryan M Namboodiri, Bethany...</td>\n",
       "      <td>2018</td>\n",
       "      <td>10.1016/j.imbio.2017.10.028</td>\n",
       "      <td>2</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31387361</td>\n",
       "      <td>Defects in the Exocyst-Cilia Machinery Cause B...</td>\n",
       "      <td>Diana Fulmer, Katelynn Toomer, Lilong Guo, Kel...</td>\n",
       "      <td>2019</td>\n",
       "      <td>10.1161/CIRCULATIONAHA.119.038376</td>\n",
       "      <td>34</td>\n",
       "      <td>1.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29027980</td>\n",
       "      <td>The Plasticizer Bisphenol A Perturbs the Hepat...</td>\n",
       "      <td>Ludivine Renaud, Willian A da Silveira, E Star...</td>\n",
       "      <td>2017</td>\n",
       "      <td>10.3390/genes8100269</td>\n",
       "      <td>20</td>\n",
       "      <td>1.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>29309429</td>\n",
       "      <td>ShinyGPA: An interactive visualization toolkit...</td>\n",
       "      <td>Emma Kortemeier, Paula S Ramos, Kelly J Hunt, ...</td>\n",
       "      <td>2018</td>\n",
       "      <td>10.1371/journal.pone.0190949</td>\n",
       "      <td>2</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>909</th>\n",
       "      <td>32276990</td>\n",
       "      <td>Mesenchymal and MAPK Expression Signatures Ass...</td>\n",
       "      <td>Josh Lewis Stern, Grace Hibshman, Kevin Hu, Sa...</td>\n",
       "      <td>2020</td>\n",
       "      <td>10.1158/1541-7786.MCR-19-1244</td>\n",
       "      <td>17</td>\n",
       "      <td>1.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910</th>\n",
       "      <td>32525984</td>\n",
       "      <td>Tissue- and development-stage-specific mRNA an...</td>\n",
       "      <td>Anshuman Panda, Anupama Yadav, Huwate Yeerna, ...</td>\n",
       "      <td>2020</td>\n",
       "      <td>10.1093/nar/gkaa485</td>\n",
       "      <td>12</td>\n",
       "      <td>0.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>911</th>\n",
       "      <td>36371231</td>\n",
       "      <td>Differential regulation of TNFα and IL-6 expre...</td>\n",
       "      <td>Ida Deichaite, Timothy J Sears, Leisa Sutton, ...</td>\n",
       "      <td>2022</td>\n",
       "      <td>10.1186/s12967-022-03731-x</td>\n",
       "      <td>2</td>\n",
       "      <td>0.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>912</th>\n",
       "      <td>37682073</td>\n",
       "      <td>Transcriptional subtypes of glottic cancer cha...</td>\n",
       "      <td>Bharat A Panuganti, Christine Carico, Harishan...</td>\n",
       "      <td>2023</td>\n",
       "      <td>10.1002/hed.27514</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>913</th>\n",
       "      <td>37306759</td>\n",
       "      <td>Syk Inhibition Reprograms Tumor-Associated Mac...</td>\n",
       "      <td>Deepak Rohila, In Hwan Park, Timothy V Pham, J...</td>\n",
       "      <td>2023</td>\n",
       "      <td>10.1158/0008-5472.CAN-22-3645</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>914 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         pmid                                              title  \\\n",
       "0    36127808  Genetic ancestry, differential gene expression...   \n",
       "1    29074302  Endogenous antibody responses to mucin 1 in a ...   \n",
       "2    31387361  Defects in the Exocyst-Cilia Machinery Cause B...   \n",
       "3    29027980  The Plasticizer Bisphenol A Perturbs the Hepat...   \n",
       "4    29309429  ShinyGPA: An interactive visualization toolkit...   \n",
       "..        ...                                                ...   \n",
       "909  32276990  Mesenchymal and MAPK Expression Signatures Ass...   \n",
       "910  32525984  Tissue- and development-stage-specific mRNA an...   \n",
       "911  36371231  Differential regulation of TNFα and IL-6 expre...   \n",
       "912  37682073  Transcriptional subtypes of glottic cancer cha...   \n",
       "913  37306759  Syk Inhibition Reprograms Tumor-Associated Mac...   \n",
       "\n",
       "                                               authors  publication_year  \\\n",
       "0    Freddy A Barragan, Lauren J Mills, Andrew R Ra...              2023   \n",
       "1    Janardan P Pandey, Aryan M Namboodiri, Bethany...              2018   \n",
       "2    Diana Fulmer, Katelynn Toomer, Lilong Guo, Kel...              2019   \n",
       "3    Ludivine Renaud, Willian A da Silveira, E Star...              2017   \n",
       "4    Emma Kortemeier, Paula S Ramos, Kelly J Hunt, ...              2018   \n",
       "..                                                 ...               ...   \n",
       "909  Josh Lewis Stern, Grace Hibshman, Kevin Hu, Sa...              2020   \n",
       "910  Anshuman Panda, Anupama Yadav, Huwate Yeerna, ...              2020   \n",
       "911  Ida Deichaite, Timothy J Sears, Leisa Sutton, ...              2022   \n",
       "912  Bharat A Panuganti, Christine Carico, Harishan...              2023   \n",
       "913  Deepak Rohila, In Hwan Park, Timothy V Pham, J...              2023   \n",
       "\n",
       "                                   doi  citation_count  \\\n",
       "0                    10.1002/cam4.5266               4   \n",
       "1          10.1016/j.imbio.2017.10.028               2   \n",
       "2    10.1161/CIRCULATIONAHA.119.038376              34   \n",
       "3                 10.3390/genes8100269              20   \n",
       "4         10.1371/journal.pone.0190949               2   \n",
       "..                                 ...             ...   \n",
       "909      10.1158/1541-7786.MCR-19-1244              17   \n",
       "910                10.1093/nar/gkaa485              12   \n",
       "911         10.1186/s12967-022-03731-x               2   \n",
       "912                  10.1002/hed.27514               0   \n",
       "913      10.1158/0008-5472.CAN-22-3645               0   \n",
       "\n",
       "     relative_citation_ratio  \n",
       "0                        NaN  \n",
       "1                       0.09  \n",
       "2                       1.95  \n",
       "3                       1.17  \n",
       "4                       0.08  \n",
       "..                       ...  \n",
       "909                     1.13  \n",
       "910                     0.76  \n",
       "911                     0.57  \n",
       "912                      NaN  \n",
       "913                      NaN  \n",
       "\n",
       "[914 rows x 7 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pmid_info_icite"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Try processing in batches for fewer API calls\n",
    "The above approach works but is even slower per iteration than the Entrez API (~15min for 1000). Try a batching approach where multiple PMIDs are sent in a single call."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_publication_info_from_pmid_icite_batch(pmids):\n",
    "    \"\"\"\n",
    "    Get publication information for a list of PMIDs using the iCite API.\n",
    "\n",
    "    :param pmids: List of PubMed IDs (str)\n",
    "    :return: DataFrame containing publication information\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Join PMIDs into a comma-separated string\n",
    "        pmid_str = ','.join(pmids)\n",
    "        \n",
    "        # Use the iCite API to get publication data for all PMIDs\n",
    "        response = requests.get(f\"https://icite.od.nih.gov/api/pubs?pmids={pmid_str}\")\n",
    "        pubs = response.json()\n",
    "\n",
    "        # Initialize an empty list to store publication information for each PMID\n",
    "        publication_info_list = []\n",
    "\n",
    "        for pub in pubs:\n",
    "            # Extract relevant information\n",
    "            publication_info = {\n",
    "                'pmid': pub.get('pmid', ''),\n",
    "                'title': pub.get('title', ''),\n",
    "                'authors': pub.get('authors', ''),\n",
    "                'publication_year': pub.get('year', ''),\n",
    "                'doi': pub.get('doi', ''),\n",
    "                'citation_count': pub.get('citation_count', ''),\n",
    "                'relative_citation_ratio': pub.get('relative_citation_ratio', ''),\n",
    "            }\n",
    "            \n",
    "            # Add data to running list\n",
    "            publication_info_list.append(publication_info)\n",
    "\n",
    "        return pd.DataFrame(publication_info_list)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error fetching information for PMIDs {pmids} from iCite API: {e}\")\n",
    "        return pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pub_info_batched(df_pmid, batch_size=10):\n",
    "    df_pmid_info = pd.DataFrame()\n",
    "\n",
    "    # Extract unique PMIDs\n",
    "    unique_pmids = df_pmid['pmid'].unique()\n",
    "\n",
    "    # Split PMIDs into batches\n",
    "    pmid_batches = [\n",
    "        unique_pmids[i : i + batch_size].astype(str) for i in range(0, len(unique_pmids), batch_size)\n",
    "    ]\n",
    "\n",
    "    for batch in tqdm(pmid_batches, ncols=80):\n",
    "        try:\n",
    "            # Use iCite API to get publication data for the batch\n",
    "            batch_info = [get_publication_info_from_pmid_icite(pmid) for pmid in batch]\n",
    "\n",
    "            # Filter out None results (failed API calls)\n",
    "            batch_info = [info for info in batch_info if info is not None]\n",
    "\n",
    "            # Combine the information with the original DataFrame\n",
    "            df_current = pd.DataFrame(batch_info)\n",
    "\n",
    "            # Add the current DataFrame to df_pmid_info\n",
    "            df_pmid_info = pd.concat([df_pmid_info, df_current], ignore_index=True)\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"Error processing batch of PMIDs: {e}\")\n",
    "\n",
    "    return df_pmid_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|▍                                          | 1/100 [00:00<01:17,  1.28it/s]C:\\Users\\dormanza\\AppData\\Local\\Temp\\2\\ipykernel_13608\\659356853.py:26: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  df_pmid_info = pd.concat([df_pmid_info, df_current], ignore_index=True)\n",
      "100%|█████████████████████████████████████████| 100/100 [01:22<00:00,  1.22it/s]\n"
     ]
    }
   ],
   "source": [
    "# No batching\n",
    "test = get_pub_info_test_loop(df_pmid.head(100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████| 10/10 [01:17<00:00,  7.77s/it]\n"
     ]
    }
   ],
   "source": [
    "# Batch size default 10\n",
    "test = get_pub_info_batched(df_pmid.head(100))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compare timing of gathering iCite PMID info in single vs batched calls\n",
    "Table gathers data from cells below\n",
    "\n",
    "| API    | PMIDs | Batch Size| Time (mm:ss)| Rate (s/pmid) |\n",
    "| ------:| -----:| ---------:| -----------:| -------------:|\n",
    "| Entrez | 100   | None      | 00:34        | 0.34   |\n",
    "| iCite  | 100   | None      | 01:22        | 0.82   |\n",
    "| iCite  | 100   | 1         | 01:18        | 0.78   |\n",
    "| iCite  | 100   | 5         | 01:18        | 0.78   |\n",
    "| iCite  | 100   | 10        | 01:18        | 0.78   |\n",
    "| iCite  | 100   | 50        | 01:19        | 0.79   |\n",
    "| Entrez | 500   | None      | 02:08        | 0.26   |\n",
    "| iCite  | 500   | None      | 07:09        | 0.86   |\n",
    "| iCite  | 500   | 100       | 06:52        | 0.82   |\n",
    "\n",
    "\n",
    "\n",
    "#### Summary\n",
    "1. The Entrez API is 3-4 times faster than the iCite API\n",
    "2. Batching the iCite API calls does not significantly improve performance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                           | 1/494 [00:00<07:24,  1.11it/s]C:\\Users\\dormanza\\AppData\\Local\\Temp\\2\\ipykernel_13608\\659356853.py:26: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  df_pmid_info = pd.concat([df_pmid_info, df_current], ignore_index=True)\n",
      " 98%|████████████████████████████████████████▎| 486/494 [06:45<00:06,  1.30it/s]"
     ]
    }
   ],
   "source": [
    "# 500 pmids, no batching\n",
    "test = get_pub_info_test_loop(df_pmid.head(500))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████| 5/5 [07:09<00:00, 85.85s/it]\n"
     ]
    }
   ],
   "source": [
    "# 500 pmids, batch size 100\n",
    "test = get_pub_info_batched(df_pmid.head(500), batch_size=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████| 20/20 [01:18<00:00,  3.92s/it]\n"
     ]
    }
   ],
   "source": [
    "# 100 pmids, batch size 5\n",
    "test = get_pub_info_batched(df_pmid.head(100), batch_size=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████| 2/2 [01:18<00:00, 39.35s/it]\n"
     ]
    }
   ],
   "source": [
    "# 100 pmids, batch size 50\n",
    "test = get_pub_info_batched(df_pmid.head(100), batch_size=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|▍                                          | 1/100 [00:00<01:09,  1.43it/s]C:\\Users\\dormanza\\AppData\\Local\\Temp\\2\\ipykernel_13608\\2002022982.py:24: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  df_pmid_info = pd.concat([df_pmid_info, df_current], ignore_index=True)\n",
      "100%|█████████████████████████████████████████| 100/100 [01:18<00:00,  1.27it/s]\n"
     ]
    }
   ],
   "source": [
    "# 100 pmids, batch size 1\n",
    "test = get_pub_info_batched(df_pmid.head(100), batch_size=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate through each unique PMID with tqdm progress bar\n",
    "def get_pub_info_test_loop_entrez(df_pmid):\n",
    "\n",
    "    df_pmid_info = pd.DataFrame()\n",
    "\n",
    "    for pmid in tqdm(df_pmid['pmid'].unique(), \n",
    "                    #total=remaining_pmid_count, \n",
    "                    ncols=80):\n",
    "        try:\n",
    "            # Use PubMed API to get publication data\n",
    "            publication_info = gpub.get_publication_info_from_pmid(pmid)\n",
    "\n",
    "            if publication_info:\n",
    "                # Combine the information with the original DataFrame\n",
    "                df_current = pd.DataFrame({\n",
    "                    'pmid': pmid,\n",
    "                    'title': publication_info['title'],\n",
    "                    'authors': publication_info['authors'],\n",
    "                    'publication_year': publication_info['publication_year'],\n",
    "                    # 'doi': publication_info['doi'],\n",
    "                    # 'citation_count': publication_info['citation_count'],\n",
    "                    # 'relative_citation_ratio': publication_info['relative_citation_ratio']\n",
    "                }, index=[0])\n",
    "\n",
    "                # Add the current DataFrame to df_pmid_info\n",
    "                df_pmid_info = pd.concat([df_pmid_info, df_current], ignore_index=True)\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"Error processing PMID {pmid}: {e}\")\n",
    "            # Fill in fields with NaN if not available\n",
    "            df_current = pd.DataFrame({\n",
    "                'pmid': pmid,\n",
    "                'title': pd.NA,\n",
    "                'authors': pd.NA,\n",
    "                'publication_year': pd.NA,\n",
    "                # 'doi': pd.NA,\n",
    "                # 'citation_count': pd.NA,\n",
    "                # 'relative_citation_ratio': pd.NA\n",
    "            }, index=[0])\n",
    "\n",
    "            # Add the current DataFrame to df_pmid_info\n",
    "            df_pmid_info = pd.concat([df_pmid_info, df_current], ignore_index=True)\n",
    "\n",
    "    return df_pmid_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████| 100/100 [00:34<00:00,  2.89it/s]\n"
     ]
    }
   ],
   "source": [
    "# 100 pmids, Entrez method\n",
    "test = get_pub_info_test_loop_entrez(df_pmid.head(100))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 89%|████████████████████████████████████▋    | 442/494 [01:52<00:10,  4.75it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error fetching information for PMID 33579955: list index out of range\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 95%|███████████████████████████████████████  | 470/494 [02:02<00:04,  4.98it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error fetching information for PMID 33574288: list index out of range\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████| 494/494 [02:08<00:00,  3.86it/s]\n"
     ]
    }
   ],
   "source": [
    "# 100 pmids, Entrez method\n",
    "test = get_pub_info_test_loop_entrez(df_pmid.head(500))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### See if pulling fewer fields from iCite is faster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_publication_info_from_pmid_icite(pmid, fields='all'):\n",
    "    \"\"\"\n",
    "    Get publication information for a given PMID using the iCite API.\n",
    "\n",
    "    :param pmid: PubMed ID (str)\n",
    "    :return: Dictionary containing publication information\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Use the iCite API to get publication data\n",
    "        if fields == ['all']:\n",
    "            response = requests.get(f\"https://icite.od.nih.gov/api/pubs/{pmid}\")\n",
    "        \n",
    "        # If a list of fields is provided, include only those in the response\n",
    "        else:\n",
    "            field_str = ','.join(fields)\n",
    "            response = requests.get(f\"https://icite.od.nih.gov/api/pubs/{pmid}\"\n",
    "                                    f\"&fl={field_str}\")\n",
    "        pub = response.json()\n",
    "\n",
    "        # Extract relevant information\n",
    "        publication_info = {\n",
    "            'publication_id': pub.get('pmid', ''),\n",
    "            # 'title': pub.get('title', ''),\n",
    "            # 'authors': pub.get('authors', ''),\n",
    "            # 'publication_year': pub.get('year', ''),\n",
    "            'doi':pub.get('doi', ''),\n",
    "            'citation_count': pub.get('citation_count', ''),\n",
    "            'relative_citation_ratio': pub.get('relative_citation_ratio', ''),\n",
    "        }\n",
    "\n",
    "        return publication_info\n",
    "\n",
    "    except Exception as e:\n",
    "        # Use tqdm.write() instead of print() for long processes\n",
    "        tqdm.write(f\"Error fetching information for PMID {pmid} from iCite API: {e}\")\n",
    "        #print(f\"Error fetching information for PMID {pmid} from iCite API: {e}\")\n",
    "        return None\n",
    "\n",
    "# Iterate through each unique PMID with tqdm progress bar\n",
    "def get_pub_info_test_loop(df_pmid, fields='all'):\n",
    "\n",
    "    df_pmid_info = pd.DataFrame()\n",
    "\n",
    "    for pmid in tqdm(df_pmid['pmid'].unique(), \n",
    "                    #total=remaining_pmid_count, \n",
    "                    ncols=80):\n",
    "        try:\n",
    "            # Use PubMed API to get publication data\n",
    "            publication_info = get_publication_info_from_pmid_icite(pmid, fields)\n",
    "\n",
    "            if publication_info:\n",
    "                # Combine the information with the original DataFrame\n",
    "                df_current = pd.DataFrame({\n",
    "                    'pmid': pmid,\n",
    "                    # 'title': publication_info['title'],\n",
    "                    # 'authors': publication_info['authors'],\n",
    "                    # 'publication_year': publication_info['publication_year'],\n",
    "                    'doi': publication_info['doi'],\n",
    "                    'citation_count': publication_info['citation_count'],\n",
    "                    'relative_citation_ratio': publication_info['relative_citation_ratio']\n",
    "                }, index=[0])\n",
    "\n",
    "                # Add the current DataFrame to df_pmid_info\n",
    "                df_pmid_info = pd.concat([df_pmid_info, df_current], ignore_index=True)\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"Error processing PMID {pmid}: {e}\")\n",
    "            # Fill in fields with NaN if not available\n",
    "            df_current = pd.DataFrame({\n",
    "                'pmid': pmid,\n",
    "                # 'title': pd.NA,\n",
    "                # 'authors': pd.NA,\n",
    "                # 'publication_year': pd.NA,\n",
    "                'doi': pd.NA,\n",
    "                'citation_count': pd.NA,\n",
    "                'relative_citation_ratio': pd.NA\n",
    "            }, index=[0])\n",
    "\n",
    "            # Add the current DataFrame to df_pmid_info\n",
    "            df_pmid_info = pd.concat([df_pmid_info, df_current], ignore_index=True)\n",
    "\n",
    "    return df_pmid_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████| 100/100 [01:19<00:00,  1.26it/s]\n"
     ]
    }
   ],
   "source": [
    "test = get_pub_info_test_loop(df_pmid.head(100), fields=['pmid', 'citation_count', 'doi', 'relative_citation_ratio'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No notable change in runtime. 1:19m vs 1:22m to complete 100 pmids, gathering either a few or all fields"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 2: Using Bulk Snapshow Download  \n",
    "\n",
    "Still using Jira Ticket: [INS-790](https://tracker.nci.nih.gov/browse/INS-790)  \n",
    "\n",
    "Pulling data from the iCite API was too slow at the scale we need. Instead, try using the bulk download of the monthly database snapshot that iCite provides on NIH FigShare: https://nih.figshare.com/collections/iCite_Database_Snapshots_NIH_Open_Citation_Collection_/4586573"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---TIMESTAMP OVERRIDE IN USE---\n",
      "---Disable this with comments in config.py for default behavior---\n"
     ]
    }
   ],
   "source": [
    "# Method to import from parent directory\n",
    "import os\n",
    "import sys\n",
    "root_dir = os.path.abspath(os.path.join(os.getcwd(), \"../\"))\n",
    "sys.path.append(root_dir)\n",
    "\n",
    "#import requests\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "# Get all existing publication functions\n",
    "import modules.gather_publication_data as gpub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Checkpoint loading instead of regathering data during development\n",
    "pmid_filename = 'gathered_pmids_20231110.csv'\n",
    "df_pmid = pd.read_csv(pmid_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coreproject</th>\n",
       "      <th>pmid</th>\n",
       "      <th>applid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>36127808</td>\n",
       "      <td>10902170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29074302</td>\n",
       "      <td>9321971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>31387361</td>\n",
       "      <td>9321971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29027980</td>\n",
       "      <td>9321971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29309429</td>\n",
       "      <td>9321971</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   coreproject      pmid    applid\n",
       "0  R01CA239701  36127808  10902170\n",
       "1  R21CA209848  29074302   9321971\n",
       "2  R21CA209848  31387361   9321971\n",
       "3  R21CA209848  29027980   9321971\n",
       "4  R21CA209848  29309429   9321971"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pmid.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define path to icite bulk csv.zip file and verify \n",
    "icite_filepath = '../data/raw/icite/2023-10/icite_metadata.zip'\n",
    "os.path.exists(icite_filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set columns of intersest to pull from iCite\n",
    "cols = ['pmid','doi','title','authors','year','citation_count','relative_citation_ratio']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>year</th>\n",
       "      <th>journal</th>\n",
       "      <th>is_research_article</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>field_citation_rate</th>\n",
       "      <th>expected_citations_per_year</th>\n",
       "      <th>...</th>\n",
       "      <th>molecular_cellular</th>\n",
       "      <th>x_coord</th>\n",
       "      <th>y_coord</th>\n",
       "      <th>apt</th>\n",
       "      <th>is_clinical</th>\n",
       "      <th>cited_by_clin</th>\n",
       "      <th>cited_by</th>\n",
       "      <th>references</th>\n",
       "      <th>provisional</th>\n",
       "      <th>last_modified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10.1016/0006-2944(75)90147-7</td>\n",
       "      <td>Formate assay in body fluids: application in m...</td>\n",
       "      <td>A B Makar, K E McMartin, M Palese, T R Tephly</td>\n",
       "      <td>1975</td>\n",
       "      <td>Biochem Med</td>\n",
       "      <td>True</td>\n",
       "      <td>101</td>\n",
       "      <td>5.203998</td>\n",
       "      <td>0.793752</td>\n",
       "      <td>...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>-0.144338</td>\n",
       "      <td>-0.250000</td>\n",
       "      <td>0.25</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27354968 34013366 30849241 21923939 36162727 3...</td>\n",
       "      <td>4972128 4332837 13672941 14203183 14161139 143...</td>\n",
       "      <td>No</td>\n",
       "      <td>10/31/2023, 08:00:32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>10.1016/0006-291x(75)90482-9</td>\n",
       "      <td>Delineation of the intimate details of the bac...</td>\n",
       "      <td>K S Bose, R H Sarma</td>\n",
       "      <td>1975</td>\n",
       "      <td>Biochem Biophys Res Commun</td>\n",
       "      <td>True</td>\n",
       "      <td>39</td>\n",
       "      <td>5.963835</td>\n",
       "      <td>0.905675</td>\n",
       "      <td>...</td>\n",
       "      <td>1.00</td>\n",
       "      <td>-0.866025</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>0.25</td>\n",
       "      <td>False</td>\n",
       "      <td>21542697.0</td>\n",
       "      <td>36225522 6267127 26376 37339783 37441102 37754...</td>\n",
       "      <td>4150960 4356257 4846745 4357832 4683494 441485...</td>\n",
       "      <td>No</td>\n",
       "      <td>10/31/2023, 08:12:21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>10.1016/0006-291x(75)90498-2</td>\n",
       "      <td>Metal substitutions incarbonic anhydrase: a ha...</td>\n",
       "      <td>R J Smith, R G Bryant</td>\n",
       "      <td>1975</td>\n",
       "      <td>Biochem Biophys Res Commun</td>\n",
       "      <td>True</td>\n",
       "      <td>17</td>\n",
       "      <td>5.358380</td>\n",
       "      <td>0.816492</td>\n",
       "      <td>...</td>\n",
       "      <td>0.57</td>\n",
       "      <td>-0.247436</td>\n",
       "      <td>-0.285714</td>\n",
       "      <td>0.25</td>\n",
       "      <td>False</td>\n",
       "      <td>37120631.0</td>\n",
       "      <td>25624746 33776281 32427033 28053241 35738176 2...</td>\n",
       "      <td>28089154 29927711 33870035 31115340 27536128 3...</td>\n",
       "      <td>No</td>\n",
       "      <td>10/31/2023, 08:12:58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>10.1016/0006-291x(75)90506-9</td>\n",
       "      <td>Effect of chloroquine on cultured fibroblasts:...</td>\n",
       "      <td>U N Wiesmann, S DiDonato, N N Herschkowitz</td>\n",
       "      <td>1975</td>\n",
       "      <td>Biochem Biophys Res Commun</td>\n",
       "      <td>True</td>\n",
       "      <td>75</td>\n",
       "      <td>5.586596</td>\n",
       "      <td>0.850108</td>\n",
       "      <td>...</td>\n",
       "      <td>0.80</td>\n",
       "      <td>-0.692820</td>\n",
       "      <td>-0.200000</td>\n",
       "      <td>0.05</td>\n",
       "      <td>False</td>\n",
       "      <td>21975914.0</td>\n",
       "      <td>564972 8907731 7060838 6734624 36936276 312582...</td>\n",
       "      <td>13663253 4271529 5021451 4607946 4374680 14907...</td>\n",
       "      <td>No</td>\n",
       "      <td>10/31/2023, 08:13:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>10.1016/0006-291x(75)90508-2</td>\n",
       "      <td>Atomic models for the polypeptide backbones of...</td>\n",
       "      <td>W A Hendrickson, K B Ward</td>\n",
       "      <td>1975</td>\n",
       "      <td>Biochem Biophys Res Commun</td>\n",
       "      <td>True</td>\n",
       "      <td>23</td>\n",
       "      <td>5.652509</td>\n",
       "      <td>0.859817</td>\n",
       "      <td>...</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.288675</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>0.05</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7118409 2619971 364941 3380793 856811 8372226 ...</td>\n",
       "      <td>4882249 5059118 14834145 1056020 5509841</td>\n",
       "      <td>No</td>\n",
       "      <td>10/31/2023, 08:13:09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   pmid                           doi  \\\n",
       "0     1  10.1016/0006-2944(75)90147-7   \n",
       "1     2  10.1016/0006-291x(75)90482-9   \n",
       "2     3  10.1016/0006-291x(75)90498-2   \n",
       "3     4  10.1016/0006-291x(75)90506-9   \n",
       "4     5  10.1016/0006-291x(75)90508-2   \n",
       "\n",
       "                                               title  \\\n",
       "0  Formate assay in body fluids: application in m...   \n",
       "1  Delineation of the intimate details of the bac...   \n",
       "2  Metal substitutions incarbonic anhydrase: a ha...   \n",
       "3  Effect of chloroquine on cultured fibroblasts:...   \n",
       "4  Atomic models for the polypeptide backbones of...   \n",
       "\n",
       "                                         authors  year  \\\n",
       "0  A B Makar, K E McMartin, M Palese, T R Tephly  1975   \n",
       "1                            K S Bose, R H Sarma  1975   \n",
       "2                          R J Smith, R G Bryant  1975   \n",
       "3     U N Wiesmann, S DiDonato, N N Herschkowitz  1975   \n",
       "4                      W A Hendrickson, K B Ward  1975   \n",
       "\n",
       "                      journal  is_research_article  citation_count  \\\n",
       "0                 Biochem Med                 True             101   \n",
       "1  Biochem Biophys Res Commun                 True              39   \n",
       "2  Biochem Biophys Res Commun                 True              17   \n",
       "3  Biochem Biophys Res Commun                 True              75   \n",
       "4  Biochem Biophys Res Commun                 True              23   \n",
       "\n",
       "   field_citation_rate  expected_citations_per_year  ...  molecular_cellular  \\\n",
       "0             5.203998                     0.793752  ...                0.50   \n",
       "1             5.963835                     0.905675  ...                1.00   \n",
       "2             5.358380                     0.816492  ...                0.57   \n",
       "3             5.586596                     0.850108  ...                0.80   \n",
       "4             5.652509                     0.859817  ...                0.33   \n",
       "\n",
       "    x_coord   y_coord   apt  is_clinical  cited_by_clin  \\\n",
       "0 -0.144338 -0.250000  0.25        False            NaN   \n",
       "1 -0.866025 -0.500000  0.25        False     21542697.0   \n",
       "2 -0.247436 -0.285714  0.25        False     37120631.0   \n",
       "3 -0.692820 -0.200000  0.05        False     21975914.0   \n",
       "4  0.288675 -0.500000  0.05        False            NaN   \n",
       "\n",
       "                                            cited_by  \\\n",
       "0  27354968 34013366 30849241 21923939 36162727 3...   \n",
       "1  36225522 6267127 26376 37339783 37441102 37754...   \n",
       "2  25624746 33776281 32427033 28053241 35738176 2...   \n",
       "3  564972 8907731 7060838 6734624 36936276 312582...   \n",
       "4  7118409 2619971 364941 3380793 856811 8372226 ...   \n",
       "\n",
       "                                          references  provisional  \\\n",
       "0  4972128 4332837 13672941 14203183 14161139 143...           No   \n",
       "1  4150960 4356257 4846745 4357832 4683494 441485...           No   \n",
       "2  28089154 29927711 33870035 31115340 27536128 3...           No   \n",
       "3  13663253 4271529 5021451 4607946 4374680 14907...           No   \n",
       "4           4882249 5059118 14834145 1056020 5509841           No   \n",
       "\n",
       "          last_modified  \n",
       "0  10/31/2023, 08:00:32  \n",
       "1  10/31/2023, 08:12:21  \n",
       "2  10/31/2023, 08:12:58  \n",
       "3  10/31/2023, 08:13:00  \n",
       "4  10/31/2023, 08:13:09  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Try reading first rows of the file\n",
    "df_icite_sample = pd.read_csv(icite_filepath, compression='zip', nrows=5)\n",
    "df_icite_sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pmid                                                                           1\n",
       "doi                                                 10.1016/0006-2944(75)90147-7\n",
       "title                          Formate assay in body fluids: application in m...\n",
       "authors                            A B Makar, K E McMartin, M Palese, T R Tephly\n",
       "year                                                                        1975\n",
       "journal                                                              Biochem Med\n",
       "is_research_article                                                         True\n",
       "citation_count                                                               101\n",
       "field_citation_rate                                                     5.203998\n",
       "expected_citations_per_year                                             0.793752\n",
       "citations_per_year                                                      2.104167\n",
       "relative_citation_ratio                                                     2.65\n",
       "nih_percentile                                                               NaN\n",
       "human                                                                       0.17\n",
       "animal                                                                      0.33\n",
       "molecular_cellular                                                           0.5\n",
       "x_coord                                                                -0.144338\n",
       "y_coord                                                                    -0.25\n",
       "apt                                                                         0.25\n",
       "is_clinical                                                                False\n",
       "cited_by_clin                                                                NaN\n",
       "cited_by                       27354968 34013366 30849241 21923939 36162727 3...\n",
       "references                     4972128 4332837 13672941 14203183 14161139 143...\n",
       "provisional                                                                   No\n",
       "last_modified                                               10/31/2023, 08:00:32\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show first row to see columns and sample values\n",
    "df_icite_sample.loc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "PMIDs seem to be serial, so we can get an idea of the count/range of rows that we'll need to access from the iCite data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lowest PMID:   1,279,509\n",
      "Highest PMID: 37,947,614\n"
     ]
    }
   ],
   "source": [
    "# Find highest PMID in the gathered PMID list.\n",
    "print(f\"Lowest PMID:   {df_pmid['pmid'].min():,}\\n\"\n",
    "      f\"Highest PMID: {df_pmid['pmid'].max():,}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Range: 36,668,105\n"
     ]
    }
   ],
   "source": [
    "print(f\"Range: {37947614-1279509:,}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write a function that can iterate through this large dataset in chunks and pull out info for any relevant PMID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# THERE IS AN UPDATED VERSION OF THIS FUNCTION LATER\n",
    "def enrich_df_with_icite(df_pmid, icite_filepath, cols=None, chunk_size=250000):\n",
    "    \"\"\"\n",
    "    Enrich a DataFrame with iCite data using chunks.\n",
    "\n",
    "    :param df_pmid (pd.DataFrame): DataFrame containing PMID-related data.\n",
    "    :param icite_filepath (str): Path to the zipped iCite CSV file.\n",
    "    :param cols (list): List of columns to select from df_pmid before merging.\n",
    "                   If None, all columns from df_pmid are used.\n",
    "    :param chunk_size (int): Number of rows to read per chunk.\n",
    "\n",
    "    :return: pd.DataFrame: Enriched DataFrame with iCite data.\n",
    "    \"\"\"\n",
    "\n",
    "    # If cols is not specified, use all columns from df_pmid\n",
    "    if cols is None:\n",
    "        cols = df_pmid.columns.tolist()\n",
    "\n",
    "    # Initialize an empty DataFrame to store the enriched data\n",
    "    df_enriched = pd.DataFrame()\n",
    "\n",
    "    # Create a tqdm wrapper around the generator to track progress\n",
    "    chunks = tqdm(pd.read_csv(icite_filepath, compression='zip', chunksize=chunk_size),\n",
    "                  desc=\"Processing iCite\", unit=\"chunk\")\n",
    "\n",
    "    # Iterate through chunks of the iCite DataFrame\n",
    "    for chunk in chunks:\n",
    "        # Merge only the specified columns from df_pmid\n",
    "        df_merged = pd.merge(df_pmid, chunk[cols], how='left', on='pmid')\n",
    "\n",
    "        # Append the merged DataFrame to the result\n",
    "        df_enriched = pd.concat([df_enriched, df_merged], ignore_index=True)\n",
    "\n",
    "    return df_enriched\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing iCite: 143chunk [07:39,  3.60s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite: 144chunk [07:43,  3.72s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite: 145chunk [07:47,  3.66s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (21,22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite: 146chunk [07:50,  3.22s/chunk]\n"
     ]
    }
   ],
   "source": [
    "# Make it go\n",
    "df_pmid_enriched = enrich_df_with_icite(df_pmid, icite_filepath, cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coreproject</th>\n",
       "      <th>pmid</th>\n",
       "      <th>applid</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>year</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>relative_citation_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>36127808</td>\n",
       "      <td>10902170</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29074302</td>\n",
       "      <td>9321971</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>31387361</td>\n",
       "      <td>9321971</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29027980</td>\n",
       "      <td>9321971</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29309429</td>\n",
       "      <td>9321971</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25626937</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>35471840</td>\n",
       "      <td>10690040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25626938</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>35793873</td>\n",
       "      <td>10690040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25626939</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>36509758</td>\n",
       "      <td>10690040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25626940</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>36775354</td>\n",
       "      <td>10690040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25626941</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>37286865</td>\n",
       "      <td>10690040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25626942 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          coreproject      pmid    applid  doi title authors  year  \\\n",
       "0         R01CA239701  36127808  10902170  NaN   NaN     NaN   NaN   \n",
       "1         R21CA209848  29074302   9321971  NaN   NaN     NaN   NaN   \n",
       "2         R21CA209848  31387361   9321971  NaN   NaN     NaN   NaN   \n",
       "3         R21CA209848  29027980   9321971  NaN   NaN     NaN   NaN   \n",
       "4         R21CA209848  29309429   9321971  NaN   NaN     NaN   NaN   \n",
       "...               ...       ...       ...  ...   ...     ...   ...   \n",
       "25626937  P50CA196530  35471840  10690040  NaN   NaN     NaN   NaN   \n",
       "25626938  P50CA196530  35793873  10690040  NaN   NaN     NaN   NaN   \n",
       "25626939  P50CA196530  36509758  10690040  NaN   NaN     NaN   NaN   \n",
       "25626940  P50CA196530  36775354  10690040  NaN   NaN     NaN   NaN   \n",
       "25626941  P50CA196530  37286865  10690040  NaN   NaN     NaN   NaN   \n",
       "\n",
       "          citation_count  relative_citation_ratio  \n",
       "0                    NaN                      NaN  \n",
       "1                    NaN                      NaN  \n",
       "2                    NaN                      NaN  \n",
       "3                    NaN                      NaN  \n",
       "4                    NaN                      NaN  \n",
       "...                  ...                      ...  \n",
       "25626937             NaN                      NaN  \n",
       "25626938             NaN                      NaN  \n",
       "25626939             NaN                      NaN  \n",
       "25626940             NaN                      NaN  \n",
       "25626941             NaN                      NaN  \n",
       "\n",
       "[25626942 rows x 9 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pmid_enriched"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All iCite values are appearing blank, but other checks show that the iCite data is there.  \n",
    "Running enrich_df_with_icite only took around 8 minutes, so I'll try loading the full iCite file as-is and manipulating afterwards."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_large_csv(filepath, chunk_size=250000):\n",
    "    \"\"\"\n",
    "    Load a large CSV file into a single DataFrame using chunks.\n",
    "\n",
    "    :param filepath (str): Path to the CSV file.\n",
    "    :param chunk_size (int): Number of rows to read per chunk.\n",
    "\n",
    "    :return: pd.DataFrame: Loaded DataFrame.\n",
    "    \"\"\"\n",
    "\n",
    "    # Initialize an empty DataFrame to store the loaded data\n",
    "    df = pd.DataFrame()\n",
    "\n",
    "    # Create a tqdm wrapper around the generator to track progress\n",
    "    chunks = tqdm(pd.read_csv(filepath, chunksize=chunk_size),\n",
    "                  desc=f\"Loading {filepath}\", unit=\"chunk\")\n",
    "\n",
    "    # Iterate through chunks of the DataFrame\n",
    "    for chunk in chunks:\n",
    "        # Append the chunk to the result DataFrame\n",
    "        df = pd.concat([df, chunk], ignore_index=True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Try loading entire iCite csv into a single df in chunks\n",
    "# df_icite_full = load_large_csv(icite_filepath, chunk_size=250000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above attempt was on track to take a few hours. If I hadn't stopped it, it might have succeeded or it might have hit a memory error.  \n",
    "\n",
    "I'll try to make the merging-by-chunks method work. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check that the pmid column is the same type in both dfs\n",
    "df_pmid['pmid'].dtypes == df_icite_sample['pmid'].dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Go back and look closely at the df_pmid_enriched output. Check to see if all are NaN. If not, explore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coreproject</th>\n",
       "      <th>pmid</th>\n",
       "      <th>applid</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>year</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>relative_citation_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>883502</th>\n",
       "      <td>R01CA047296</td>\n",
       "      <td>1511878</td>\n",
       "      <td>10650758</td>\n",
       "      <td>10.1016/0378-1119(92)90262-n</td>\n",
       "      <td>A convenient cloning vector containing the GAL...</td>\n",
       "      <td>L Raycroft, G Lozano</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>883508</th>\n",
       "      <td>R01CA053840</td>\n",
       "      <td>1339708</td>\n",
       "      <td>10683285</td>\n",
       "      <td>10.1101/sqb.1992.057.01.012</td>\n",
       "      <td>Protein tyrosine phosphatases: the problems of...</td>\n",
       "      <td>N K Tonks, Q Yang, A J Flint, M F Gebbink, B R...</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>892201</th>\n",
       "      <td>P01CA022443</td>\n",
       "      <td>1457207</td>\n",
       "      <td>10898452</td>\n",
       "      <td>10.1089/aid.1992.8.1611</td>\n",
       "      <td>Studies on the role of the V3 loop in human im...</td>\n",
       "      <td>S H Chiou, E O Freed, A T Panganiban, W R Kenealy</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>892202</th>\n",
       "      <td>P01CA022443</td>\n",
       "      <td>1501283</td>\n",
       "      <td>10898452</td>\n",
       "      <td>10.1128/JVI.66.9.5472-5478.1992</td>\n",
       "      <td>Identification and characterization of fusion ...</td>\n",
       "      <td>E O Freed, D J Myers</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>1.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>892220</th>\n",
       "      <td>P01CA022443</td>\n",
       "      <td>1373201</td>\n",
       "      <td>10898452</td>\n",
       "      <td>10.1128/JVI.66.5.3093-3100.1992</td>\n",
       "      <td>5-Azacytidine and RNA secondary structure incr...</td>\n",
       "      <td>V K Pathak, H M Temin</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>2.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25499140</th>\n",
       "      <td>P30CA014520</td>\n",
       "      <td>37800093</td>\n",
       "      <td>10905065</td>\n",
       "      <td>10.1055/s-0039-3401815</td>\n",
       "      <td>Physician Burnout and Timing of Electronic Hea...</td>\n",
       "      <td>Mark A Micek, Brian Arndt, Wen-Jan Tuan, Betsy...</td>\n",
       "      <td>2020.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25577376</th>\n",
       "      <td>P30CA069533</td>\n",
       "      <td>37840910</td>\n",
       "      <td>10888579</td>\n",
       "      <td>10.15695/jstem/v5i2.12</td>\n",
       "      <td>Approaches for Measuring Inclusive Demographic...</td>\n",
       "      <td>Megan A Mekinda, Sunita Chaudhary, Nathan L Va...</td>\n",
       "      <td>2022.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25583019</th>\n",
       "      <td>P30CA082709</td>\n",
       "      <td>37829495</td>\n",
       "      <td>10695929</td>\n",
       "      <td>10.3389/adar.2022.10400</td>\n",
       "      <td>Sex-Dependent Synaptic Remodeling of the Somat...</td>\n",
       "      <td>Gregory G Grecco, Jui Yen Huang, Braulio Muñoz...</td>\n",
       "      <td>2022.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25598850</th>\n",
       "      <td>P50CA244688</td>\n",
       "      <td>37854304</td>\n",
       "      <td>10814693</td>\n",
       "      <td>10.1016/j.ssci.2022.105763</td>\n",
       "      <td>Methods to improve the translation of evidence...</td>\n",
       "      <td>R J Guerin, R E Glasgow, A Tyler, B A Rabin, A...</td>\n",
       "      <td>2022.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25615455</th>\n",
       "      <td>U54CA254566</td>\n",
       "      <td>37786449</td>\n",
       "      <td>10691298</td>\n",
       "      <td>10.1109/JPROC.2021.3054390</td>\n",
       "      <td>A review of deep learning in medical imaging: ...</td>\n",
       "      <td>S Kevin Zhou, Hayit Greenspan, Christos Davatz...</td>\n",
       "      <td>2021.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.98</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>157761 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          coreproject      pmid    applid                              doi  \\\n",
       "883502    R01CA047296   1511878  10650758     10.1016/0378-1119(92)90262-n   \n",
       "883508    R01CA053840   1339708  10683285      10.1101/sqb.1992.057.01.012   \n",
       "892201    P01CA022443   1457207  10898452          10.1089/aid.1992.8.1611   \n",
       "892202    P01CA022443   1501283  10898452  10.1128/JVI.66.9.5472-5478.1992   \n",
       "892220    P01CA022443   1373201  10898452  10.1128/JVI.66.5.3093-3100.1992   \n",
       "...               ...       ...       ...                              ...   \n",
       "25499140  P30CA014520  37800093  10905065           10.1055/s-0039-3401815   \n",
       "25577376  P30CA069533  37840910  10888579           10.15695/jstem/v5i2.12   \n",
       "25583019  P30CA082709  37829495  10695929          10.3389/adar.2022.10400   \n",
       "25598850  P50CA244688  37854304  10814693       10.1016/j.ssci.2022.105763   \n",
       "25615455  U54CA254566  37786449  10691298       10.1109/JPROC.2021.3054390   \n",
       "\n",
       "                                                      title  \\\n",
       "883502    A convenient cloning vector containing the GAL...   \n",
       "883508    Protein tyrosine phosphatases: the problems of...   \n",
       "892201    Studies on the role of the V3 loop in human im...   \n",
       "892202    Identification and characterization of fusion ...   \n",
       "892220    5-Azacytidine and RNA secondary structure incr...   \n",
       "...                                                     ...   \n",
       "25499140  Physician Burnout and Timing of Electronic Hea...   \n",
       "25577376  Approaches for Measuring Inclusive Demographic...   \n",
       "25583019  Sex-Dependent Synaptic Remodeling of the Somat...   \n",
       "25598850  Methods to improve the translation of evidence...   \n",
       "25615455  A review of deep learning in medical imaging: ...   \n",
       "\n",
       "                                                    authors    year  \\\n",
       "883502                                 L Raycroft, G Lozano  1992.0   \n",
       "883508    N K Tonks, Q Yang, A J Flint, M F Gebbink, B R...  1992.0   \n",
       "892201    S H Chiou, E O Freed, A T Panganiban, W R Kenealy  1992.0   \n",
       "892202                                 E O Freed, D J Myers  1992.0   \n",
       "892220                                V K Pathak, H M Temin  1992.0   \n",
       "...                                                     ...     ...   \n",
       "25499140  Mark A Micek, Brian Arndt, Wen-Jan Tuan, Betsy...  2020.0   \n",
       "25577376  Megan A Mekinda, Sunita Chaudhary, Nathan L Va...  2022.0   \n",
       "25583019  Gregory G Grecco, Jui Yen Huang, Braulio Muñoz...  2022.0   \n",
       "25598850  R J Guerin, R E Glasgow, A Tyler, B A Rabin, A...  2022.0   \n",
       "25615455  S Kevin Zhou, Hayit Greenspan, Christos Davatz...  2021.0   \n",
       "\n",
       "          citation_count  relative_citation_ratio  \n",
       "883502               3.0                     0.05  \n",
       "883508              31.0                     0.55  \n",
       "892201              14.0                     0.30  \n",
       "892202              48.0                     1.13  \n",
       "892220              95.0                     2.14  \n",
       "...                  ...                      ...  \n",
       "25499140             1.0                     0.11  \n",
       "25577376             0.0                     0.00  \n",
       "25583019             1.0                     0.35  \n",
       "25598850             1.0                     0.49  \n",
       "25615455            13.0                     1.98  \n",
       "\n",
       "[157761 rows x 9 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Show rows without any NaN values\n",
    "df_pmid_enriched[~df_pmid_enriched.isnull().any(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "175527"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check number of rows in the input df of pmids with coreproject and applids\n",
    "len(df_pmid)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The goal should be for the output to have a row for each unique pmid. If no iCite data is available for a given pmid, there should still be a row with the pmid and other values listed as NaN.  \n",
    "\n",
    "Find our goal number of rows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "144658"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check number of rows in a list of unique pmids\n",
    "df_pmid.pmid.nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Try again with tweaks to the merging step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# THERE IS AN UPDATED VERSION OF THIS FUNCTION BELOW\n",
    "def enrich_df_with_icite(df_pmid, icite_filepath, cols=None, chunk_size=250000):\n",
    "    \"\"\"\n",
    "    Enrich a DataFrame with iCite data using chunks.\n",
    "\n",
    "    :param df_pmid (pd.DataFrame): DataFrame containing PMID-related data.\n",
    "    :param icite_filepath (str): Path to the zipped iCite CSV file.\n",
    "    :param cols (list): List of columns to select from df_pmid before merging.\n",
    "                   If None, all columns from df_pmid are used.\n",
    "    :param chunk_size (int): Number of rows to read per chunk.\n",
    "\n",
    "    :return: pd.DataFrame: Enriched DataFrame with iCite data.\n",
    "    \"\"\"\n",
    "\n",
    "    # If cols is not specified, use all columns from df_pmid\n",
    "    if cols is None:\n",
    "        cols = df_pmid.columns.tolist()\n",
    "\n",
    "    # Initialize an empty list to store the enriched data chunks\n",
    "    df_enriched_chunks = []\n",
    "\n",
    "    # Create a tqdm wrapper around the generator to track progress\n",
    "    chunks = tqdm(pd.read_csv(icite_filepath, compression='zip', chunksize=chunk_size),\n",
    "                  desc=\"Processing iCite\", unit=\"chunk\")\n",
    "\n",
    "    # Iterate through chunks of the iCite DataFrame\n",
    "    for chunk in chunks:\n",
    "        # Merge only the specified columns from df_pmid where 'pmid' values match\n",
    "        df_merged = pd.merge(df_pmid, chunk[cols], how='left', on='pmid')\n",
    "\n",
    "        # Append the merged DataFrame to the list\n",
    "        df_enriched_chunks.append(df_merged)\n",
    "\n",
    "    # Concatenate all the chunks into the final enriched DataFrame\n",
    "    df_enriched = pd.concat(df_enriched_chunks, ignore_index=True)\n",
    "\n",
    "    return df_enriched\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing iCite: 0chunk [00:00, ?chunk/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing iCite: 143chunk [06:00,  2.48s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite: 144chunk [06:03,  2.44s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite: 145chunk [06:05,  2.40s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (21,22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite: 146chunk [06:07,  2.52s/chunk]\n"
     ]
    }
   ],
   "source": [
    "# Try again\n",
    "df_pmid_enriched = enrich_df_with_icite(df_pmid, icite_filepath, cols, chunk_size=250000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coreproject</th>\n",
       "      <th>pmid</th>\n",
       "      <th>applid</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>year</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>relative_citation_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>36127808</td>\n",
       "      <td>10902170</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29074302</td>\n",
       "      <td>9321971</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>31387361</td>\n",
       "      <td>9321971</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29027980</td>\n",
       "      <td>9321971</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29309429</td>\n",
       "      <td>9321971</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25626937</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>35471840</td>\n",
       "      <td>10690040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25626938</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>35793873</td>\n",
       "      <td>10690040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25626939</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>36509758</td>\n",
       "      <td>10690040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25626940</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>36775354</td>\n",
       "      <td>10690040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25626941</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>37286865</td>\n",
       "      <td>10690040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25626942 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          coreproject      pmid    applid  doi title authors  year  \\\n",
       "0         R01CA239701  36127808  10902170  NaN   NaN     NaN   NaN   \n",
       "1         R21CA209848  29074302   9321971  NaN   NaN     NaN   NaN   \n",
       "2         R21CA209848  31387361   9321971  NaN   NaN     NaN   NaN   \n",
       "3         R21CA209848  29027980   9321971  NaN   NaN     NaN   NaN   \n",
       "4         R21CA209848  29309429   9321971  NaN   NaN     NaN   NaN   \n",
       "...               ...       ...       ...  ...   ...     ...   ...   \n",
       "25626937  P50CA196530  35471840  10690040  NaN   NaN     NaN   NaN   \n",
       "25626938  P50CA196530  35793873  10690040  NaN   NaN     NaN   NaN   \n",
       "25626939  P50CA196530  36509758  10690040  NaN   NaN     NaN   NaN   \n",
       "25626940  P50CA196530  36775354  10690040  NaN   NaN     NaN   NaN   \n",
       "25626941  P50CA196530  37286865  10690040  NaN   NaN     NaN   NaN   \n",
       "\n",
       "          citation_count  relative_citation_ratio  \n",
       "0                    NaN                      NaN  \n",
       "1                    NaN                      NaN  \n",
       "2                    NaN                      NaN  \n",
       "3                    NaN                      NaN  \n",
       "4                    NaN                      NaN  \n",
       "...                  ...                      ...  \n",
       "25626937             NaN                      NaN  \n",
       "25626938             NaN                      NaN  \n",
       "25626939             NaN                      NaN  \n",
       "25626940             NaN                      NaN  \n",
       "25626941             NaN                      NaN  \n",
       "\n",
       "[25626942 rows x 9 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pmid_enriched"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Way too many rows are showing up, and most of them are NaN..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coreproject</th>\n",
       "      <th>pmid</th>\n",
       "      <th>applid</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>year</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>relative_citation_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>883502</th>\n",
       "      <td>R01CA047296</td>\n",
       "      <td>1511878</td>\n",
       "      <td>10650758</td>\n",
       "      <td>10.1016/0378-1119(92)90262-n</td>\n",
       "      <td>A convenient cloning vector containing the GAL...</td>\n",
       "      <td>L Raycroft, G Lozano</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>883508</th>\n",
       "      <td>R01CA053840</td>\n",
       "      <td>1339708</td>\n",
       "      <td>10683285</td>\n",
       "      <td>10.1101/sqb.1992.057.01.012</td>\n",
       "      <td>Protein tyrosine phosphatases: the problems of...</td>\n",
       "      <td>N K Tonks, Q Yang, A J Flint, M F Gebbink, B R...</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>891728</th>\n",
       "      <td>K12HD000849</td>\n",
       "      <td>1325845</td>\n",
       "      <td>10746928</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Changes in Na,K-ATPase gene expression during ...</td>\n",
       "      <td>S K Chambers, M Gilmore-Hebert, B M Kacinski, ...</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>892201</th>\n",
       "      <td>P01CA022443</td>\n",
       "      <td>1457207</td>\n",
       "      <td>10898452</td>\n",
       "      <td>10.1089/aid.1992.8.1611</td>\n",
       "      <td>Studies on the role of the V3 loop in human im...</td>\n",
       "      <td>S H Chiou, E O Freed, A T Panganiban, W R Kenealy</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>892202</th>\n",
       "      <td>P01CA022443</td>\n",
       "      <td>1501283</td>\n",
       "      <td>10898452</td>\n",
       "      <td>10.1128/JVI.66.9.5472-5478.1992</td>\n",
       "      <td>Identification and characterization of fusion ...</td>\n",
       "      <td>E O Freed, D J Myers</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>1.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25625277</th>\n",
       "      <td>P50CA217674</td>\n",
       "      <td>37737674</td>\n",
       "      <td>10687031</td>\n",
       "      <td>10.14309/ajg.0000000000002508</td>\n",
       "      <td>A Study of Dietary Patterns Derived by Cluster...</td>\n",
       "      <td>Xiaotao Zhang, Carrie R Daniel, Valeria Solter...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25625284</th>\n",
       "      <td>P50CA217674</td>\n",
       "      <td>37835569</td>\n",
       "      <td>10687031</td>\n",
       "      <td>10.3390/cancers15194875</td>\n",
       "      <td>The Gut Microbiome as a Biomarker and Therapeu...</td>\n",
       "      <td>Betul Gok Yavuz, Saumil Datar, Shadi Chamseddi...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25625290</th>\n",
       "      <td>U54CA274367</td>\n",
       "      <td>37873404</td>\n",
       "      <td>10697365</td>\n",
       "      <td>10.1101/2023.09.30.560293</td>\n",
       "      <td>A Specialized Epithelial Cell Type Regulating ...</td>\n",
       "      <td>Jia Li, Alan J Simmons, Sophie Chiron, Marisol...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25625295</th>\n",
       "      <td>U54CA274371</td>\n",
       "      <td>37745323</td>\n",
       "      <td>10708199</td>\n",
       "      <td>10.1101/2023.09.17.557982</td>\n",
       "      <td>Digitize your Biology! Modeling multicellular ...</td>\n",
       "      <td>Jeanette A I Johnson, Genevieve L Stein-O'Brie...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25625856</th>\n",
       "      <td>P50CA070907</td>\n",
       "      <td>37816332</td>\n",
       "      <td>10845884</td>\n",
       "      <td>10.1016/j.ccell.2023.09.003</td>\n",
       "      <td>Systematic investigation of mitochondrial tran...</td>\n",
       "      <td>Hongyi Zhang, Xuexin Yu, Jianfeng Ye, Huiyu Li...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175447 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          coreproject      pmid    applid                              doi  \\\n",
       "883502    R01CA047296   1511878  10650758     10.1016/0378-1119(92)90262-n   \n",
       "883508    R01CA053840   1339708  10683285      10.1101/sqb.1992.057.01.012   \n",
       "891728    K12HD000849   1325845  10746928                              NaN   \n",
       "892201    P01CA022443   1457207  10898452          10.1089/aid.1992.8.1611   \n",
       "892202    P01CA022443   1501283  10898452  10.1128/JVI.66.9.5472-5478.1992   \n",
       "...               ...       ...       ...                              ...   \n",
       "25625277  P50CA217674  37737674  10687031    10.14309/ajg.0000000000002508   \n",
       "25625284  P50CA217674  37835569  10687031          10.3390/cancers15194875   \n",
       "25625290  U54CA274367  37873404  10697365        10.1101/2023.09.30.560293   \n",
       "25625295  U54CA274371  37745323  10708199        10.1101/2023.09.17.557982   \n",
       "25625856  P50CA070907  37816332  10845884      10.1016/j.ccell.2023.09.003   \n",
       "\n",
       "                                                      title  \\\n",
       "883502    A convenient cloning vector containing the GAL...   \n",
       "883508    Protein tyrosine phosphatases: the problems of...   \n",
       "891728    Changes in Na,K-ATPase gene expression during ...   \n",
       "892201    Studies on the role of the V3 loop in human im...   \n",
       "892202    Identification and characterization of fusion ...   \n",
       "...                                                     ...   \n",
       "25625277  A Study of Dietary Patterns Derived by Cluster...   \n",
       "25625284  The Gut Microbiome as a Biomarker and Therapeu...   \n",
       "25625290  A Specialized Epithelial Cell Type Regulating ...   \n",
       "25625295  Digitize your Biology! Modeling multicellular ...   \n",
       "25625856  Systematic investigation of mitochondrial tran...   \n",
       "\n",
       "                                                    authors    year  \\\n",
       "883502                                 L Raycroft, G Lozano  1992.0   \n",
       "883508    N K Tonks, Q Yang, A J Flint, M F Gebbink, B R...  1992.0   \n",
       "891728    S K Chambers, M Gilmore-Hebert, B M Kacinski, ...  1992.0   \n",
       "892201    S H Chiou, E O Freed, A T Panganiban, W R Kenealy  1992.0   \n",
       "892202                                 E O Freed, D J Myers  1992.0   \n",
       "...                                                     ...     ...   \n",
       "25625277  Xiaotao Zhang, Carrie R Daniel, Valeria Solter...  2023.0   \n",
       "25625284  Betul Gok Yavuz, Saumil Datar, Shadi Chamseddi...  2023.0   \n",
       "25625290  Jia Li, Alan J Simmons, Sophie Chiron, Marisol...  2023.0   \n",
       "25625295  Jeanette A I Johnson, Genevieve L Stein-O'Brie...  2023.0   \n",
       "25625856  Hongyi Zhang, Xuexin Yu, Jianfeng Ye, Huiyu Li...  2023.0   \n",
       "\n",
       "          citation_count  relative_citation_ratio  \n",
       "883502               3.0                     0.05  \n",
       "883508              31.0                     0.55  \n",
       "891728              11.0                     0.27  \n",
       "892201              14.0                     0.30  \n",
       "892202              48.0                     1.13  \n",
       "...                  ...                      ...  \n",
       "25625277             0.0                      NaN  \n",
       "25625284             0.0                      NaN  \n",
       "25625290             0.0                      NaN  \n",
       "25625295             0.0                      NaN  \n",
       "25625856             0.0                      NaN  \n",
       "\n",
       "[175447 rows x 9 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for rows where at last one of the iCite col values is not NaN\n",
    "df_pmid_enriched.dropna(subset=['doi',\n",
    "                                'title',\n",
    "                                'authors',\n",
    "                                'year',\n",
    "                                'citation_count',\n",
    "                                'relative_citation_ratio'], how='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "coreproject      1771\n",
       "pmid           144658\n",
       "applid           1783\n",
       "dtype: int64"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get number of unique values in all cols in input\n",
    "df_pmid.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coreproject</th>\n",
       "      <th>pmid</th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64337</th>\n",
       "      <td>P30CA016672</td>\n",
       "      <td>29904738</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95152</th>\n",
       "      <td>P30CA051008</td>\n",
       "      <td>9795182</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8660</th>\n",
       "      <td>P30CA006973</td>\n",
       "      <td>19380450</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18894</th>\n",
       "      <td>P30CA008748</td>\n",
       "      <td>35293090</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95155</th>\n",
       "      <td>P30CA051008</td>\n",
       "      <td>9846989</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56815</th>\n",
       "      <td>P30CA016520</td>\n",
       "      <td>23649625</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56816</th>\n",
       "      <td>P30CA016520</td>\n",
       "      <td>23658517</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56817</th>\n",
       "      <td>P30CA016520</td>\n",
       "      <td>23664401</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56818</th>\n",
       "      <td>P30CA016520</td>\n",
       "      <td>23666239</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165987</th>\n",
       "      <td>UM1HG006348</td>\n",
       "      <td>37880260</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>165988 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        coreproject      pmid    0\n",
       "64337   P30CA016672  29904738  292\n",
       "95152   P30CA051008   9795182  292\n",
       "8660    P30CA006973  19380450  292\n",
       "18894   P30CA008748  35293090  292\n",
       "95155   P30CA051008   9846989  292\n",
       "...             ...       ...  ...\n",
       "56815   P30CA016520  23649625  146\n",
       "56816   P30CA016520  23658517  146\n",
       "56817   P30CA016520  23664401  146\n",
       "56818   P30CA016520  23666239  146\n",
       "165987  UM1HG006348  37880260  146\n",
       "\n",
       "[165988 rows x 3 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check to see how many rows have each unique coreproject-pmid combination\n",
    "df_pmid_enriched.groupby(['coreproject','pmid']).size().reset_index().sort_values(by=0, ascending = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Still seeing many empty duplicate rows. The minimum number of duplicates is 146, which is the same as the number of chunks. Must be an issue with duplication during merging."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# THIS FUNCTION IS IMPROVED AND REPLACED BY `get_icite_data_for_pmids` BELOW\n",
    "def enrich_df_with_icite(df_pmid, icite_filepath, cols=None, chunk_size=250000):\n",
    "    \"\"\"\n",
    "    Enrich a DataFrame with iCite data using chunks.\n",
    "\n",
    "    :param df_pmid (pd.DataFrame): DataFrame containing PMID-related data.\n",
    "    :param icite_filepath (str): Path to the zipped iCite CSV file.\n",
    "    :param cols (list): List of columns to select from df_pmid before merging.\n",
    "                   If None, all columns from df_pmid are used.\n",
    "    :param chunk_size (int): Number of rows to read per chunk.\n",
    "\n",
    "    :return: pd.DataFrame: Enriched DataFrame with iCite data.\n",
    "    \"\"\"\n",
    "\n",
    "    # If cols is not specified, use all columns from df_pmid\n",
    "    if cols is None:\n",
    "        cols = df_pmid.columns.tolist()\n",
    "\n",
    "    # Initialize an empty list to store the enriched data chunks\n",
    "    df_enriched_chunks = []\n",
    "\n",
    "    # Create a tqdm wrapper around the generator to track progress\n",
    "    chunks = tqdm(pd.read_csv(icite_filepath, compression='zip', chunksize=chunk_size),\n",
    "                  desc=\"Processing iCite\", unit=\"chunk\")\n",
    "\n",
    "    # Iterate through chunks of the iCite DataFrame\n",
    "    for chunk in chunks:\n",
    "        # Merge only the specified columns from df_pmid where 'pmid' values match\n",
    "        df_merged = pd.merge(df_pmid, chunk[cols], how='left', on='pmid')\n",
    "\n",
    "        # Drop rows where all 'cols' except 'pmid' are blank\n",
    "        df_merged = df_merged.dropna(subset=cols[1:], how='all')\n",
    "\n",
    "        # Append the merged DataFrame to the list\n",
    "        df_enriched_chunks.append(df_merged)\n",
    "\n",
    "    # Concatenate all the chunks into the final enriched DataFrame\n",
    "    df_enriched = pd.concat(df_enriched_chunks, ignore_index=True)\n",
    "\n",
    "    return df_enriched\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing iCite: 143chunk [06:19,  2.50s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite: 144chunk [06:21,  2.51s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite: 145chunk [06:24,  2.46s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (21,22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite: 146chunk [06:25,  2.64s/chunk]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coreproject</th>\n",
       "      <th>pmid</th>\n",
       "      <th>applid</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>year</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>relative_citation_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R01CA047296</td>\n",
       "      <td>1511878</td>\n",
       "      <td>10650758</td>\n",
       "      <td>10.1016/0378-1119(92)90262-n</td>\n",
       "      <td>A convenient cloning vector containing the GAL...</td>\n",
       "      <td>L Raycroft, G Lozano</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R01CA053840</td>\n",
       "      <td>1339708</td>\n",
       "      <td>10683285</td>\n",
       "      <td>10.1101/sqb.1992.057.01.012</td>\n",
       "      <td>Protein tyrosine phosphatases: the problems of...</td>\n",
       "      <td>N K Tonks, Q Yang, A J Flint, M F Gebbink, B R...</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K12HD000849</td>\n",
       "      <td>1325845</td>\n",
       "      <td>10746928</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Changes in Na,K-ATPase gene expression during ...</td>\n",
       "      <td>S K Chambers, M Gilmore-Hebert, B M Kacinski, ...</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P01CA022443</td>\n",
       "      <td>1457207</td>\n",
       "      <td>10898452</td>\n",
       "      <td>10.1089/aid.1992.8.1611</td>\n",
       "      <td>Studies on the role of the V3 loop in human im...</td>\n",
       "      <td>S H Chiou, E O Freed, A T Panganiban, W R Kenealy</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P01CA022443</td>\n",
       "      <td>1501283</td>\n",
       "      <td>10898452</td>\n",
       "      <td>10.1128/JVI.66.9.5472-5478.1992</td>\n",
       "      <td>Identification and characterization of fusion ...</td>\n",
       "      <td>E O Freed, D J Myers</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>1.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175442</th>\n",
       "      <td>P50CA217674</td>\n",
       "      <td>37737674</td>\n",
       "      <td>10687031</td>\n",
       "      <td>10.14309/ajg.0000000000002508</td>\n",
       "      <td>A Study of Dietary Patterns Derived by Cluster...</td>\n",
       "      <td>Xiaotao Zhang, Carrie R Daniel, Valeria Solter...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175443</th>\n",
       "      <td>P50CA217674</td>\n",
       "      <td>37835569</td>\n",
       "      <td>10687031</td>\n",
       "      <td>10.3390/cancers15194875</td>\n",
       "      <td>The Gut Microbiome as a Biomarker and Therapeu...</td>\n",
       "      <td>Betul Gok Yavuz, Saumil Datar, Shadi Chamseddi...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175444</th>\n",
       "      <td>U54CA274367</td>\n",
       "      <td>37873404</td>\n",
       "      <td>10697365</td>\n",
       "      <td>10.1101/2023.09.30.560293</td>\n",
       "      <td>A Specialized Epithelial Cell Type Regulating ...</td>\n",
       "      <td>Jia Li, Alan J Simmons, Sophie Chiron, Marisol...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175445</th>\n",
       "      <td>U54CA274371</td>\n",
       "      <td>37745323</td>\n",
       "      <td>10708199</td>\n",
       "      <td>10.1101/2023.09.17.557982</td>\n",
       "      <td>Digitize your Biology! Modeling multicellular ...</td>\n",
       "      <td>Jeanette A I Johnson, Genevieve L Stein-O'Brie...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175446</th>\n",
       "      <td>P50CA070907</td>\n",
       "      <td>37816332</td>\n",
       "      <td>10845884</td>\n",
       "      <td>10.1016/j.ccell.2023.09.003</td>\n",
       "      <td>Systematic investigation of mitochondrial tran...</td>\n",
       "      <td>Hongyi Zhang, Xuexin Yu, Jianfeng Ye, Huiyu Li...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175447 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        coreproject      pmid    applid                              doi  \\\n",
       "0       R01CA047296   1511878  10650758     10.1016/0378-1119(92)90262-n   \n",
       "1       R01CA053840   1339708  10683285      10.1101/sqb.1992.057.01.012   \n",
       "2       K12HD000849   1325845  10746928                              NaN   \n",
       "3       P01CA022443   1457207  10898452          10.1089/aid.1992.8.1611   \n",
       "4       P01CA022443   1501283  10898452  10.1128/JVI.66.9.5472-5478.1992   \n",
       "...             ...       ...       ...                              ...   \n",
       "175442  P50CA217674  37737674  10687031    10.14309/ajg.0000000000002508   \n",
       "175443  P50CA217674  37835569  10687031          10.3390/cancers15194875   \n",
       "175444  U54CA274367  37873404  10697365        10.1101/2023.09.30.560293   \n",
       "175445  U54CA274371  37745323  10708199        10.1101/2023.09.17.557982   \n",
       "175446  P50CA070907  37816332  10845884      10.1016/j.ccell.2023.09.003   \n",
       "\n",
       "                                                    title  \\\n",
       "0       A convenient cloning vector containing the GAL...   \n",
       "1       Protein tyrosine phosphatases: the problems of...   \n",
       "2       Changes in Na,K-ATPase gene expression during ...   \n",
       "3       Studies on the role of the V3 loop in human im...   \n",
       "4       Identification and characterization of fusion ...   \n",
       "...                                                   ...   \n",
       "175442  A Study of Dietary Patterns Derived by Cluster...   \n",
       "175443  The Gut Microbiome as a Biomarker and Therapeu...   \n",
       "175444  A Specialized Epithelial Cell Type Regulating ...   \n",
       "175445  Digitize your Biology! Modeling multicellular ...   \n",
       "175446  Systematic investigation of mitochondrial tran...   \n",
       "\n",
       "                                                  authors    year  \\\n",
       "0                                    L Raycroft, G Lozano  1992.0   \n",
       "1       N K Tonks, Q Yang, A J Flint, M F Gebbink, B R...  1992.0   \n",
       "2       S K Chambers, M Gilmore-Hebert, B M Kacinski, ...  1992.0   \n",
       "3       S H Chiou, E O Freed, A T Panganiban, W R Kenealy  1992.0   \n",
       "4                                    E O Freed, D J Myers  1992.0   \n",
       "...                                                   ...     ...   \n",
       "175442  Xiaotao Zhang, Carrie R Daniel, Valeria Solter...  2023.0   \n",
       "175443  Betul Gok Yavuz, Saumil Datar, Shadi Chamseddi...  2023.0   \n",
       "175444  Jia Li, Alan J Simmons, Sophie Chiron, Marisol...  2023.0   \n",
       "175445  Jeanette A I Johnson, Genevieve L Stein-O'Brie...  2023.0   \n",
       "175446  Hongyi Zhang, Xuexin Yu, Jianfeng Ye, Huiyu Li...  2023.0   \n",
       "\n",
       "        citation_count  relative_citation_ratio  \n",
       "0                  3.0                     0.05  \n",
       "1                 31.0                     0.55  \n",
       "2                 11.0                     0.27  \n",
       "3                 14.0                     0.30  \n",
       "4                 48.0                     1.13  \n",
       "...                ...                      ...  \n",
       "175442             0.0                      NaN  \n",
       "175443             0.0                      NaN  \n",
       "175444             0.0                      NaN  \n",
       "175445             0.0                      NaN  \n",
       "175446             0.0                      NaN  \n",
       "\n",
       "[175447 rows x 9 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Try again\n",
    "df_pmid_enriched = enrich_df_with_icite(df_pmid, icite_filepath, cols, chunk_size=250000)\n",
    "df_pmid_enriched"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This looks better, but still misses any pmid rows without any icite data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coreproject</th>\n",
       "      <th>pmid</th>\n",
       "      <th>applid</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>year</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>relative_citation_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>36127808</td>\n",
       "      <td>10902170</td>\n",
       "      <td>10.1002/cam4.5266</td>\n",
       "      <td>Genetic ancestry, differential gene expression...</td>\n",
       "      <td>Freddy A Barragan, Lauren J Mills, Andrew R Ra...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29074302</td>\n",
       "      <td>9321971</td>\n",
       "      <td>10.1016/j.imbio.2017.10.028</td>\n",
       "      <td>Endogenous antibody responses to mucin 1 in a ...</td>\n",
       "      <td>Janardan P Pandey, Aryan M Namboodiri, Bethany...</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>31387361</td>\n",
       "      <td>9321971</td>\n",
       "      <td>10.1161/CIRCULATIONAHA.119.038376</td>\n",
       "      <td>Defects in the Exocyst-Cilia Machinery Cause B...</td>\n",
       "      <td>Diana Fulmer, Katelynn Toomer, Lilong Guo, Kel...</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>1.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29027980</td>\n",
       "      <td>9321971</td>\n",
       "      <td>10.3390/genes8100269</td>\n",
       "      <td>The Plasticizer Bisphenol A Perturbs the Hepat...</td>\n",
       "      <td>Ludivine Renaud, Willian A da Silveira, E Star...</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29309429</td>\n",
       "      <td>9321971</td>\n",
       "      <td>10.1371/journal.pone.0190949</td>\n",
       "      <td>ShinyGPA: An interactive visualization toolkit...</td>\n",
       "      <td>Emma Kortemeier, Paula S Ramos, Kelly J Hunt, ...</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194594</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>35471840</td>\n",
       "      <td>10690040</td>\n",
       "      <td>10.1146/annurev-immunol-070621-030155</td>\n",
       "      <td>Resistance Mechanisms to Anti-PD Cancer Immuno...</td>\n",
       "      <td>Matthew D Vesely, Tianxiang Zhang, Lieping Chen</td>\n",
       "      <td>2022.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>12.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194595</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>35793873</td>\n",
       "      <td>10690040</td>\n",
       "      <td>10.1136/jitc-2022-005025</td>\n",
       "      <td>Quantitative tissue analysis and role of myelo...</td>\n",
       "      <td>Brian S Henick, Franz Villarroel-Espindola, Il...</td>\n",
       "      <td>2022.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194596</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>36509758</td>\n",
       "      <td>10690040</td>\n",
       "      <td>10.1038/s41467-022-34889-z</td>\n",
       "      <td>Brain metastatic outgrowth and osimertinib res...</td>\n",
       "      <td>Sally J Adua, Anna Arnal-Estapé, Minghui Zhao,...</td>\n",
       "      <td>2022.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194597</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>36775354</td>\n",
       "      <td>10690040</td>\n",
       "      <td>10.1038/s41374-022-00796-6</td>\n",
       "      <td>Quantitative assessment of Siglec-15 expressio...</td>\n",
       "      <td>Saba Shafi, Thazin Nwe Aung, Vasiliki Xirou, N...</td>\n",
       "      <td>2022.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194598</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>37286865</td>\n",
       "      <td>10690040</td>\n",
       "      <td>10.1007/s10555-023-10109-y</td>\n",
       "      <td>Epigenetic markers and therapeutic targets for...</td>\n",
       "      <td>Carolyn J Kravitz, Qin Yan, Don X Nguyen</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>194599 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        coreproject      pmid    applid  \\\n",
       "0       R01CA239701  36127808  10902170   \n",
       "1       R21CA209848  29074302   9321971   \n",
       "2       R21CA209848  31387361   9321971   \n",
       "3       R21CA209848  29027980   9321971   \n",
       "4       R21CA209848  29309429   9321971   \n",
       "...             ...       ...       ...   \n",
       "194594  P50CA196530  35471840  10690040   \n",
       "194595  P50CA196530  35793873  10690040   \n",
       "194596  P50CA196530  36509758  10690040   \n",
       "194597  P50CA196530  36775354  10690040   \n",
       "194598  P50CA196530  37286865  10690040   \n",
       "\n",
       "                                          doi  \\\n",
       "0                           10.1002/cam4.5266   \n",
       "1                 10.1016/j.imbio.2017.10.028   \n",
       "2           10.1161/CIRCULATIONAHA.119.038376   \n",
       "3                        10.3390/genes8100269   \n",
       "4                10.1371/journal.pone.0190949   \n",
       "...                                       ...   \n",
       "194594  10.1146/annurev-immunol-070621-030155   \n",
       "194595               10.1136/jitc-2022-005025   \n",
       "194596             10.1038/s41467-022-34889-z   \n",
       "194597             10.1038/s41374-022-00796-6   \n",
       "194598             10.1007/s10555-023-10109-y   \n",
       "\n",
       "                                                    title  \\\n",
       "0       Genetic ancestry, differential gene expression...   \n",
       "1       Endogenous antibody responses to mucin 1 in a ...   \n",
       "2       Defects in the Exocyst-Cilia Machinery Cause B...   \n",
       "3       The Plasticizer Bisphenol A Perturbs the Hepat...   \n",
       "4       ShinyGPA: An interactive visualization toolkit...   \n",
       "...                                                   ...   \n",
       "194594  Resistance Mechanisms to Anti-PD Cancer Immuno...   \n",
       "194595  Quantitative tissue analysis and role of myelo...   \n",
       "194596  Brain metastatic outgrowth and osimertinib res...   \n",
       "194597  Quantitative assessment of Siglec-15 expressio...   \n",
       "194598  Epigenetic markers and therapeutic targets for...   \n",
       "\n",
       "                                                  authors    year  \\\n",
       "0       Freddy A Barragan, Lauren J Mills, Andrew R Ra...  2023.0   \n",
       "1       Janardan P Pandey, Aryan M Namboodiri, Bethany...  2018.0   \n",
       "2       Diana Fulmer, Katelynn Toomer, Lilong Guo, Kel...  2019.0   \n",
       "3       Ludivine Renaud, Willian A da Silveira, E Star...  2017.0   \n",
       "4       Emma Kortemeier, Paula S Ramos, Kelly J Hunt, ...  2018.0   \n",
       "...                                                   ...     ...   \n",
       "194594    Matthew D Vesely, Tianxiang Zhang, Lieping Chen  2022.0   \n",
       "194595  Brian S Henick, Franz Villarroel-Espindola, Il...  2022.0   \n",
       "194596  Sally J Adua, Anna Arnal-Estapé, Minghui Zhao,...  2022.0   \n",
       "194597  Saba Shafi, Thazin Nwe Aung, Vasiliki Xirou, N...  2022.0   \n",
       "194598           Carolyn J Kravitz, Qin Yan, Don X Nguyen  2023.0   \n",
       "\n",
       "        citation_count  relative_citation_ratio  \n",
       "0                  4.0                      NaN  \n",
       "1                  2.0                     0.09  \n",
       "2                 34.0                     1.95  \n",
       "3                 20.0                     1.17  \n",
       "4                  2.0                     0.08  \n",
       "...                ...                      ...  \n",
       "194594            71.0                    12.39  \n",
       "194595             1.0                     0.16  \n",
       "194596             1.0                     0.19  \n",
       "194597             3.0                     0.43  \n",
       "194598             0.0                      NaN  \n",
       "\n",
       "[194599 rows x 9 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge enriched df back to the df_pmid to fill the blanks\n",
    "df_pmid_icite = pd.merge(df_pmid, df_pmid_enriched, how='left', on=['coreproject','pmid','applid'])\n",
    "df_pmid_icite"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Too many rows appear when the original df_pmid is merged back with the enriched..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coreproject</th>\n",
       "      <th>pmid</th>\n",
       "      <th>applid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R01CA239701</td>\n",
       "      <td>36127808</td>\n",
       "      <td>10902170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29074302</td>\n",
       "      <td>9321971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>31387361</td>\n",
       "      <td>9321971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29027980</td>\n",
       "      <td>9321971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R21CA209848</td>\n",
       "      <td>29309429</td>\n",
       "      <td>9321971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175522</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>35471840</td>\n",
       "      <td>10690040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175523</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>35793873</td>\n",
       "      <td>10690040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175524</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>36509758</td>\n",
       "      <td>10690040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175525</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>36775354</td>\n",
       "      <td>10690040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175526</th>\n",
       "      <td>P50CA196530</td>\n",
       "      <td>37286865</td>\n",
       "      <td>10690040</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175527 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        coreproject      pmid    applid\n",
       "0       R01CA239701  36127808  10902170\n",
       "1       R21CA209848  29074302   9321971\n",
       "2       R21CA209848  31387361   9321971\n",
       "3       R21CA209848  29027980   9321971\n",
       "4       R21CA209848  29309429   9321971\n",
       "...             ...       ...       ...\n",
       "175522  P50CA196530  35471840  10690040\n",
       "175523  P50CA196530  35793873  10690040\n",
       "175524  P50CA196530  36509758  10690040\n",
       "175525  P50CA196530  36775354  10690040\n",
       "175526  P50CA196530  37286865  10690040\n",
       "\n",
       "[175527 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pmid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coreproject</th>\n",
       "      <th>pmid</th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64337</th>\n",
       "      <td>P30CA016672</td>\n",
       "      <td>29904738</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95152</th>\n",
       "      <td>P30CA051008</td>\n",
       "      <td>9795182</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8660</th>\n",
       "      <td>P30CA006973</td>\n",
       "      <td>19380450</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18894</th>\n",
       "      <td>P30CA008748</td>\n",
       "      <td>35293090</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95155</th>\n",
       "      <td>P30CA051008</td>\n",
       "      <td>9846989</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56815</th>\n",
       "      <td>P30CA016520</td>\n",
       "      <td>23649625</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56816</th>\n",
       "      <td>P30CA016520</td>\n",
       "      <td>23658517</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56817</th>\n",
       "      <td>P30CA016520</td>\n",
       "      <td>23664401</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56818</th>\n",
       "      <td>P30CA016520</td>\n",
       "      <td>23666239</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165987</th>\n",
       "      <td>UM1HG006348</td>\n",
       "      <td>37880260</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>165988 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        coreproject      pmid  0\n",
       "64337   P30CA016672  29904738  2\n",
       "95152   P30CA051008   9795182  2\n",
       "8660    P30CA006973  19380450  2\n",
       "18894   P30CA008748  35293090  2\n",
       "95155   P30CA051008   9846989  2\n",
       "...             ...       ... ..\n",
       "56815   P30CA016520  23649625  1\n",
       "56816   P30CA016520  23658517  1\n",
       "56817   P30CA016520  23664401  1\n",
       "56818   P30CA016520  23666239  1\n",
       "165987  UM1HG006348  37880260  1\n",
       "\n",
       "[165988 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for duplicates of the same coreproject-pmid combination \n",
    "df_pmid.groupby(['coreproject','pmid']).size().reset_index().sort_values(by=0, ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I'm still getting odd numbers and unintuitive counts when trying to merge the original pmid list with the data pulled from iCite. This may be due to the extra 'coreproject' and 'applid' columns in df_pmid causing duplications or other problems during the merge.  \n",
    "\n",
    "Try a new approach focused on pulling unique pmids and adding icite data for each. The output will not contain coreproject or applid. The coreproject column can be joined back in at the end using pmid."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# THERE IS AN UDPATED VERSION OF THIS FUNCTION BELOW\n",
    "def get_icite_data_for_pmids(df_pmid, icite_filepath, cols, \n",
    "                             chunk_size=250000, chunk_count_est=None):\n",
    "    \"\"\"\n",
    "    Get iCite data for unique PMIDs from a DataFrame using chunks.\n",
    "\n",
    "    :param df_pmid (pd.DataFrame): DataFrame containing PMID-related data.\n",
    "    :param icite_filepath (str): Path to the zipped iCite CSV file.\n",
    "    :param cols (list): Columns to pull from iCite and include in output df\n",
    "    :param chunk_size (int): Number of rows to read per chunk.\n",
    "    :param chunk_count (int): Estimated number of chunks. If None, ignored\n",
    "\n",
    "    :return: pd.DataFrame: DataFrame with specified columns\n",
    "    \"\"\"\n",
    "    # Get unique PMIDs from df_pmid\n",
    "    unique_pmids = df_pmid['pmid'].unique()\n",
    "\n",
    "    # Initialize an empty list to store the enriched data chunks\n",
    "    df_enriched_chunks = []\n",
    "\n",
    "    # Create a tqdm wrapper around the generator to track progress\n",
    "    chunks = tqdm(pd.read_csv(icite_filepath, compression='zip', chunksize=chunk_size),\n",
    "                  desc=\"Processing iCite\", unit=\"chunk\", total=chunk_count_est)\n",
    "\n",
    "    # Iterate through chunks of the iCite DataFrame\n",
    "    for chunk in chunks:\n",
    "        # Filter the chunk to include only rows with PMIDs in unique_pmids\n",
    "        chunk_filtered = chunk[chunk['pmid'].isin(unique_pmids)]\n",
    "\n",
    "        # Append the filtered DataFrame to the list with selected columns\n",
    "        df_enriched_chunks.append(chunk_filtered[cols])\n",
    "\n",
    "    # Concatenate all the chunks into the final enriched DataFrame\n",
    "    df_enriched = pd.concat(df_enriched_chunks, ignore_index=True)\n",
    "\n",
    "    return df_enriched\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing iCite: 0chunk [00:00, ?chunk/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing iCite: 143chunk [05:49,  2.58s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite: 144chunk [05:51,  2.51s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite: 145chunk [05:54,  2.43s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (21,22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite: 146chunk [05:55,  2.44s/chunk]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>year</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>relative_citation_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1279509</td>\n",
       "      <td>10.1203/00006450-199210000-00018</td>\n",
       "      <td>Expression and regulation of L-selectin on eos...</td>\n",
       "      <td>J B Smith, R D Kunjummen, T K Kishimoto, D C A...</td>\n",
       "      <td>1992</td>\n",
       "      <td>25</td>\n",
       "      <td>0.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1280555</td>\n",
       "      <td>10.1002/cyto.990130707</td>\n",
       "      <td>Streptavidin-based quantitative staining of in...</td>\n",
       "      <td>P Srivastava, T L Sladek, M N Goodman, J W Jac...</td>\n",
       "      <td>1992</td>\n",
       "      <td>14</td>\n",
       "      <td>0.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1281066</td>\n",
       "      <td>10.1002/cyto.990130808</td>\n",
       "      <td>Reticulocyte quantification by flow cytometry,...</td>\n",
       "      <td>K J Schimenti, K Lacerna, A Wamble, L Maston, ...</td>\n",
       "      <td>1992</td>\n",
       "      <td>38</td>\n",
       "      <td>1.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1282437</td>\n",
       "      <td>10.1101/gr.2.2.137</td>\n",
       "      <td>Development of a sensitive reverse transcripta...</td>\n",
       "      <td>S S Tan, J H Weis</td>\n",
       "      <td>1992</td>\n",
       "      <td>49</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1283327</td>\n",
       "      <td>10.1002/gcc.2870050414</td>\n",
       "      <td>Sublocalization of the chromosome 5 breakpoint...</td>\n",
       "      <td>S W Morris, J T Foust, M B Valentine, W M Robe...</td>\n",
       "      <td>1992</td>\n",
       "      <td>14</td>\n",
       "      <td>0.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144591</th>\n",
       "      <td>37928187</td>\n",
       "      <td>10.3389/fcimb.2023.1270935</td>\n",
       "      <td>Co-infection and co-localization of Kaposi sar...</td>\n",
       "      <td>Peter Julius, Guobin Kang, Stepfanie Siyumbwa,...</td>\n",
       "      <td>2023</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144592</th>\n",
       "      <td>37928542</td>\n",
       "      <td>10.3389/fimmu.2023.1236514</td>\n",
       "      <td>Features of the TCR repertoire associate with ...</td>\n",
       "      <td>Mateusz Pospiech, Mukund Tamizharasan, Yu-Chun...</td>\n",
       "      <td>2023</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144593</th>\n",
       "      <td>37928819</td>\n",
       "      <td>10.1007/978-3-031-33842-7_6</td>\n",
       "      <td>Leveraging 2D Deep Learning ImageNet-trained m...</td>\n",
       "      <td>Bhakti Baheti, Sarthak Pati, Bjoern Menze, Spy...</td>\n",
       "      <td>2023</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144594</th>\n",
       "      <td>37930127</td>\n",
       "      <td>10.1097/HC9.0000000000000315</td>\n",
       "      <td>Patient-reported symptoms and interest in symp...</td>\n",
       "      <td>Andrew M Moon, Sarah Cook, Rachel M Swier, Han...</td>\n",
       "      <td>2023</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144595</th>\n",
       "      <td>37930454</td>\n",
       "      <td>10.1186/s41181-023-00223-1</td>\n",
       "      <td>&lt;sup&gt;89&lt;/sup&gt;Zr-leukocyte labelling for cell t...</td>\n",
       "      <td>Maryke Kahts, Hua Guo, Harikrishna Kommidi, Ya...</td>\n",
       "      <td>2023</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>144596 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            pmid                               doi  \\\n",
       "0        1279509  10.1203/00006450-199210000-00018   \n",
       "1        1280555            10.1002/cyto.990130707   \n",
       "2        1281066            10.1002/cyto.990130808   \n",
       "3        1282437                10.1101/gr.2.2.137   \n",
       "4        1283327            10.1002/gcc.2870050414   \n",
       "...          ...                               ...   \n",
       "144591  37928187        10.3389/fcimb.2023.1270935   \n",
       "144592  37928542        10.3389/fimmu.2023.1236514   \n",
       "144593  37928819       10.1007/978-3-031-33842-7_6   \n",
       "144594  37930127      10.1097/HC9.0000000000000315   \n",
       "144595  37930454        10.1186/s41181-023-00223-1   \n",
       "\n",
       "                                                    title  \\\n",
       "0       Expression and regulation of L-selectin on eos...   \n",
       "1       Streptavidin-based quantitative staining of in...   \n",
       "2       Reticulocyte quantification by flow cytometry,...   \n",
       "3       Development of a sensitive reverse transcripta...   \n",
       "4       Sublocalization of the chromosome 5 breakpoint...   \n",
       "...                                                   ...   \n",
       "144591  Co-infection and co-localization of Kaposi sar...   \n",
       "144592  Features of the TCR repertoire associate with ...   \n",
       "144593  Leveraging 2D Deep Learning ImageNet-trained m...   \n",
       "144594  Patient-reported symptoms and interest in symp...   \n",
       "144595  <sup>89</sup>Zr-leukocyte labelling for cell t...   \n",
       "\n",
       "                                                  authors  year  \\\n",
       "0       J B Smith, R D Kunjummen, T K Kishimoto, D C A...  1992   \n",
       "1       P Srivastava, T L Sladek, M N Goodman, J W Jac...  1992   \n",
       "2       K J Schimenti, K Lacerna, A Wamble, L Maston, ...  1992   \n",
       "3                                       S S Tan, J H Weis  1992   \n",
       "4       S W Morris, J T Foust, M B Valentine, W M Robe...  1992   \n",
       "...                                                   ...   ...   \n",
       "144591  Peter Julius, Guobin Kang, Stepfanie Siyumbwa,...  2023   \n",
       "144592  Mateusz Pospiech, Mukund Tamizharasan, Yu-Chun...  2023   \n",
       "144593  Bhakti Baheti, Sarthak Pati, Bjoern Menze, Spy...  2023   \n",
       "144594  Andrew M Moon, Sarah Cook, Rachel M Swier, Han...  2023   \n",
       "144595  Maryke Kahts, Hua Guo, Harikrishna Kommidi, Ya...  2023   \n",
       "\n",
       "        citation_count  relative_citation_ratio  \n",
       "0                   25                     0.67  \n",
       "1                   14                     0.43  \n",
       "2                   38                     1.45  \n",
       "3                   49                     1.00  \n",
       "4                   14                     0.29  \n",
       "...                ...                      ...  \n",
       "144591               0                      NaN  \n",
       "144592               0                      NaN  \n",
       "144593               0                      NaN  \n",
       "144594               0                      NaN  \n",
       "144595               0                      NaN  \n",
       "\n",
       "[144596 rows x 7 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Try the new function that doesn't include coreproject or applid in the output\n",
    "df_pmid_icite = get_icite_data_for_pmids(df_pmid, icite_filepath, cols)\n",
    "df_pmid_icite"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looks good at a glance. Runtime is about 6min"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "144658"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the expected row count again\n",
    "df_pmid.pmid.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1279509</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96401</th>\n",
       "      <td>30131384</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96395</th>\n",
       "      <td>30130351</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96396</th>\n",
       "      <td>30130469</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96397</th>\n",
       "      <td>30130544</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48191</th>\n",
       "      <td>23153072</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48190</th>\n",
       "      <td>23152935</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48189</th>\n",
       "      <td>23152800</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48188</th>\n",
       "      <td>23152566</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144595</th>\n",
       "      <td>37930454</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>144596 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            pmid  0\n",
       "0        1279509  1\n",
       "96401   30131384  1\n",
       "96395   30130351  1\n",
       "96396   30130469  1\n",
       "96397   30130544  1\n",
       "...          ... ..\n",
       "48191   23153072  1\n",
       "48190   23152935  1\n",
       "48189   23152800  1\n",
       "48188   23152566  1\n",
       "144595  37930454  1\n",
       "\n",
       "[144596 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for duplicate pmids in the icite-enriched df\n",
    "df_pmid_icite.groupby('pmid').size().reset_index().sort_values(by=0,ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36127808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29074302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31387361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29027980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>29309429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144653</th>\n",
       "      <td>28651562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144654</th>\n",
       "      <td>31887619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144655</th>\n",
       "      <td>32251621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144656</th>\n",
       "      <td>33148678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144657</th>\n",
       "      <td>36509758</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>144658 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            pmid\n",
       "0       36127808\n",
       "1       29074302\n",
       "2       31387361\n",
       "3       29027980\n",
       "4       29309429\n",
       "...          ...\n",
       "144653  28651562\n",
       "144654  31887619\n",
       "144655  32251621\n",
       "144656  33148678\n",
       "144657  36509758\n",
       "\n",
       "[144658 rows x 1 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Try approach where unique pmids are a 1-column df instead of a list\n",
    "df_unique_pmids = pd.DataFrame(df_pmid['pmid'].unique().tolist(), columns=['pmid'])\n",
    "df_unique_pmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>year</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>relative_citation_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36127808</td>\n",
       "      <td>10.1002/cam4.5266</td>\n",
       "      <td>Genetic ancestry, differential gene expression...</td>\n",
       "      <td>Freddy A Barragan, Lauren J Mills, Andrew R Ra...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29074302</td>\n",
       "      <td>10.1016/j.imbio.2017.10.028</td>\n",
       "      <td>Endogenous antibody responses to mucin 1 in a ...</td>\n",
       "      <td>Janardan P Pandey, Aryan M Namboodiri, Bethany...</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31387361</td>\n",
       "      <td>10.1161/CIRCULATIONAHA.119.038376</td>\n",
       "      <td>Defects in the Exocyst-Cilia Machinery Cause B...</td>\n",
       "      <td>Diana Fulmer, Katelynn Toomer, Lilong Guo, Kel...</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>1.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29027980</td>\n",
       "      <td>10.3390/genes8100269</td>\n",
       "      <td>The Plasticizer Bisphenol A Perturbs the Hepat...</td>\n",
       "      <td>Ludivine Renaud, Willian A da Silveira, E Star...</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>29309429</td>\n",
       "      <td>10.1371/journal.pone.0190949</td>\n",
       "      <td>ShinyGPA: An interactive visualization toolkit...</td>\n",
       "      <td>Emma Kortemeier, Paula S Ramos, Kelly J Hunt, ...</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144653</th>\n",
       "      <td>28651562</td>\n",
       "      <td>10.1186/s12859-017-1711-z</td>\n",
       "      <td>GRAPE: a pathway template method to characteri...</td>\n",
       "      <td>Michael I Klein, David F Stern, Hongyu Zhao</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144654</th>\n",
       "      <td>31887619</td>\n",
       "      <td>10.1016/j.oraloncology.2019.104554</td>\n",
       "      <td>Transfer RNA methyltransferase gene NSUN2 mRNA...</td>\n",
       "      <td>Lingeng Lu, Stephen G Gaffney, Vincent L Canna...</td>\n",
       "      <td>2020.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144655</th>\n",
       "      <td>32251621</td>\n",
       "      <td>10.1016/S1470-2045(20)30111-X</td>\n",
       "      <td>Pembrolizumab for management of patients with ...</td>\n",
       "      <td>Sarah B Goldberg, Kurt A Schalper, Scott N Get...</td>\n",
       "      <td>2020.0</td>\n",
       "      <td>276.0</td>\n",
       "      <td>18.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144656</th>\n",
       "      <td>33148678</td>\n",
       "      <td>10.1158/1940-6207.CAPR-20-0394</td>\n",
       "      <td>Clearing the Haze: What Do We Still Need to Le...</td>\n",
       "      <td>Lisa M Fucito, Hannah Malinosky, Stephen R Bal...</td>\n",
       "      <td>2021.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144657</th>\n",
       "      <td>36509758</td>\n",
       "      <td>10.1038/s41467-022-34889-z</td>\n",
       "      <td>Brain metastatic outgrowth and osimertinib res...</td>\n",
       "      <td>Sally J Adua, Anna Arnal-Estapé, Minghui Zhao,...</td>\n",
       "      <td>2022.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>144658 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            pmid                                 doi  \\\n",
       "0       36127808                   10.1002/cam4.5266   \n",
       "1       29074302         10.1016/j.imbio.2017.10.028   \n",
       "2       31387361   10.1161/CIRCULATIONAHA.119.038376   \n",
       "3       29027980                10.3390/genes8100269   \n",
       "4       29309429        10.1371/journal.pone.0190949   \n",
       "...          ...                                 ...   \n",
       "144653  28651562           10.1186/s12859-017-1711-z   \n",
       "144654  31887619  10.1016/j.oraloncology.2019.104554   \n",
       "144655  32251621       10.1016/S1470-2045(20)30111-X   \n",
       "144656  33148678      10.1158/1940-6207.CAPR-20-0394   \n",
       "144657  36509758          10.1038/s41467-022-34889-z   \n",
       "\n",
       "                                                    title  \\\n",
       "0       Genetic ancestry, differential gene expression...   \n",
       "1       Endogenous antibody responses to mucin 1 in a ...   \n",
       "2       Defects in the Exocyst-Cilia Machinery Cause B...   \n",
       "3       The Plasticizer Bisphenol A Perturbs the Hepat...   \n",
       "4       ShinyGPA: An interactive visualization toolkit...   \n",
       "...                                                   ...   \n",
       "144653  GRAPE: a pathway template method to characteri...   \n",
       "144654  Transfer RNA methyltransferase gene NSUN2 mRNA...   \n",
       "144655  Pembrolizumab for management of patients with ...   \n",
       "144656  Clearing the Haze: What Do We Still Need to Le...   \n",
       "144657  Brain metastatic outgrowth and osimertinib res...   \n",
       "\n",
       "                                                  authors    year  \\\n",
       "0       Freddy A Barragan, Lauren J Mills, Andrew R Ra...  2023.0   \n",
       "1       Janardan P Pandey, Aryan M Namboodiri, Bethany...  2018.0   \n",
       "2       Diana Fulmer, Katelynn Toomer, Lilong Guo, Kel...  2019.0   \n",
       "3       Ludivine Renaud, Willian A da Silveira, E Star...  2017.0   \n",
       "4       Emma Kortemeier, Paula S Ramos, Kelly J Hunt, ...  2018.0   \n",
       "...                                                   ...     ...   \n",
       "144653        Michael I Klein, David F Stern, Hongyu Zhao  2017.0   \n",
       "144654  Lingeng Lu, Stephen G Gaffney, Vincent L Canna...  2020.0   \n",
       "144655  Sarah B Goldberg, Kurt A Schalper, Scott N Get...  2020.0   \n",
       "144656  Lisa M Fucito, Hannah Malinosky, Stephen R Bal...  2021.0   \n",
       "144657  Sally J Adua, Anna Arnal-Estapé, Minghui Zhao,...  2022.0   \n",
       "\n",
       "        citation_count  relative_citation_ratio  \n",
       "0                  4.0                      NaN  \n",
       "1                  2.0                     0.09  \n",
       "2                 34.0                     1.95  \n",
       "3                 20.0                     1.17  \n",
       "4                  2.0                     0.08  \n",
       "...                ...                      ...  \n",
       "144653             6.0                     0.18  \n",
       "144654            15.0                     0.94  \n",
       "144655           276.0                    18.48  \n",
       "144656             0.0                     0.00  \n",
       "144657             1.0                     0.19  \n",
       "\n",
       "[144658 rows x 7 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fill in the blanks in the enriched df with any pmids not found in icite\n",
    "test = pd.merge(df_unique_pmids, df_pmid_icite, how='left', on='pmid')\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a list of columns specific to iCite data that excludes pmid\n",
    "icite_cols =  [ 'doi',\n",
    "                'title',\n",
    "                'authors',\n",
    "                'year',\n",
    "                'citation_count',\n",
    "                'relative_citation_ratio']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>year</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>relative_citation_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36127808</td>\n",
       "      <td>10.1002/cam4.5266</td>\n",
       "      <td>Genetic ancestry, differential gene expression...</td>\n",
       "      <td>Freddy A Barragan, Lauren J Mills, Andrew R Ra...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29074302</td>\n",
       "      <td>10.1016/j.imbio.2017.10.028</td>\n",
       "      <td>Endogenous antibody responses to mucin 1 in a ...</td>\n",
       "      <td>Janardan P Pandey, Aryan M Namboodiri, Bethany...</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31387361</td>\n",
       "      <td>10.1161/CIRCULATIONAHA.119.038376</td>\n",
       "      <td>Defects in the Exocyst-Cilia Machinery Cause B...</td>\n",
       "      <td>Diana Fulmer, Katelynn Toomer, Lilong Guo, Kel...</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>1.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29027980</td>\n",
       "      <td>10.3390/genes8100269</td>\n",
       "      <td>The Plasticizer Bisphenol A Perturbs the Hepat...</td>\n",
       "      <td>Ludivine Renaud, Willian A da Silveira, E Star...</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>29309429</td>\n",
       "      <td>10.1371/journal.pone.0190949</td>\n",
       "      <td>ShinyGPA: An interactive visualization toolkit...</td>\n",
       "      <td>Emma Kortemeier, Paula S Ramos, Kelly J Hunt, ...</td>\n",
       "      <td>2018.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144653</th>\n",
       "      <td>28651562</td>\n",
       "      <td>10.1186/s12859-017-1711-z</td>\n",
       "      <td>GRAPE: a pathway template method to characteri...</td>\n",
       "      <td>Michael I Klein, David F Stern, Hongyu Zhao</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144654</th>\n",
       "      <td>31887619</td>\n",
       "      <td>10.1016/j.oraloncology.2019.104554</td>\n",
       "      <td>Transfer RNA methyltransferase gene NSUN2 mRNA...</td>\n",
       "      <td>Lingeng Lu, Stephen G Gaffney, Vincent L Canna...</td>\n",
       "      <td>2020.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144655</th>\n",
       "      <td>32251621</td>\n",
       "      <td>10.1016/S1470-2045(20)30111-X</td>\n",
       "      <td>Pembrolizumab for management of patients with ...</td>\n",
       "      <td>Sarah B Goldberg, Kurt A Schalper, Scott N Get...</td>\n",
       "      <td>2020.0</td>\n",
       "      <td>276.0</td>\n",
       "      <td>18.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144656</th>\n",
       "      <td>33148678</td>\n",
       "      <td>10.1158/1940-6207.CAPR-20-0394</td>\n",
       "      <td>Clearing the Haze: What Do We Still Need to Le...</td>\n",
       "      <td>Lisa M Fucito, Hannah Malinosky, Stephen R Bal...</td>\n",
       "      <td>2021.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144657</th>\n",
       "      <td>36509758</td>\n",
       "      <td>10.1038/s41467-022-34889-z</td>\n",
       "      <td>Brain metastatic outgrowth and osimertinib res...</td>\n",
       "      <td>Sally J Adua, Anna Arnal-Estapé, Minghui Zhao,...</td>\n",
       "      <td>2022.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>144596 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            pmid                                 doi  \\\n",
       "0       36127808                   10.1002/cam4.5266   \n",
       "1       29074302         10.1016/j.imbio.2017.10.028   \n",
       "2       31387361   10.1161/CIRCULATIONAHA.119.038376   \n",
       "3       29027980                10.3390/genes8100269   \n",
       "4       29309429        10.1371/journal.pone.0190949   \n",
       "...          ...                                 ...   \n",
       "144653  28651562           10.1186/s12859-017-1711-z   \n",
       "144654  31887619  10.1016/j.oraloncology.2019.104554   \n",
       "144655  32251621       10.1016/S1470-2045(20)30111-X   \n",
       "144656  33148678      10.1158/1940-6207.CAPR-20-0394   \n",
       "144657  36509758          10.1038/s41467-022-34889-z   \n",
       "\n",
       "                                                    title  \\\n",
       "0       Genetic ancestry, differential gene expression...   \n",
       "1       Endogenous antibody responses to mucin 1 in a ...   \n",
       "2       Defects in the Exocyst-Cilia Machinery Cause B...   \n",
       "3       The Plasticizer Bisphenol A Perturbs the Hepat...   \n",
       "4       ShinyGPA: An interactive visualization toolkit...   \n",
       "...                                                   ...   \n",
       "144653  GRAPE: a pathway template method to characteri...   \n",
       "144654  Transfer RNA methyltransferase gene NSUN2 mRNA...   \n",
       "144655  Pembrolizumab for management of patients with ...   \n",
       "144656  Clearing the Haze: What Do We Still Need to Le...   \n",
       "144657  Brain metastatic outgrowth and osimertinib res...   \n",
       "\n",
       "                                                  authors    year  \\\n",
       "0       Freddy A Barragan, Lauren J Mills, Andrew R Ra...  2023.0   \n",
       "1       Janardan P Pandey, Aryan M Namboodiri, Bethany...  2018.0   \n",
       "2       Diana Fulmer, Katelynn Toomer, Lilong Guo, Kel...  2019.0   \n",
       "3       Ludivine Renaud, Willian A da Silveira, E Star...  2017.0   \n",
       "4       Emma Kortemeier, Paula S Ramos, Kelly J Hunt, ...  2018.0   \n",
       "...                                                   ...     ...   \n",
       "144653        Michael I Klein, David F Stern, Hongyu Zhao  2017.0   \n",
       "144654  Lingeng Lu, Stephen G Gaffney, Vincent L Canna...  2020.0   \n",
       "144655  Sarah B Goldberg, Kurt A Schalper, Scott N Get...  2020.0   \n",
       "144656  Lisa M Fucito, Hannah Malinosky, Stephen R Bal...  2021.0   \n",
       "144657  Sally J Adua, Anna Arnal-Estapé, Minghui Zhao,...  2022.0   \n",
       "\n",
       "        citation_count  relative_citation_ratio  \n",
       "0                  4.0                      NaN  \n",
       "1                  2.0                     0.09  \n",
       "2                 34.0                     1.95  \n",
       "3                 20.0                     1.17  \n",
       "4                  2.0                     0.08  \n",
       "...                ...                      ...  \n",
       "144653             6.0                     0.18  \n",
       "144654            15.0                     0.94  \n",
       "144655           276.0                    18.48  \n",
       "144656             0.0                     0.00  \n",
       "144657             1.0                     0.19  \n",
       "\n",
       "[144596 rows x 7 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for rows where at last one of the iCite col values is not NaN\n",
    "test.dropna(subset=icite_cols, how='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>year</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>relative_citation_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>844</th>\n",
       "      <td>37945900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2377</th>\n",
       "      <td>31993221</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3020</th>\n",
       "      <td>37932419</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3858</th>\n",
       "      <td>37947614</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3880</th>\n",
       "      <td>37936688</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132248</th>\n",
       "      <td>31026409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138240</th>\n",
       "      <td>37945902</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140474</th>\n",
       "      <td>32123530</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142744</th>\n",
       "      <td>25513417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142892</th>\n",
       "      <td>25513250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            pmid  doi title authors  year  citation_count  \\\n",
       "844     37945900  NaN   NaN     NaN   NaN             NaN   \n",
       "2377    31993221  NaN   NaN     NaN   NaN             NaN   \n",
       "3020    37932419  NaN   NaN     NaN   NaN             NaN   \n",
       "3858    37947614  NaN   NaN     NaN   NaN             NaN   \n",
       "3880    37936688  NaN   NaN     NaN   NaN             NaN   \n",
       "...          ...  ...   ...     ...   ...             ...   \n",
       "132248  31026409  NaN   NaN     NaN   NaN             NaN   \n",
       "138240  37945902  NaN   NaN     NaN   NaN             NaN   \n",
       "140474  32123530  NaN   NaN     NaN   NaN             NaN   \n",
       "142744  25513417  NaN   NaN     NaN   NaN             NaN   \n",
       "142892  25513250  NaN   NaN     NaN   NaN             NaN   \n",
       "\n",
       "        relative_citation_ratio  \n",
       "844                         NaN  \n",
       "2377                        NaN  \n",
       "3020                        NaN  \n",
       "3858                        NaN  \n",
       "3880                        NaN  \n",
       "...                         ...  \n",
       "132248                      NaN  \n",
       "138240                      NaN  \n",
       "140474                      NaN  \n",
       "142744                      NaN  \n",
       "142892                      NaN  \n",
       "\n",
       "[62 rows x 7 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get rows where ALL icite-specific values are NaN\n",
    "test[test[icite_cols].isna().all(axis=1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>year</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>relative_citation_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>36127808</td>\n",
       "      <td>10.1002/cam4.5266</td>\n",
       "      <td>Genetic ancestry, differential gene expression...</td>\n",
       "      <td>Freddy A Barragan, Lauren J Mills, Andrew R Ra...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>30755818</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Synergistic effects of SHP2 and PI3K pathway i...</td>\n",
       "      <td>Bowen Sun, Nathaniel R Jensen, Dongjun Chung, ...</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>35895854</td>\n",
       "      <td>10.1111/biom.13727</td>\n",
       "      <td>A Bayesian multivariate mixture model for high...</td>\n",
       "      <td>Carter Allen, Yuzhou Chang, Brian Neelon, Won ...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>37715500</td>\n",
       "      <td>10.1002/sim.9911</td>\n",
       "      <td>A Bayesian framework for pathway-guided identi...</td>\n",
       "      <td>Zequn Sun, Dongjun Chung, Brian Neelon, Andrew...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>37335961</td>\n",
       "      <td>10.1200/CCI.22.00138</td>\n",
       "      <td>BatMan: Mitigating Batch Effects Via Stratific...</td>\n",
       "      <td>Ai Ni, Mengling Liu, Li-Xuan Qin</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144593</th>\n",
       "      <td>35338489</td>\n",
       "      <td>10.1111/biom.13665</td>\n",
       "      <td>Clustering high-dimensional data via feature s...</td>\n",
       "      <td>Tianqi Liu, Yu Lu, Biqing Zhu, Hongyu Zhao</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144619</th>\n",
       "      <td>37121213</td>\n",
       "      <td>10.1016/j.lungcan.2023.107211</td>\n",
       "      <td>Comparative genomics between matched solid and...</td>\n",
       "      <td>Gavitt A Woodard, Vivianne Ding, Christina Cho...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144633</th>\n",
       "      <td>36960400</td>\n",
       "      <td>10.1158/2767-9764.CRC-22-0334</td>\n",
       "      <td>Quantitative, Spatially Defined Expression of ...</td>\n",
       "      <td>Thazin N Aung, Niki Gavrielatou, Ioannis A Vat...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144634</th>\n",
       "      <td>37365174</td>\n",
       "      <td>10.1038/s41467-023-39514-1</td>\n",
       "      <td>An optogenetic-phosphoproteomic study reveals ...</td>\n",
       "      <td>Wenping Zhou, Wenxue Li, Shisheng Wang, Barbor...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144638</th>\n",
       "      <td>36913537</td>\n",
       "      <td>10.1158/1078-0432.CCR-22-3497</td>\n",
       "      <td>Efficacy of Osimertinib in Patients with Lung ...</td>\n",
       "      <td>Michael J Grant, Jacqueline V Aredo, Jacquelin...</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14654 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            pmid                            doi  \\\n",
       "0       36127808              10.1002/cam4.5266   \n",
       "7       30755818                            NaN   \n",
       "11      35895854             10.1111/biom.13727   \n",
       "23      37715500               10.1002/sim.9911   \n",
       "70      37335961           10.1200/CCI.22.00138   \n",
       "...          ...                            ...   \n",
       "144593  35338489             10.1111/biom.13665   \n",
       "144619  37121213  10.1016/j.lungcan.2023.107211   \n",
       "144633  36960400  10.1158/2767-9764.CRC-22-0334   \n",
       "144634  37365174     10.1038/s41467-023-39514-1   \n",
       "144638  36913537  10.1158/1078-0432.CCR-22-3497   \n",
       "\n",
       "                                                    title  \\\n",
       "0       Genetic ancestry, differential gene expression...   \n",
       "7       Synergistic effects of SHP2 and PI3K pathway i...   \n",
       "11      A Bayesian multivariate mixture model for high...   \n",
       "23      A Bayesian framework for pathway-guided identi...   \n",
       "70      BatMan: Mitigating Batch Effects Via Stratific...   \n",
       "...                                                   ...   \n",
       "144593  Clustering high-dimensional data via feature s...   \n",
       "144619  Comparative genomics between matched solid and...   \n",
       "144633  Quantitative, Spatially Defined Expression of ...   \n",
       "144634  An optogenetic-phosphoproteomic study reveals ...   \n",
       "144638  Efficacy of Osimertinib in Patients with Lung ...   \n",
       "\n",
       "                                                  authors    year  \\\n",
       "0       Freddy A Barragan, Lauren J Mills, Andrew R Ra...  2023.0   \n",
       "7       Bowen Sun, Nathaniel R Jensen, Dongjun Chung, ...  2019.0   \n",
       "11      Carter Allen, Yuzhou Chang, Brian Neelon, Won ...  2023.0   \n",
       "23      Zequn Sun, Dongjun Chung, Brian Neelon, Andrew...  2023.0   \n",
       "70                       Ai Ni, Mengling Liu, Li-Xuan Qin  2023.0   \n",
       "...                                                   ...     ...   \n",
       "144593         Tianqi Liu, Yu Lu, Biqing Zhu, Hongyu Zhao  2023.0   \n",
       "144619  Gavitt A Woodard, Vivianne Ding, Christina Cho...  2023.0   \n",
       "144633  Thazin N Aung, Niki Gavrielatou, Ioannis A Vat...  2023.0   \n",
       "144634  Wenping Zhou, Wenxue Li, Shisheng Wang, Barbor...  2023.0   \n",
       "144638  Michael J Grant, Jacqueline V Aredo, Jacquelin...  2023.0   \n",
       "\n",
       "        citation_count  relative_citation_ratio  \n",
       "0                  4.0                      NaN  \n",
       "7                 13.0                     0.64  \n",
       "11                 2.0                      NaN  \n",
       "23                 0.0                      NaN  \n",
       "70                 0.0                      NaN  \n",
       "...                ...                      ...  \n",
       "144593             2.0                      NaN  \n",
       "144619             1.0                      NaN  \n",
       "144633             0.0                      NaN  \n",
       "144634             0.0                      NaN  \n",
       "144638             0.0                      NaN  \n",
       "\n",
       "[14654 rows x 7 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get rows where ANY icite-specific values are NaN\n",
    "test[test[icite_cols].isna().any(axis=1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Detour to improve progress estimations with tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import zipfile\n",
    "import io\n",
    "\n",
    "def get_csv_rowcount(filepath, encoding='utf-8'):\n",
    "    \"\"\"\n",
    "    Get the number of rows in a CSV file inside a zip archive.\n",
    "\n",
    "    :param filepath (str): Path to the zipped CSV file.\n",
    "    :param encoding (str): Encoding of the CSV file. Default 'utf-8'.\n",
    "\n",
    "    :return: int: Number of rows in the CSV file.\n",
    "    \"\"\"\n",
    "\n",
    "    with zipfile.ZipFile(filepath, 'r') as zipped_file:\n",
    "        # Assuming the CSV file inside the zip has the same name as the zip file\n",
    "        csv_filename = zipped_file.namelist()[0]\n",
    "        \n",
    "        with zipped_file.open(csv_filename) as csv_file:\n",
    "            # Use io.TextIOWrapper to handle decoding from bytes to text\n",
    "            with io.TextIOWrapper(csv_file, encoding=encoding) as f:\n",
    "                return sum(1 for line in f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "icite_file_rowcount = get_csv_rowcount(icite_filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Row count:     36458403\n",
      "Chunk size:      250000\n",
      "Chunk count:        146\n"
     ]
    }
   ],
   "source": [
    "chunk_size = 250000\n",
    "print(f\"Row count:   {icite_file_rowcount:>10}\\n\"\n",
    "      f\"Chunk size:  {chunk_size:>10}\\n\"\n",
    "      f\"Chunk count: {-(-icite_file_rowcount // chunk_size):>10}\") # Divide and use negatives to round up"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Time to count the number of rows is longer than desired (~2.5min), but could still be valuable. The total number of chunks can be used with tqdm `total` to give a more accurate progress bar. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finalize the iCite function. Add step to use unique PMID dataframe so that any PMIDs without iCite data will still be listed (with NaN for all iCite columns).  \n",
    "Also add the ability to specify estimated total chunk count for the tqdm progress bar. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_icite_data_for_pmids(df_pmid, icite_filepath, cols, \n",
    "                             chunk_size=250000, chunk_count_est=None):\n",
    "    \"\"\"\n",
    "    Get iCite data for unique PMIDs from a DataFrame using chunks.\n",
    "\n",
    "    :param df_pmid (pd.DataFrame): DataFrame containing PMID-related data.\n",
    "    :param icite_filepath (str): Path to the zipped iCite CSV file.\n",
    "    :param cols (list): Columns to pull from iCite and include in output df\n",
    "    :param chunk_size (int): Number of rows to read per chunk.\n",
    "    :param chunk_count (int): Estimated number of chunks. If None, ignored\n",
    "\n",
    "    :return: pd.DataFrame: DataFrame with specified columns\n",
    "    \"\"\"\n",
    "    # Get unique PMIDs from df_pmid\n",
    "    unique_pmids = df_pmid['pmid'].unique()\n",
    "\n",
    "    # Initialize an empty list to store the enriched data chunks\n",
    "    df_enriched_chunks = []\n",
    "\n",
    "    # Create a tqdm wrapper around the generator to track progress\n",
    "    chunks = tqdm(pd.read_csv(icite_filepath, compression='zip', \n",
    "                              chunksize=chunk_size),\n",
    "                  desc=\"Processing iCite\", unit=\"chunk\", total=chunk_count_est)\n",
    "\n",
    "    # Iterate through chunks of the iCite DataFrame\n",
    "    for chunk in chunks:\n",
    "        # Filter the chunk to include only rows with PMIDs in unique_pmids\n",
    "        chunk_filtered = chunk[chunk['pmid'].isin(unique_pmids)]\n",
    "\n",
    "        # Append the filtered DataFrame to the list with selected columns\n",
    "        df_enriched_chunks.append(chunk_filtered[cols])\n",
    "\n",
    "    # Concatenate all the chunks into the final enriched DataFrame\n",
    "    df_enriched = pd.concat(df_enriched_chunks, ignore_index=True)\n",
    "\n",
    "    # Include unique PMIDs not found in iCite with NaN in iCite columns\n",
    "    df_unique_pmids = pd.DataFrame(df_pmid['pmid'].unique(), columns=['pmid'])\n",
    "    df_pmids_with_icite = pd.merge(df_unique_pmids, df_enriched, \n",
    "                                   how='left', on='pmid')\n",
    "\n",
    "    # Sort by PMID for consistent handling\n",
    "    df_pmids_with_icite.sort_values(by='pmid', inplace=True, ignore_index=True)\n",
    "\n",
    "    return df_pmids_with_icite\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing iCite:   0%|          | 0/146 [00:00<?, ?chunk/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing iCite:  98%|█████████▊| 143/146 [05:44<00:07,  2.62s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite:  99%|█████████▊| 144/146 [05:47<00:05,  2.54s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite:  99%|█████████▉| 145/146 [05:49<00:02,  2.44s/chunk]c:\\Users\\dormanza\\AppData\\Local\\anaconda3\\envs\\ins-data-env\\Lib\\site-packages\\tqdm\\std.py:1178: DtypeWarning: Columns (21,22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  for obj in iterable:\n",
      "Processing iCite: 100%|██████████| 146/146 [05:50<00:00,  2.40s/chunk]\n"
     ]
    }
   ],
   "source": [
    "df_pmids_with_icite = get_icite_data_for_pmids(df_pmid, icite_filepath, cols, \n",
    "                                               chunk_size=250000, \n",
    "                                               chunk_count_est=146)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>doi</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>year</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>relative_citation_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1279509</td>\n",
       "      <td>10.1203/00006450-199210000-00018</td>\n",
       "      <td>Expression and regulation of L-selectin on eos...</td>\n",
       "      <td>J B Smith, R D Kunjummen, T K Kishimoto, D C A...</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1280555</td>\n",
       "      <td>10.1002/cyto.990130707</td>\n",
       "      <td>Streptavidin-based quantitative staining of in...</td>\n",
       "      <td>P Srivastava, T L Sladek, M N Goodman, J W Jac...</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1281066</td>\n",
       "      <td>10.1002/cyto.990130808</td>\n",
       "      <td>Reticulocyte quantification by flow cytometry,...</td>\n",
       "      <td>K J Schimenti, K Lacerna, A Wamble, L Maston, ...</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>1.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1282437</td>\n",
       "      <td>10.1101/gr.2.2.137</td>\n",
       "      <td>Development of a sensitive reverse transcripta...</td>\n",
       "      <td>S S Tan, J H Weis</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1283327</td>\n",
       "      <td>10.1002/gcc.2870050414</td>\n",
       "      <td>Sublocalization of the chromosome 5 breakpoint...</td>\n",
       "      <td>S W Morris, J T Foust, M B Valentine, W M Robe...</td>\n",
       "      <td>1992.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144653</th>\n",
       "      <td>37947333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144654</th>\n",
       "      <td>37947334</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144655</th>\n",
       "      <td>37947335</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144656</th>\n",
       "      <td>37947337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144657</th>\n",
       "      <td>37947614</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>144658 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            pmid                               doi  \\\n",
       "0        1279509  10.1203/00006450-199210000-00018   \n",
       "1        1280555            10.1002/cyto.990130707   \n",
       "2        1281066            10.1002/cyto.990130808   \n",
       "3        1282437                10.1101/gr.2.2.137   \n",
       "4        1283327            10.1002/gcc.2870050414   \n",
       "...          ...                               ...   \n",
       "144653  37947333                               NaN   \n",
       "144654  37947334                               NaN   \n",
       "144655  37947335                               NaN   \n",
       "144656  37947337                               NaN   \n",
       "144657  37947614                               NaN   \n",
       "\n",
       "                                                    title  \\\n",
       "0       Expression and regulation of L-selectin on eos...   \n",
       "1       Streptavidin-based quantitative staining of in...   \n",
       "2       Reticulocyte quantification by flow cytometry,...   \n",
       "3       Development of a sensitive reverse transcripta...   \n",
       "4       Sublocalization of the chromosome 5 breakpoint...   \n",
       "...                                                   ...   \n",
       "144653                                                NaN   \n",
       "144654                                                NaN   \n",
       "144655                                                NaN   \n",
       "144656                                                NaN   \n",
       "144657                                                NaN   \n",
       "\n",
       "                                                  authors    year  \\\n",
       "0       J B Smith, R D Kunjummen, T K Kishimoto, D C A...  1992.0   \n",
       "1       P Srivastava, T L Sladek, M N Goodman, J W Jac...  1992.0   \n",
       "2       K J Schimenti, K Lacerna, A Wamble, L Maston, ...  1992.0   \n",
       "3                                       S S Tan, J H Weis  1992.0   \n",
       "4       S W Morris, J T Foust, M B Valentine, W M Robe...  1992.0   \n",
       "...                                                   ...     ...   \n",
       "144653                                                NaN     NaN   \n",
       "144654                                                NaN     NaN   \n",
       "144655                                                NaN     NaN   \n",
       "144656                                                NaN     NaN   \n",
       "144657                                                NaN     NaN   \n",
       "\n",
       "        citation_count  relative_citation_ratio  \n",
       "0                 25.0                     0.67  \n",
       "1                 14.0                     0.43  \n",
       "2                 38.0                     1.45  \n",
       "3                 49.0                     1.00  \n",
       "4                 14.0                     0.29  \n",
       "...                ...                      ...  \n",
       "144653             NaN                      NaN  \n",
       "144654             NaN                      NaN  \n",
       "144655             NaN                      NaN  \n",
       "144656             NaN                      NaN  \n",
       "144657             NaN                      NaN  \n",
       "\n",
       "[144658 rows x 7 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pmids_with_icite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export as csv for development checkpoint use\n",
    "df_pmids_with_icite.to_csv('pmids_with_icite_20231201.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The CSV export has some odd behavior when opened in Excel. The author list seems to overflow into the next line, causing an isolated shift in that record.  \n",
    "This seems due to very logn author lists hitting the Excel cell character limit of 32,767.  \n",
    "\n",
    "It is not a problem with the dataframe or CSV itself, and should not cause problems when loaded downstream. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "119879    True\n",
       "Name: authors, dtype: bool"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The author name 'Petrovski' is split in two when loaded in Excel. Check that it is present in df\n",
    "df_pmids_with_icite[df_pmids_with_icite['pmid'] == 33634751]['authors'].str.contains('Goran Petrovski')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ins-data-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
